<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Stephen Woloszynek" />

<meta name="date" content="2020-04-30" />

<title>Predicting Functional Content</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>



<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Predicting Functional Content</h1>
<h4 class="author">Stephen Woloszynek</h4>
<h4 class="date">2020-04-30</h4>



<p>Currently, we provide two approaches for prediction functional content from 16S rRNA amplicon abundance data: PICRUSt, which utilizes GreenGreens 13.5 assigned OTUs, and Tax4Fun, which can handle Silva assignments.</p>
<div id="picrust" class="section level2">
<h2>PICRUSt</h2>
<p>We’ll start by making functional predictions using PICRUSt on the Gevers et al. inflammatory bowel disease dataset. The dataset can be easily accessed by simply typing <strong>GEVERS</strong>, which is a list that contains an OTU table, a dataframe of metadata, and a taxonomy table. First, note that the OTU table has both rownames and column names that correspond to the metadata and taxonomy table, respectively. The latter is most important for functional prediction, since the algorithm looks for these names when mapping the OTU table to functional annotations. For PICRUSt, these names have to be OTU ids, so they’ll be long integer codes.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="kw">library</span>(themetagenomics)</a>
<a class="sourceLine" id="cb1-2" data-line-number="2"><span class="co">#&gt; Loading required package: Rcpp</span></a>
<a class="sourceLine" id="cb1-3" data-line-number="3"></a>
<a class="sourceLine" id="cb1-4" data-line-number="4">GEVERS<span class="op">$</span>OTU[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>,<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]</a>
<a class="sourceLine" id="cb1-5" data-line-number="5"><span class="co">#&gt;            4463892 4433823 288362 4381553 345362</span></a>
<a class="sourceLine" id="cb1-6" data-line-number="6"><span class="co">#&gt; SRR1567209     344       8      0       2     95</span></a>
<a class="sourceLine" id="cb1-7" data-line-number="7"><span class="co">#&gt; SRR1636130     448       1    517      43      4</span></a>
<a class="sourceLine" id="cb1-8" data-line-number="8"><span class="co">#&gt; SRR1212542     434    1239    185       3    332</span></a>
<a class="sourceLine" id="cb1-9" data-line-number="9"><span class="co">#&gt; SRR1566548       2      72    662    2263    713</span></a>
<a class="sourceLine" id="cb1-10" data-line-number="10"><span class="co">#&gt; SRR1212635     723    2479     33       3    199</span></a></code></pre></div>
<p>These column names correspond with the row names in the taxonomy table:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1">GEVERS<span class="op">$</span>TAX[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>,<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>]</a>
<a class="sourceLine" id="cb2-2" data-line-number="2"><span class="co">#&gt;         Kingdom       Phylum              Class                   </span></a>
<a class="sourceLine" id="cb2-3" data-line-number="3"><span class="co">#&gt; 4463892 &quot;k__Bacteria&quot; &quot;p__Bacteroidetes&quot;  &quot;c__Bacteroidia&quot;        </span></a>
<a class="sourceLine" id="cb2-4" data-line-number="4"><span class="co">#&gt; 4433823 &quot;k__Bacteria&quot; &quot;p__Bacteroidetes&quot;  &quot;c__Bacteroidia&quot;        </span></a>
<a class="sourceLine" id="cb2-5" data-line-number="5"><span class="co">#&gt; 288362  &quot;k__Bacteria&quot; &quot;p__Firmicutes&quot;     &quot;c__Clostridia&quot;         </span></a>
<a class="sourceLine" id="cb2-6" data-line-number="6"><span class="co">#&gt; 4381553 &quot;k__Bacteria&quot; &quot;p__Bacteroidetes&quot;  &quot;c__Bacteroidia&quot;        </span></a>
<a class="sourceLine" id="cb2-7" data-line-number="7"><span class="co">#&gt; 345362  &quot;k__Bacteria&quot; &quot;p__Proteobacteria&quot; &quot;c__Gammaproteobacteria&quot;</span></a></code></pre></div>
<p>To run PICRUSt, we first have to download the reference files. We can choose between KO terms or COG terms, or we can simply download the entire set of files, which is what the download command defaults to. We’ll stick with KO terms for now, and we’ll download it to a temporary directly. It’s probably best for you to download them to a permanant location. If you would like to download these files manually, the repo can be found here: <a href="https://gitlab.com/sw1/themetagenomics_data/" class="uri">https://gitlab.com/sw1/themetagenomics_data/</a>.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1">tmp &lt;-<span class="st"> </span><span class="kw">tempdir</span>()</a>
<a class="sourceLine" id="cb3-2" data-line-number="2"><span class="kw">download_ref</span>(tmp,<span class="dt">reference=</span><span class="st">'gg_ko'</span>,<span class="dt">overwrite=</span><span class="ot">FALSE</span>)</a></code></pre></div>
<p>We now have our GreenGenes KO terms reference file in a temporary directory for easy access. Before we perform the actual prediction, we could manually normalize for OTU copy number via the <em>cnn</em> command, but we provide an argument within the PICRUSt function to make our approach a little more streamlined. Now, we’ll run PICRUSt. Our implementation uses Rcpp, so it’s fast, but behaves analogously to the python scripts you may be familiar with.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1"><span class="kw">system.time</span>(FUNCTIONS &lt;-<span class="st"> </span><span class="kw">picrust</span>(GEVERS<span class="op">$</span>OTU,<span class="dt">rows_are_taxa=</span><span class="ot">FALSE</span>,</a>
<a class="sourceLine" id="cb4-2" data-line-number="2">                                 <span class="dt">reference=</span><span class="st">'gg_ko'</span>,<span class="dt">reference_path=</span>tmp,</a>
<a class="sourceLine" id="cb4-3" data-line-number="3">                                 <span class="dt">cn_normalize=</span><span class="ot">TRUE</span>,<span class="dt">sample_normalize=</span><span class="ot">FALSE</span>,</a>
<a class="sourceLine" id="cb4-4" data-line-number="4">                                 <span class="dt">drop=</span><span class="ot">TRUE</span>))</a>
<a class="sourceLine" id="cb4-5" data-line-number="5"><span class="co">#&gt;    user  system elapsed </span></a>
<a class="sourceLine" id="cb4-6" data-line-number="6"><span class="co">#&gt;  33.912   0.392  34.326</span></a></code></pre></div>
<p>The sample_normalize flag simply controls whether you want raw counts or relative abundances as your output. The output is another list with 3 elements: the function table that contains the KO term counts across samples, the KEGG metadata that describes the KO terms, and PICRUSt specific metadata that has the NSTI quality control score for each OTU.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1">FUNCTIONS<span class="op">$</span>fxn_table[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>,<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]</a>
<a class="sourceLine" id="cb5-2" data-line-number="2"><span class="co">#&gt;            K01361 K01362 K02249 K05844 K05845</span></a>
<a class="sourceLine" id="cb5-3" data-line-number="3"><span class="co">#&gt; SRR1567209      5   1498      0     31     55</span></a>
<a class="sourceLine" id="cb5-4" data-line-number="4"><span class="co">#&gt; SRR1636130      0    828      0      5     46</span></a>
<a class="sourceLine" id="cb5-5" data-line-number="5"><span class="co">#&gt; SRR1212542      0    815      0     66     65</span></a>
<a class="sourceLine" id="cb5-6" data-line-number="6"><span class="co">#&gt; SRR1566548      0   3791      0    169    158</span></a>
<a class="sourceLine" id="cb5-7" data-line-number="7"><span class="co">#&gt; SRR1212635     13   1037      0    760    187</span></a></code></pre></div>
<p>The metadata file is a list of lists.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1"><span class="kw">names</span>(FUNCTIONS<span class="op">$</span>fxn_meta)</a>
<a class="sourceLine" id="cb6-2" data-line-number="2"><span class="co">#&gt; [1] &quot;KEGG_Description&quot; &quot;KEGG_Pathways&quot;</span></a></code></pre></div>
<p>which contains the descriptions for each KO term</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" data-line-number="1"><span class="kw">head</span>(FUNCTIONS<span class="op">$</span>fxn_meta<span class="op">$</span>KEGG_Description)</a>
<a class="sourceLine" id="cb7-2" data-line-number="2"><span class="co">#&gt; $K01361</span></a>
<a class="sourceLine" id="cb7-3" data-line-number="3"><span class="co">#&gt; $K01361[[1]]</span></a>
<a class="sourceLine" id="cb7-4" data-line-number="4"><span class="co">#&gt; [1] &quot;lactocepin [ec:3.4.21.96]&quot;</span></a>
<a class="sourceLine" id="cb7-5" data-line-number="5"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb7-6" data-line-number="6"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb7-7" data-line-number="7"><span class="co">#&gt; $K01362</span></a>
<a class="sourceLine" id="cb7-8" data-line-number="8"><span class="co">#&gt; $K01362[[1]]</span></a>
<a class="sourceLine" id="cb7-9" data-line-number="9"><span class="co">#&gt; [1] &quot;none&quot;</span></a>
<a class="sourceLine" id="cb7-10" data-line-number="10"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb7-11" data-line-number="11"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb7-12" data-line-number="12"><span class="co">#&gt; $K02249</span></a>
<a class="sourceLine" id="cb7-13" data-line-number="13"><span class="co">#&gt; $K02249[[1]]</span></a>
<a class="sourceLine" id="cb7-14" data-line-number="14"><span class="co">#&gt; [1] &quot;competence protein comgg&quot;</span></a>
<a class="sourceLine" id="cb7-15" data-line-number="15"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb7-16" data-line-number="16"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb7-17" data-line-number="17"><span class="co">#&gt; $K05844</span></a>
<a class="sourceLine" id="cb7-18" data-line-number="18"><span class="co">#&gt; $K05844[[1]]</span></a>
<a class="sourceLine" id="cb7-19" data-line-number="19"><span class="co">#&gt; [1] &quot;ribosomal protein s6 modification protein&quot;</span></a>
<a class="sourceLine" id="cb7-20" data-line-number="20"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb7-21" data-line-number="21"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb7-22" data-line-number="22"><span class="co">#&gt; $K05845</span></a>
<a class="sourceLine" id="cb7-23" data-line-number="23"><span class="co">#&gt; $K05845[[1]]</span></a>
<a class="sourceLine" id="cb7-24" data-line-number="24"><span class="co">#&gt; [1] &quot;osmoprotectant transport system substrate-binding protein&quot;</span></a>
<a class="sourceLine" id="cb7-25" data-line-number="25"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb7-26" data-line-number="26"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb7-27" data-line-number="27"><span class="co">#&gt; $K05846</span></a>
<a class="sourceLine" id="cb7-28" data-line-number="28"><span class="co">#&gt; $K05846[[1]]</span></a>
<a class="sourceLine" id="cb7-29" data-line-number="29"><span class="co">#&gt; [1] &quot;osmoprotectant transport system permease protein&quot;</span></a></code></pre></div>
<p>and the hierarchy information</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" data-line-number="1"><span class="kw">head</span>(FUNCTIONS<span class="op">$</span>fxn_meta<span class="op">$</span>KEGG_Pathways)</a>
<a class="sourceLine" id="cb8-2" data-line-number="2"><span class="co">#&gt; $K01361</span></a>
<a class="sourceLine" id="cb8-3" data-line-number="3"><span class="co">#&gt; $K01361[[1]]</span></a>
<a class="sourceLine" id="cb8-4" data-line-number="4"><span class="co">#&gt; [1] &quot;genetic information processing&quot;   &quot;folding, sorting and degradation&quot;</span></a>
<a class="sourceLine" id="cb8-5" data-line-number="5"><span class="co">#&gt; [3] &quot;chaperones and folding catalysts&quot;</span></a>
<a class="sourceLine" id="cb8-6" data-line-number="6"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb8-7" data-line-number="7"><span class="co">#&gt; $K01361[[2]]</span></a>
<a class="sourceLine" id="cb8-8" data-line-number="8"><span class="co">#&gt; [1] &quot;metabolism&quot;      &quot;enzyme families&quot; &quot;peptidases&quot;     </span></a>
<a class="sourceLine" id="cb8-9" data-line-number="9"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb8-10" data-line-number="10"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb8-11" data-line-number="11"><span class="co">#&gt; $K01362</span></a>
<a class="sourceLine" id="cb8-12" data-line-number="12"><span class="co">#&gt; $K01362[[1]]</span></a>
<a class="sourceLine" id="cb8-13" data-line-number="13"><span class="co">#&gt; [1] &quot;unclassified&quot;          &quot;metabolism&quot;            &quot;amino acid metabolism&quot;</span></a>
<a class="sourceLine" id="cb8-14" data-line-number="14"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb8-15" data-line-number="15"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb8-16" data-line-number="16"><span class="co">#&gt; $K02249</span></a>
<a class="sourceLine" id="cb8-17" data-line-number="17"><span class="co">#&gt; $K02249[[1]]</span></a>
<a class="sourceLine" id="cb8-18" data-line-number="18"><span class="co">#&gt; [1] &quot;environmental information processing&quot;</span></a>
<a class="sourceLine" id="cb8-19" data-line-number="19"><span class="co">#&gt; [2] &quot;membrane transport&quot;                  </span></a>
<a class="sourceLine" id="cb8-20" data-line-number="20"><span class="co">#&gt; [3] &quot;secretion system&quot;                    </span></a>
<a class="sourceLine" id="cb8-21" data-line-number="21"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb8-22" data-line-number="22"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb8-23" data-line-number="23"><span class="co">#&gt; $K05844</span></a>
<a class="sourceLine" id="cb8-24" data-line-number="24"><span class="co">#&gt; $K05844[[1]]</span></a>
<a class="sourceLine" id="cb8-25" data-line-number="25"><span class="co">#&gt; [1] &quot;genetic information processing&quot; &quot;translation&quot;                   </span></a>
<a class="sourceLine" id="cb8-26" data-line-number="26"><span class="co">#&gt; [3] &quot;ribosome biogenesis&quot;           </span></a>
<a class="sourceLine" id="cb8-27" data-line-number="27"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb8-28" data-line-number="28"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb8-29" data-line-number="29"><span class="co">#&gt; $K05845</span></a>
<a class="sourceLine" id="cb8-30" data-line-number="30"><span class="co">#&gt; $K05845[[1]]</span></a>
<a class="sourceLine" id="cb8-31" data-line-number="31"><span class="co">#&gt; [1] &quot;environmental information processing&quot;</span></a>
<a class="sourceLine" id="cb8-32" data-line-number="32"><span class="co">#&gt; [2] &quot;membrane transport&quot;                  </span></a>
<a class="sourceLine" id="cb8-33" data-line-number="33"><span class="co">#&gt; [3] &quot;abc transporters&quot;                    </span></a>
<a class="sourceLine" id="cb8-34" data-line-number="34"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb8-35" data-line-number="35"><span class="co">#&gt; $K05845[[2]]</span></a>
<a class="sourceLine" id="cb8-36" data-line-number="36"><span class="co">#&gt; [1] &quot;environmental information processing&quot;</span></a>
<a class="sourceLine" id="cb8-37" data-line-number="37"><span class="co">#&gt; [2] &quot;membrane transport&quot;                  </span></a>
<a class="sourceLine" id="cb8-38" data-line-number="38"><span class="co">#&gt; [3] &quot;transporters&quot;                        </span></a>
<a class="sourceLine" id="cb8-39" data-line-number="39"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb8-40" data-line-number="40"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb8-41" data-line-number="41"><span class="co">#&gt; $K05846</span></a>
<a class="sourceLine" id="cb8-42" data-line-number="42"><span class="co">#&gt; $K05846[[1]]</span></a>
<a class="sourceLine" id="cb8-43" data-line-number="43"><span class="co">#&gt; [1] &quot;environmental information processing&quot;</span></a>
<a class="sourceLine" id="cb8-44" data-line-number="44"><span class="co">#&gt; [2] &quot;membrane transport&quot;                  </span></a>
<a class="sourceLine" id="cb8-45" data-line-number="45"><span class="co">#&gt; [3] &quot;abc transporters&quot;                    </span></a>
<a class="sourceLine" id="cb8-46" data-line-number="46"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb8-47" data-line-number="47"><span class="co">#&gt; $K05846[[2]]</span></a>
<a class="sourceLine" id="cb8-48" data-line-number="48"><span class="co">#&gt; [1] &quot;environmental information processing&quot;</span></a>
<a class="sourceLine" id="cb8-49" data-line-number="49"><span class="co">#&gt; [2] &quot;membrane transport&quot;                  </span></a>
<a class="sourceLine" id="cb8-50" data-line-number="50"><span class="co">#&gt; [3] &quot;transporters&quot;</span></a></code></pre></div>
<p>Each element in these lists are named, so they’re easily accessible. For example, say we wanted information on K05846:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" data-line-number="1">FUNCTIONS<span class="op">$</span>fxn_meta<span class="op">$</span>KEGG_Description[<span class="st">'K05846'</span>]</a>
<a class="sourceLine" id="cb9-2" data-line-number="2"><span class="co">#&gt; $K05846</span></a>
<a class="sourceLine" id="cb9-3" data-line-number="3"><span class="co">#&gt; $K05846[[1]]</span></a>
<a class="sourceLine" id="cb9-4" data-line-number="4"><span class="co">#&gt; [1] &quot;osmoprotectant transport system permease protein&quot;</span></a>
<a class="sourceLine" id="cb9-5" data-line-number="5">FUNCTIONS<span class="op">$</span>fxn_meta<span class="op">$</span>KEGG_Pathways[<span class="st">'K05846'</span>]</a>
<a class="sourceLine" id="cb9-6" data-line-number="6"><span class="co">#&gt; $K05846</span></a>
<a class="sourceLine" id="cb9-7" data-line-number="7"><span class="co">#&gt; $K05846[[1]]</span></a>
<a class="sourceLine" id="cb9-8" data-line-number="8"><span class="co">#&gt; [1] &quot;environmental information processing&quot;</span></a>
<a class="sourceLine" id="cb9-9" data-line-number="9"><span class="co">#&gt; [2] &quot;membrane transport&quot;                  </span></a>
<a class="sourceLine" id="cb9-10" data-line-number="10"><span class="co">#&gt; [3] &quot;abc transporters&quot;                    </span></a>
<a class="sourceLine" id="cb9-11" data-line-number="11"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb9-12" data-line-number="12"><span class="co">#&gt; $K05846[[2]]</span></a>
<a class="sourceLine" id="cb9-13" data-line-number="13"><span class="co">#&gt; [1] &quot;environmental information processing&quot;</span></a>
<a class="sourceLine" id="cb9-14" data-line-number="14"><span class="co">#&gt; [2] &quot;membrane transport&quot;                  </span></a>
<a class="sourceLine" id="cb9-15" data-line-number="15"><span class="co">#&gt; [3] &quot;transporters&quot;</span></a></code></pre></div>
<p>The hierarchy information is ordered based on its depth, so environmental information processing is the most general level, whereas abc transporters is the most specific.</p>
</div>
<div id="tax4fun" class="section level2">
<h2>Tax4Fun</h2>
<p>If we used Silva assignments instead of GreenGenes, we no longer can use PICRUSt. Instead, we can take advantage of Tax4Fun. For this, we’ll use the David et al. time series dataset, which is accessible via the <strong>DAVID</strong> command. Like GEVERS, DAVID is a list containing an abundance table, a metadata dataframe, and a taxonomy table. Note, however, that we are no longer working with OTU IDs (since this table was created via the Dada2 pipeline).</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" data-line-number="1">DAVID<span class="op">$</span>ABUND[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>,<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]</a>
<a class="sourceLine" id="cb10-2" data-line-number="2"><span class="co">#&gt;           00001 00002 00003 00004 00005</span></a>
<a class="sourceLine" id="cb10-3" data-line-number="3"><span class="co">#&gt; ERR531441  7232 44702     0  4562 17174</span></a>
<a class="sourceLine" id="cb10-4" data-line-number="4"><span class="co">#&gt; ERR531442  1807     0   756  5386  3899</span></a>
<a class="sourceLine" id="cb10-5" data-line-number="5"><span class="co">#&gt; ERR531443 20396 19382     0  9155 10203</span></a>
<a class="sourceLine" id="cb10-6" data-line-number="6"><span class="co">#&gt; ERR531444 14325   686     0   394 11702</span></a>
<a class="sourceLine" id="cb10-7" data-line-number="7"><span class="co">#&gt; ERR531445  3534  6751 13093  4323  3533</span></a></code></pre></div>
<p>The column names are arbitrary codes that represent the final sequences from the Dada2 error model. Of more use to us is the difference in the taxonomy names</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" data-line-number="1">DAVID<span class="op">$</span>TAX[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>,<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>]</a>
<a class="sourceLine" id="cb11-2" data-line-number="2"><span class="co">#&gt;       Kingdom    Phylum          Class        </span></a>
<a class="sourceLine" id="cb11-3" data-line-number="3"><span class="co">#&gt; 00001 &quot;Bacteria&quot; &quot;Bacteroidetes&quot; &quot;Bacteroidia&quot;</span></a>
<a class="sourceLine" id="cb11-4" data-line-number="4"><span class="co">#&gt; 00002 &quot;Bacteria&quot; &quot;Firmicutes&quot;    &quot;Clostridia&quot; </span></a>
<a class="sourceLine" id="cb11-5" data-line-number="5"><span class="co">#&gt; 00003 &quot;Bacteria&quot; &quot;Bacteroidetes&quot; &quot;Bacteroidia&quot;</span></a>
<a class="sourceLine" id="cb11-6" data-line-number="6"><span class="co">#&gt; 00004 &quot;Bacteria&quot; &quot;Firmicutes&quot;    &quot;Clostridia&quot; </span></a>
<a class="sourceLine" id="cb11-7" data-line-number="7"><span class="co">#&gt; 00005 &quot;Bacteria&quot; &quot;Bacteroidetes&quot; &quot;Bacteroidia&quot;</span></a></code></pre></div>
<p>Note that they no longer contain the taxonomy prefixes we saw in the GreenGenes assignments. To generate predictions, we again need to download reference files. We’ll use the same command, but change the reference argument to “silva_ko.”</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" data-line-number="1">tmp &lt;-<span class="st"> </span><span class="kw">tempdir</span>()</a>
<a class="sourceLine" id="cb12-2" data-line-number="2"><span class="kw">download_ref</span>(tmp,<span class="dt">reference=</span><span class="st">'silva_ko'</span>,<span class="dt">overwrite=</span><span class="ot">FALSE</span>)</a></code></pre></div>
<p>Now, we’ll use the <strong>t4f</strong> command, which takes quite a few of Tax4Fun specific arguments. We can choose the protein domain classification method performed when generating the references (UPROC or PAUDA) and whether we’d prefer to use short or long read references. Unlike PICRUSt, copy number normalization occurs with respect to the KO terms and <em>not</em> the OTUs, but in terms of the function, the argument is the same as in the picrust command. Sample normalization also occurs during the mapping step and not after predictions are made, so your decision to sample normalize may be influenced accordingly.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb13-1" data-line-number="1"><span class="kw">system.time</span>(FUNCTIONS &lt;-<span class="st"> </span><span class="kw">t4f</span>(DAVID<span class="op">$</span>ABUND,<span class="dt">rows_are_taxa=</span><span class="ot">FALSE</span>,<span class="dt">tax_table=</span>DAVID<span class="op">$</span>TAX,</a>
<a class="sourceLine" id="cb13-2" data-line-number="2">                             <span class="dt">reference_path=</span>tmp,<span class="dt">type=</span><span class="st">'uproc'</span>,<span class="dt">short=</span><span class="ot">TRUE</span>,</a>
<a class="sourceLine" id="cb13-3" data-line-number="3">                             <span class="dt">cn_normalize=</span><span class="ot">TRUE</span>,<span class="dt">sample_normalize=</span><span class="ot">TRUE</span>,<span class="dt">drop=</span><span class="ot">TRUE</span>))</a>
<a class="sourceLine" id="cb13-4" data-line-number="4"><span class="co">#&gt;    user  system elapsed </span></a>
<a class="sourceLine" id="cb13-5" data-line-number="5"><span class="co">#&gt;  10.892   0.824  11.728</span></a></code></pre></div>
<p>The output is analogous to what we saw when using PICRUSt except for the method metadata. For PICRUSt, this contained the NSTI quality score; for Tax4Fun, on the other hand, this contains the FTU scores, the fraction of OTUs that weren’t mapped to KO terms.</p>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
