<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="PLN team" />

<meta name="date" content="2020-06-21" />

<title>Data importation in PLNmodels</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
code.sourceCode > span { display: inline-block; line-height: 1.25; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>



<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Data importation in PLNmodels</h1>
<h4 class="author">PLN team</h4>
<h4 class="date">2020-06-21</h4>


<div id="TOC">
<ul>
<li><a href="#preliminaries">Preliminaries</a></li>
<li><a href="#format-description">Format description</a></li>
<li><a href="#computing-offsets">Computing offsets</a></li>
<li><a href="#building-data-frame-using-prepare_data">Building data frame using <code>prepare_data</code></a></li>
<li><a href="#importing-data-from-biom-and-phyloseq-objects-using-prepare_data_from_phyloseqbiom">Importing data from biom and phyloseq objects using <code>prepare_data_from_[phyloseq|biom]</code></a><ul>
<li><a href="#reading-from-a-biom-file">Reading from a biom file</a></li>
<li><a href="#reading-from-a-phyloseq-class-object">Reading from a phyloseq-class object</a></li>
</ul></li>
<li><a href="#mathematical-details-about-the-offsets">Mathematical details about the offsets</a><ul>
<li><a href="#total-sum-scaling">Total Sum Scaling</a></li>
<li><a href="#cumulative-sum-scaling">Cumulative Sum Scaling</a></li>
<li><a href="#relative-log-expression">Relative Log Expression</a></li>
<li><a href="#geometric-mean-of-pairwise-ratio">Geometric Mean of Pairwise Ratio</a></li>
</ul></li>
<li><a href="#references">References</a></li>
</ul>
</div>

<div id="preliminaries" class="section level2">
<h2>Preliminaries</h2>
<p>This vignette documents the data format used in <strong>PLNmodel</strong> by <code>PLN</code> and its variants. It also shows how to create an object in the proper format for further analyses from (i) tabular data, (ii) biom-class objects and (iii) phyloseq-class objects.</p>
</div>
<div id="format-description" class="section level2">
<h2>Format description</h2>
<p>We illustrate the format using trichoptera data set, a full description of which can be found in <a href="Trichoptera.html">the corresponding vignette</a>.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a><span class="kw">library</span>(PLNmodels)</span>
<span id="cb1-2"><a href="#cb1-2"></a><span class="kw">data</span>(trichoptera)</span></code></pre></div>
<p>The trichoptera data set is a list made of two data frames: <code>Abundance</code> (hereafter referred to as the <em>counts</em>) and <code>Covariate</code> (hereafter the <em>covariates</em>).</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1"></a><span class="kw">str</span>(trichoptera, <span class="dt">max.level =</span> <span class="dv">1</span>)</span></code></pre></div>
<pre><code>## List of 2
##  $ Abundance:&#39;data.frame&#39;:   49 obs. of  17 variables:
##  $ Covariate:&#39;data.frame&#39;:   49 obs. of  7 variables:</code></pre>
<p>The covariates include, among others, the wind, pressure and humidity.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1"></a><span class="kw">names</span>(trichoptera<span class="op">$</span>Covariate)</span></code></pre></div>
<pre><code>## [1] &quot;Temperature&quot;   &quot;Wind&quot;          &quot;Pressure&quot;      &quot;Humidity&quot;     
## [5] &quot;Cloudiness&quot;    &quot;Precipitation&quot; &quot;Group&quot;</code></pre>
<p>In the PLN framework, we model the counts from the covariates, let’s say wind and pressure, using a Poisson Log-Normal model. Most models in R use the so-called <em>formula interface</em> and it would thus be natural to write something like</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1"></a><span class="kw">PLN</span>(Abundance <span class="op">~</span><span class="st"> </span>Wind <span class="op">+</span><span class="st"> </span>Pressure, <span class="dt">data =</span> trichoptera)</span></code></pre></div>
<p>Unfortunately and unlike many generalized linear models, the response in PLN is intrinsically <strong>multivariate</strong>: it has 17 dimensions in our example. The left hand side (LHS) must encode a multivariate response across multiple samples, using a 2D-array (e.g. a matrix or a data frame).</p>
<p>We must therefore prepare a data structure where <code>Abundance</code> refers to a count <em>matrix</em> whereas <code>Wind</code> and <code>Pressure</code> refer to <em>vectors</em> before feeding it to <code>PLN</code>. That’s the purpose of <code>prepare_data</code>.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1"></a>trichoptera2 &lt;-<span class="st"> </span><span class="kw">prepare_data</span>(<span class="dt">counts     =</span> trichoptera<span class="op">$</span>Abundance, </span>
<span id="cb7-2"><a href="#cb7-2"></a>                             <span class="dt">covariates =</span> trichoptera<span class="op">$</span>Covariate)</span>
<span id="cb7-3"><a href="#cb7-3"></a><span class="kw">str</span>(trichoptera2)</span></code></pre></div>
<pre><code>## &#39;data.frame&#39;:    49 obs. of  9 variables:
##  $ Abundance    : num [1:49, 1:17] 0 0 0 0 0 0 0 0 0 0 ...
##   ..- attr(*, &quot;dimnames&quot;)=List of 2
##   .. ..$ : chr [1:49] &quot;1&quot; &quot;2&quot; &quot;3&quot; &quot;4&quot; ...
##   .. ..$ : chr [1:17] &quot;Che&quot; &quot;Hyc&quot; &quot;Hym&quot; &quot;Hys&quot; ...
##  $ Temperature  : num  18.7 19.8 22 23 22.5 23.9 15 17.2 15.4 14.1 ...
##  $ Wind         : num  -2.3 -2.7 -0.7 2.3 2.3 -2 -4.7 -1 -2.7 -3.7 ...
##  $ Pressure     : num  998 1000 997 991 990 ...
##  $ Humidity     : num  60 63 73 71 62 64 93 84 88 75 ...
##  $ Cloudiness   : num  19 0 6 81 50 50 100 19 69 6 ...
##  $ Precipitation: num  0 0 0 0 0 0 1.6 0 1.6 0 ...
##  $ Group        : Factor w/ 12 levels &quot;1&quot;,&quot;2&quot;,&quot;3&quot;,&quot;4&quot;,..: 1 1 1 1 1 1 1 1 1 1 ...
##  $ Offset       : num  29 13 38 192 79 18 8 34 12 4 ...</code></pre>
<p>If you look carefully, you can notice a few difference between <code>trichoptera</code> and <code>trichoptera2</code>:</p>
<ul>
<li>the first is a <code>list</code> whereas the second is a <code>data.frame</code><a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a>;</li>
<li><code>Abundance</code> is a matrix-column of <code>trichoptera2</code> that you can extract using the usual functions <code>[</code> and <code>[[</code> to retrieve the count matrix;</li>
<li><code>trichoptera2</code> has an additional <code>Offset</code> column (more on that later).</li>
</ul>
</div>
<div id="computing-offsets" class="section level2">
<h2>Computing offsets</h2>
<p>It is common practice when modeling count data to introduce an offset term to control for different sampling efforts, exposures, baselines, etc. The <em>proper way</em> to compute sample-specific offsets in still debated and may vary depending on the field. There are nevertheless a few popular methods:</p>
<ul>
<li>Total Sum Scaling (TSS), where the offset of a sample is the total count in that sample</li>
<li>Cumulative Sum Scaling (CSS), introduced in <span class="citation">(Paulson et al. <a href="#ref-CSS" role="doc-biblioref">2013</a>)</span>, where the offset of a sample if the cumulative sum of counts in that sample, up to a quantile determined in a data driven way.</li>
<li>Relative Log-Expression (RLE), implemented in <span class="citation">(Anders and Huber <a href="#ref-DESeq2" role="doc-biblioref">2010</a>)</span>, where all samples are used to compute a reference sample, each sample is compared to the reference sample using log-ratios and the offset is the median log-ratio.</li>
<li>Geometric Mean of Pairwise Ratio (GMPR), introduced in <span class="citation">(Chen et al. <a href="#ref-GMPR" role="doc-biblioref">2018</a>)</span> where each sample is compared to each other to compute a median log-ratio and the offset of a sample is the geometric means of those pairwise ratios.</li>
</ul>
<p>Each of these offset be computed from a counts matrix using the <code>compute_offset</code> function and changing its <code>offset</code> argument:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1"></a><span class="co">## same as compute_offset(trichoptera$Abundance, offset = &quot;TSS&quot;)</span></span>
<span id="cb9-2"><a href="#cb9-2"></a><span class="kw">compute_offset</span>(trichoptera<span class="op">$</span>Abundance) </span></code></pre></div>
<pre><code>##    1    2    3    4    5    6    7    8    9   10   11   12   13   14   15   16 
##   29   13   38  192   79   18    8   34   12    4    4    3   49   33  600  172 
##   17   18   19   20   21   22   23   24   25   26   27   28   29   30   31   32 
##   58   51   56  127   35   13   17    3   27   40   44    8    9 1599 2980   88 
##   33   34   35   36   37   38   39   40   41   42   43   44   45   46   47   48 
##  135  327   66   90   63   15   14   20   70   53   95   43   62  149   16   31 
##   49 
##   86</code></pre>
<p>In this particular example, the counts are too sparse and sophisticated offset methods all fail (numeric output hidden)</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1"></a><span class="kw">compute_offset</span>(trichoptera<span class="op">$</span>Abundance, <span class="st">&quot;CSS&quot;</span>)</span></code></pre></div>
<pre><code>## Warning in offset_function(counts, ...): Some samples only have 1 positive
## values. Can&#39;t compute quantiles and fall back to TSS normalization</code></pre>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1"></a><span class="kw">compute_offset</span>(trichoptera<span class="op">$</span>Abundance, <span class="st">&quot;RLE&quot;</span>)</span></code></pre></div>
<pre><code>## Warning in offset_function(counts, ...): Because of high sparsity, some samples
## have null or infinite offset.</code></pre>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1"></a><span class="kw">compute_offset</span>(trichoptera<span class="op">$</span>Abundance, <span class="st">&quot;GMPR&quot;</span>)</span></code></pre></div>
<p>We can mitigate this problem for the RLE offset by adding pseudocounts to the counts although doing so has its own drawbacks.</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1"></a><span class="kw">compute_offset</span>(trichoptera<span class="op">$</span>Abundance, <span class="st">&quot;RLE&quot;</span>, <span class="dt">pseudocounts =</span> <span class="dv">1</span>)</span></code></pre></div>
<pre><code>##         1         2         3         4         5         6         7         8 
## 0.9186270 0.8349121 0.8570257 0.9186270 0.9186270 0.8349121 0.8192245 0.8570257 
##         9        10        11        12        13        14        15        16 
## 0.7322797 0.7322797 0.6321923 0.6321923 0.9240361 0.9186270 1.3788037 0.9240361 
##        17        18        19        20        21        22        23        24 
## 0.9186270 0.9240361 0.9240361 1.7140514 0.8908577 0.8570257 0.8349121 0.6321923 
##        25        26        27        28        29        30        31        32 
## 0.9240361 0.9240361 0.9186270 0.8570257 0.8349121 2.7721084 3.2934218 0.9584503 
##        33        34        35        36        37        38        39        40 
## 1.0406547 0.9584503 0.9584503 0.9186270 0.9584503 0.8908577 0.8349121 0.7322797 
##        41        42        43        44        45        46        47        48 
## 0.9240361 0.9240361 0.9584503 0.9584503 1.2643846 1.7140514 0.8233555 0.8908577 
##        49 
## 0.9186270</code></pre>
</div>
<div id="building-data-frame-using-prepare_data" class="section level2">
<h2>Building data frame using <code>prepare_data</code></h2>
<p>We’ll already learned that <code>prepare_data</code> can join counts and covariates into a single data.frame. It can also compute offset through <code>compute_offset</code> and does so by default with <code>offset = &quot;TSS&quot;</code>, hence the <code>Offset</code> column in <code>trichoptera2</code>. You can change the offset method and provide additional arguments that will passed on to <code>compute_offset</code>.</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1"></a><span class="kw">str</span>(<span class="kw">prepare_data</span>(trichoptera<span class="op">$</span>Abundance, </span>
<span id="cb18-2"><a href="#cb18-2"></a>             trichoptera<span class="op">$</span>Covariate, </span>
<span id="cb18-3"><a href="#cb18-3"></a>             <span class="dt">offset =</span> <span class="st">&quot;RLE&quot;</span>, <span class="dt">pseudocounts =</span> <span class="dv">1</span>))</span></code></pre></div>
<pre><code>## &#39;data.frame&#39;:    49 obs. of  9 variables:
##  $ Abundance    : num [1:49, 1:17] 0 0 0 0 0 0 0 0 0 0 ...
##   ..- attr(*, &quot;dimnames&quot;)=List of 2
##   .. ..$ : chr [1:49] &quot;1&quot; &quot;2&quot; &quot;3&quot; &quot;4&quot; ...
##   .. ..$ : chr [1:17] &quot;Che&quot; &quot;Hyc&quot; &quot;Hym&quot; &quot;Hys&quot; ...
##  $ Temperature  : num  18.7 19.8 22 23 22.5 23.9 15 17.2 15.4 14.1 ...
##  $ Wind         : num  -2.3 -2.7 -0.7 2.3 2.3 -2 -4.7 -1 -2.7 -3.7 ...
##  $ Pressure     : num  998 1000 997 991 990 ...
##  $ Humidity     : num  60 63 73 71 62 64 93 84 88 75 ...
##  $ Cloudiness   : num  19 0 6 81 50 50 100 19 69 6 ...
##  $ Precipitation: num  0 0 0 0 0 0 1.6 0 1.6 0 ...
##  $ Group        : Factor w/ 12 levels &quot;1&quot;,&quot;2&quot;,&quot;3&quot;,&quot;4&quot;,..: 1 1 1 1 1 1 1 1 1 1 ...
##  $ Offset       : num  0.919 0.835 0.857 0.919 0.919 ...</code></pre>
<p>Different communities use different standard for the count data where samples are either or columns of the counts matrix. <code>prepare_data</code> uses heuristics to guess the direction of the counts matrix (or fail informatively doing so) and automatically transpose it if needed.</p>
<p>Finally, <code>prepare_data</code> enforces sample-consistency between the counts and the covariates and automatically trims away: - samples for which only covariates or only counts are available; - samples with no positive counts</p>
<p>For example, if we remove the first sample from the counts and the last one from the covariates, we end up with 49 - 2 = 47 samples left, as expected.</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1"></a><span class="kw">nrow</span>(<span class="kw">prepare_data</span>(trichoptera<span class="op">$</span>Abundance[<span class="op">-</span><span class="dv">1</span>, ], <span class="co">## remove first sample</span></span>
<span id="cb20-2"><a href="#cb20-2"></a>                  trichoptera<span class="op">$</span>Covariate[<span class="op">-</span><span class="dv">49</span>,]  <span class="co">## remove last sample</span></span>
<span id="cb20-3"><a href="#cb20-3"></a>                  ))</span></code></pre></div>
<pre><code>## [1] 47</code></pre>
</div>
<div id="importing-data-from-biom-and-phyloseq-objects-using-prepare_data_from_phyloseqbiom" class="section level2">
<h2>Importing data from biom and phyloseq objects using <code>prepare_data_from_[phyloseq|biom]</code></h2>
<p>Community composition data are quite popular in microbial ecology and usually stored in flat files using the <a href="http://biom-format.org/">biom format</a> and/or imported in R as phyloseq-class objects <span class="citation">(McMurdie <a href="#ref-phyloseq" role="doc-biblioref">2013</a>)</span> using the Bioconductor <a href="https://joey711.github.io/phyloseq/">phyloseq</a> package.</p>
<!-- We provide helper functions to directly import data from a biom file (or biom-class object) and a phyloseq-class object.  -->
<p>We show here how to import data from a biom file (or biom-class object) and form a phyloseq-class object.</p>
<div id="reading-from-a-biom-file" class="section level3">
<h3>Reading from a biom file</h3>
<p>Reading from a biom file requires the bioconductor package <a href="https://www.bioconductor.org/packages/release/bioc/html/biomformat.html">biomformat</a>. This package is <strong>not</strong> a standard dependency of PLNmodels and needs to be installed separately.</p>
<p>You can easily prepare your data from a biom file using the following steps:</p>
<ul>
<li>read your biom file with <code>biomformat::read_biom()</code></li>
<li>extract the count table with <code>biomformat::biom_data()</code></li>
<li>extract the covariates with <code>biomformat::sample_metadata()</code> (or build your own)</li>
<li>feed them to <code>prepare_data</code></li>
</ul>
<p>as illustrated below: <!-- Note that the covariates **must** be stored in the biom object as they are automatically extracted, reading a biom without covariates results in an error.  --></p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1"></a><span class="co">## If biomformat is not installed, uncomment the following lines</span></span>
<span id="cb22-2"><a href="#cb22-2"></a><span class="co"># if (!requireNamespace(&quot;BiocManager&quot;, quietly = TRUE)) {</span></span>
<span id="cb22-3"><a href="#cb22-3"></a><span class="co">#   install.packages(&quot;BiocManager&quot;)</span></span>
<span id="cb22-4"><a href="#cb22-4"></a><span class="co"># }</span></span>
<span id="cb22-5"><a href="#cb22-5"></a><span class="co"># BiocManager::install(&quot;biomformat&quot;)</span></span>
<span id="cb22-6"><a href="#cb22-6"></a><span class="kw">library</span>(biomformat)</span>
<span id="cb22-7"><a href="#cb22-7"></a>biomfile &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;rich_dense_otu_table.biom&quot;</span>, <span class="dt">package =</span> <span class="st">&quot;biomformat&quot;</span>)</span>
<span id="cb22-8"><a href="#cb22-8"></a>biom &lt;-<span class="st"> </span>biomformat<span class="op">::</span><span class="kw">read_biom</span>(biomfile)</span>
<span id="cb22-9"><a href="#cb22-9"></a><span class="co">## extract counts</span></span>
<span id="cb22-10"><a href="#cb22-10"></a>counts &lt;-<span class="st"> </span><span class="kw">as</span>(biomformat<span class="op">::</span><span class="kw">biom_data</span>(biom), <span class="st">&quot;matrix&quot;</span>)</span>
<span id="cb22-11"><a href="#cb22-11"></a><span class="co">## extract covariates (or prepare your own)</span></span>
<span id="cb22-12"><a href="#cb22-12"></a>covariates &lt;-<span class="st"> </span>biomformat<span class="op">::</span><span class="kw">sample_metadata</span>(biom)</span>
<span id="cb22-13"><a href="#cb22-13"></a><span class="co">## prepare data</span></span>
<span id="cb22-14"><a href="#cb22-14"></a>my_data &lt;-<span class="st"> </span><span class="kw">prepare_data</span>(<span class="dt">counts =</span> counts, <span class="dt">covariates =</span> covariates)</span>
<span id="cb22-15"><a href="#cb22-15"></a><span class="kw">str</span>(my_data)</span></code></pre></div>
</div>
<div id="reading-from-a-phyloseq-class-object" class="section level3">
<h3>Reading from a phyloseq-class object</h3>
<p>Likewise, preparing data from a phyloseq-class object requires the bioconductor package <a href="https://www.bioconductor.org/packages/release/bioc/html/phyloseq.html">phyloseq</a>. This package is <strong>not</strong> a standard dependency of PLNmodels and needs to be installed separately.</p>
<p>You can easily prepare your data from a phyloseq object using the following steps:</p>
<ul>
<li>extract the count table with <code>phyloseq::otu_table()</code></li>
<li>extract the covariates with <code>phyloseq::sample_data()</code> (or build your own)</li>
<li>feed them to <code>prepare_data</code></li>
</ul>
<p>as illustrated below:</p>
<!-- Note that the covariates **must** be stored in the phyloseq-class object as they are automatically extracted, importing a phyloseq object with no `sample_data` component results in an error.  -->
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1"></a><span class="co">## If biomformat is not installed, uncomment the following lines</span></span>
<span id="cb23-2"><a href="#cb23-2"></a><span class="co"># if (!requireNamespace(&quot;BiocManager&quot;, quietly = TRUE)) {</span></span>
<span id="cb23-3"><a href="#cb23-3"></a><span class="co">#   install.packages(&quot;BiocManager&quot;)</span></span>
<span id="cb23-4"><a href="#cb23-4"></a><span class="co"># }</span></span>
<span id="cb23-5"><a href="#cb23-5"></a><span class="co"># BiocManager::install(&quot;phyloseq&quot;)</span></span>
<span id="cb23-6"><a href="#cb23-6"></a><span class="kw">library</span>(phyloseq)</span>
<span id="cb23-7"><a href="#cb23-7"></a><span class="kw">data</span>(<span class="st">&quot;enterotype&quot;</span>)</span>
<span id="cb23-8"><a href="#cb23-8"></a><span class="co">## extract counts</span></span>
<span id="cb23-9"><a href="#cb23-9"></a>counts &lt;-<span class="st"> </span><span class="kw">as</span>(phyloseq<span class="op">::</span><span class="kw">otu_table</span>(enterotype), <span class="st">&quot;matrix&quot;</span>)</span>
<span id="cb23-10"><a href="#cb23-10"></a><span class="co">## extract covariates (or prepare your own)</span></span>
<span id="cb23-11"><a href="#cb23-11"></a>covariates &lt;-<span class="st"> </span>phyloseq<span class="op">::</span><span class="kw">sample_data</span>(enterotype)</span>
<span id="cb23-12"><a href="#cb23-12"></a><span class="co">## prepare data</span></span>
<span id="cb23-13"><a href="#cb23-13"></a>my_data &lt;-<span class="st"> </span><span class="kw">prepare_data</span>(<span class="dt">counts =</span> counts, <span class="dt">covariates =</span> covariates)</span>
<span id="cb23-14"><a href="#cb23-14"></a><span class="kw">str</span>(my_data)</span></code></pre></div>
</div>
</div>
<div id="mathematical-details-about-the-offsets" class="section level2">
<h2>Mathematical details about the offsets</h2>
<p>We detail here the mathematical background behind the various offsets and the way they are computed. Note <span class="math inline">\(\mathbf{C} = (c_{ij})\)</span> the counts matrix where <span class="math inline">\(c_{ij}\)</span> is the count of species <span class="math inline">\(j\)</span> in sample <span class="math inline">\(i\)</span>. Assume that there are <span class="math inline">\(J\)</span> species in total. The offset of sample <span class="math inline">\(i\)</span> is noted <span class="math inline">\(O_i\)</span> and computed in the following way.</p>
<div id="total-sum-scaling" class="section level3">
<h3>Total Sum Scaling</h3>
<p>Offsets are simply the total counts of a sample: <span class="math display">\[
O_i = \sum_{j=1}^J c_{ij}
\]</span></p>
</div>
<div id="cumulative-sum-scaling" class="section level3">
<h3>Cumulative Sum Scaling</h3>
<p>Positive counts are used to compute sample-specific quantiles <span class="math inline">\(q_i^l\)</span> and cumulative sums <span class="math inline">\(s_i^l\)</span> defined as <span class="math display">\[
q_i^l = \min \{q \text{ such that } \sum_j 1_{c_{ij} \leq q} \geq l \sum_j 1_{c_{ij} &gt; 0} \} \qquad s_i^l = \sum_{j: c_{ij} \leq q_i^l} c_{ij}
\]</span> The sample-specific quantiles are then used to compute reference quantiles defined as <span class="math inline">\(q^l = \text{median} \{q^i_l\}\)</span> and median average deviation around the quantile <span class="math inline">\(q^l\)</span> as <span class="math inline">\(d^l = \text{median} |q_i^l - q^l|\)</span>. The method then searches for the smallest quantile <span class="math inline">\(l\)</span> for which it detects instability, defined as large relative increase in the <span class="math inline">\(d^l\)</span>. Formally, <span class="math inline">\(\hat{l}\)</span> is the smallest <span class="math inline">\(l\)</span> satisfying <span class="math inline">\(\frac{d^{l+1} - d^l}{d^l} \geq 0.1\)</span>. The scaling sample-specific offset are then chosen as: <span class="math display">\[
O_i = s_i^{\hat{l}} / \text{median}_i \{ s_i^{\hat{l}} \}
\]</span> Dividing by the median of the <span class="math inline">\(s_i^{\hat{l}}\)</span> ensures that offsets are centered around <span class="math inline">\(1\)</span> and compare sizes differences with respect to the reference sample. Note also that the reference quantiles <span class="math inline">\(q^l\)</span> can be computed using either the median (default, as in the original <span class="citation">Paulson et al. (<a href="#ref-CSS" role="doc-biblioref">2013</a>)</span> paper) or the mean, by specifying <code>reference = mean</code>, as implemented in <code>metagenomeseq</code>.</p>
</div>
<div id="relative-log-expression" class="section level3">
<h3>Relative Log Expression</h3>
<p>A reference sample <span class="math inline">\((q_j)_j\)</span> is first built by computing the geometric means of each species count: <span class="math display">\[
q_j = \exp \left( \frac{1}{n} \sum_{i} \log(c_{ij})\right)
\]</span> Each sample is then compared to the reference sample to compute one ratio per species and the final offset <span class="math inline">\(O_i\)</span> is the median of those ratios: <span class="math display">\[
O_i = \text{median}_j \frac{c_{ij}}{q_j}
\]</span> The method fails when no species is shared across all sample (as all <span class="math inline">\(q_j\)</span> are then <span class="math inline">\(0\)</span>) or when a sample shares less than 50% of species with the reference (in which case the median of the ratios may be null or infinite). The problem can be alleviated by adding pseudocounts to the <span class="math inline">\(c_{ij}\)</span> with <code>pseudocounts = 1</code>.</p>
</div>
<div id="geometric-mean-of-pairwise-ratio" class="section level3">
<h3>Geometric Mean of Pairwise Ratio</h3>
<p>This method is similar to RLE but does create a reference sample. Instead, each sample is compared to each other to compute a median ratio (similar to RLE) <span class="math display">\[
r_{ii&#39;} = {\text{median}}_{j: c_{ij}.c_{i&#39;j} &gt; 0} \frac{c_{ij}}{c_{i&#39;j}}
\]</span> The offset is then taken as the median of all the <span class="math inline">\(r_{ii&#39;}\)</span>: <span class="math display">\[
O_i = \text{median}_{i&#39; != i} r_{ii&#39;}
\]</span> The method fails when there is only one sample in the data set or when a sample shares no species with any other.</p>
</div>
</div>
<div id="references" class="section level2 unnumbered">
<h2>References</h2>
<div id="refs" class="references">
<div id="ref-DESeq2">
<p>Anders, Simon, and Wolfgang Huber. 2010. “Differential Expression Analysis for Sequence Count Data.” <em>Genome Biology</em> 11 (10): R106. <a href="https://doi.org/10.1186/gb-2010-11-10-r106">https://doi.org/10.1186/gb-2010-11-10-r106</a>.</p>
</div>
<div id="ref-GMPR">
<p>Chen, Li, James Reeve, Lujun Zhang, Shengbing Huang, Xuefeng Wang, and Jun Chen. 2018. “GMPR: A Robust Normalization Method for Zero-Inflated Count Data with Application to Microbiome Sequencing Data.” <em>PeerJ</em> 6 (April): e4600. <a href="https://doi.org/10.7717/peerj.4600">https://doi.org/10.7717/peerj.4600</a>.</p>
</div>
<div id="ref-phyloseq">
<p>McMurdie, Paul J. AND Holmes. 2013. “Phyloseq: An R Package for Reproducible Interactive Analysis and Graphics of Microbiome Census Data.” <em>PLoS ONE</em> 8 (4): e61217. <a href="https://doi.org/10.1371/journal.pone.0061217">https://doi.org/10.1371/journal.pone.0061217</a>.</p>
</div>
<div id="ref-CSS">
<p>Paulson, Joseph N, O. Colin Stine, Héctor Corrada Bravo, and Mihai Pop. 2013. “Differential Abundance Analysis for Microbial Marker-Gene Surveys.” <em>Nat Methods</em> 10 (September): 1200–1202. <a href="https://doi.org/10.1038/nmeth.2658">https://doi.org/10.1038/nmeth.2658</a>.</p>
</div>
</div>
</div>
<div class="footnotes">
<hr />
<ol>
<li id="fn1"><p>although a <code>data.frame</code> is technically a <code>list</code><a href="#fnref1" class="footnote-back">↩︎</a></p></li>
</ol>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
