<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Legana Fingerhut" />

<meta name="date" content="2020-05-11" />

<title>Introduction to ampir</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>



<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Introduction to ampir</h1>
<h4 class="author">Legana Fingerhut</h4>
<h4 class="date">2020-05-11</h4>



<div id="background" class="section level1">
<h1>Background</h1>
<p>The <strong>ampir</strong> (short for <strong>a</strong>nti<strong>m</strong>icrobial <strong>p</strong>eptide prediction <strong>i</strong>n <strong>r</strong> ) package was designed to be a fast and user-friendly method to predict antimicrobial peptides (AMPs) from any given size protein dataset. <strong>ampir</strong> uses a <em>supervised statistical machine learning</em> approach to predict AMPs. It incorporates two support vector machine classification models, “precursor” and “mature” that have been trained on publicly available antimicrobial peptide data. The default model, “precursor” is best suited for full length proteins and the “mature” model is best suited for small mature proteins (&lt;60 amino acids). <strong>ampir</strong> also accepts custom (user trained) models based on the <a href="https://github.com/topepo/caret">caret</a> package. Please see the <strong>ampir</strong> <em>“How to train your model”</em> vignette for details.</p>
<div id="usage" class="section level2">
<h2>Usage</h2>
<p>Standard input to <strong>ampir</strong> is a <code>data.frame</code> with sequence names in the first column and protein sequences in the second column.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="kw">library</span>(ampir)</a></code></pre></div>
<p>Read in a FASTA formatted file as a <code>data.frame</code> with <code>read_faa()</code></p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1">my_protein_df &lt;-<span class="st"> </span><span class="kw">read_faa</span>(<span class="kw">system.file</span>(<span class="st">&quot;extdata/little_test.fasta&quot;</span>, <span class="dt">package =</span> <span class="st">&quot;ampir&quot;</span>))</a></code></pre></div>
<table>
<thead>
<tr class="header">
<th align="left">seq_name</th>
<th align="left">seq_aa</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">tr|G1P6H5|G1P6H5_MYOLU</td>
<td align="left">MALTVRIQAACLLLLLLASLTSYSLLLSQTTQLADLQTQDTAGAT…</td>
</tr>
<tr class="even">
<td align="left">tr|L5L3D0|L5L3D0_PTEAL</td>
<td align="left">MKPLLIVFVFLIFWDPALAGLNPISSEMYKKCYGNGICRLECYTS…</td>
</tr>
<tr class="odd">
<td align="left">tr|A0A183U1F1|A0A183U1F1_TOXCA</td>
<td align="left">LLRLYSPLVMFATRRVLLCLLVIYLLAQPIHSSWLKKTYKKLENS…</td>
</tr>
<tr class="even">
<td align="left">tr|Q5F4I1|Q5F4I1_DROPS</td>
<td align="left">MNFYKIFIFVALILAISVGQSEAGWLKKLGKRLERVGQHTRDATI…</td>
</tr>
<tr class="odd">
<td align="left">tr|A7S075|A7S075_NEMVE</td>
<td align="left">MFLKVVVVLLAVELSVAQSARQRVRPLDRKAGRKRFAPIFPRQCS…</td>
</tr>
<tr class="even">
<td align="left">tr|F1DFM9|F1DFM9_9CNID</td>
<td align="left">MKVLVILFGAMLVLMEFQKASAATLLEDFDDDDDLLDDGGDFDLE…</td>
</tr>
<tr class="odd">
<td align="left">tr|Q5XV93|Q5XV93_ARATH</td>
<td align="left">MSKREYERQLANEEDEQLRNFQAAVAARSAILHEPKEAALPPPAP…</td>
</tr>
<tr class="even">
<td align="left">tr|Q2XXN9|Q2XXN9_POGBA</td>
<td align="left">MRFLYLLFAVAFLFSVQAEDAELEQEQQGDPWEGLDEFQDQPPDD…</td>
</tr>
</tbody>
</table>
<p>Calculate the probability that each protein is an antimicrobial peptide with <code>predict_amps()</code> using the default “precursor” model.</p>
<p><em>Note that amino acid sequences that are shorter than 10 amino acids long and/or contain anything other than the standard 20 amino acids are not evaluated and will contain an <code>NA</code> as their <code>prob_AMP</code> value.</em></p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1">my_prediction &lt;-<span class="st"> </span><span class="kw">predict_amps</span>(my_protein_df, <span class="dt">model =</span> <span class="st">&quot;precursor&quot;</span>)</a></code></pre></div>
<table>
<thead>
<tr class="header">
<th align="left">seq_name</th>
<th align="left">seq_aa</th>
<th align="right">prob_AMP</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">tr|G1P6H5|G1P6H5_MYOLU</td>
<td align="left">MALTVRIQAACLLLLLLASLTSYSLLLSQTTQLADLQTQDTAGAT…</td>
<td align="right">0.359</td>
</tr>
<tr class="even">
<td align="left">tr|L5L3D0|L5L3D0_PTEAL</td>
<td align="left">MKPLLIVFVFLIFWDPALAGLNPISSEMYKKCYGNGICRLECYTS…</td>
<td align="right">0.839</td>
</tr>
<tr class="odd">
<td align="left">tr|A0A183U1F1|A0A183U1F1_TOXCA</td>
<td align="left">LLRLYSPLVMFATRRVLLCLLVIYLLAQPIHSSWLKKTYKKLENS…</td>
<td align="right">0.019</td>
</tr>
<tr class="even">
<td align="left">tr|Q5F4I1|Q5F4I1_DROPS</td>
<td align="left">MNFYKIFIFVALILAISVGQSEAGWLKKLGKRLERVGQHTRDATI…</td>
<td align="right">0.986</td>
</tr>
<tr class="odd">
<td align="left">tr|A7S075|A7S075_NEMVE</td>
<td align="left">MFLKVVVVLLAVELSVAQSARQRVRPLDRKAGRKRFAPIFPRQCS…</td>
<td align="right">0.023</td>
</tr>
<tr class="even">
<td align="left">tr|F1DFM9|F1DFM9_9CNID</td>
<td align="left">MKVLVILFGAMLVLMEFQKASAATLLEDFDDDDDLLDDGGDFDLE…</td>
<td align="right">0.237</td>
</tr>
<tr class="odd">
<td align="left">tr|Q5XV93|Q5XV93_ARATH</td>
<td align="left">MSKREYERQLANEEDEQLRNFQAAVAARSAILHEPKEAALPPPAP…</td>
<td align="right">0.010</td>
</tr>
<tr class="even">
<td align="left">tr|Q2XXN9|Q2XXN9_POGBA</td>
<td align="left">MRFLYLLFAVAFLFSVQAEDAELEQEQQGDPWEGLDEFQDQPPDD…</td>
<td align="right">0.650</td>
</tr>
</tbody>
</table>
<p>Predicted proteins with a specified predicted probability value could then be extracted and written to a FASTA file:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1">my_predicted_amps &lt;-<span class="st"> </span>my_protein_df[my_prediction<span class="op">$</span>prob_AMP <span class="op">&gt;</span><span class="st"> </span><span class="fl">0.8</span>,]</a></code></pre></div>
<table>
<thead>
<tr class="header">
<th></th>
<th align="left">seq_name</th>
<th align="left">seq_aa</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>2</td>
<td align="left">tr|L5L3D0|L5L3D0_PTEAL</td>
<td align="left">MKPLLIVFVFLIFWDPALAGLNPISSEMYKKCYGNGICRLECYTS…</td>
</tr>
<tr class="even">
<td>4</td>
<td align="left">tr|Q5F4I1|Q5F4I1_DROPS</td>
<td align="left">MNFYKIFIFVALILAISVGQSEAGWLKKLGKRLERVGQHTRDATI…</td>
</tr>
</tbody>
</table>
<p>Write the <code>data.frame</code> with sequence names in the first column and protein sequences in the second column to a FASTA formatted file with <code>df_to_faa()</code></p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1"><span class="kw">df_to_faa</span>(my_predicted_amps, <span class="kw">tempfile</span>(<span class="st">&quot;my_predicted_amps.fasta&quot;</span>, <span class="kw">tempdir</span>()))</a></code></pre></div>
</div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
