<!DOCTYPE html>
<html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8"/>

<title>Introduction</title>

<script type="text/javascript">
window.onload = function() {
  var imgs = document.getElementsByTagName('img'), i, img;
  for (i = 0; i < imgs.length; i++) {
    img = imgs[i];
    // center an image if it is the only element of its parent
    if (img.parentElement.childElementCount === 1)
      img.parentElement.style.textAlign = 'center';
  }
};
</script>

<!-- Styles for R syntax highlighter -->
<style type="text/css">
   pre .operator,
   pre .paren {
     color: rgb(104, 118, 135)
   }

   pre .literal {
     color: #990073
   }

   pre .number {
     color: #099;
   }

   pre .comment {
     color: #998;
     font-style: italic
   }

   pre .keyword {
     color: #900;
     font-weight: bold
   }

   pre .identifier {
     color: rgb(0, 0, 0);
   }

   pre .string {
     color: #d14;
   }
</style>

<!-- R syntax highlighter -->
<script type="text/javascript">
var hljs=new function(){function m(p){return p.replace(/&/gm,"&amp;").replace(/</gm,"&lt;")}function f(r,q,p){return RegExp(q,"m"+(r.cI?"i":"")+(p?"g":""))}function b(r){for(var p=0;p<r.childNodes.length;p++){var q=r.childNodes[p];if(q.nodeName=="CODE"){return q}if(!(q.nodeType==3&&q.nodeValue.match(/\s+/))){break}}}function h(t,s){var p="";for(var r=0;r<t.childNodes.length;r++){if(t.childNodes[r].nodeType==3){var q=t.childNodes[r].nodeValue;if(s){q=q.replace(/\n/g,"")}p+=q}else{if(t.childNodes[r].nodeName=="BR"){p+="\n"}else{p+=h(t.childNodes[r])}}}if(/MSIE [678]/.test(navigator.userAgent)){p=p.replace(/\r/g,"\n")}return p}function a(s){var r=s.className.split(/\s+/);r=r.concat(s.parentNode.className.split(/\s+/));for(var q=0;q<r.length;q++){var p=r[q].replace(/^language-/,"");if(e[p]){return p}}}function c(q){var p=[];(function(s,t){for(var r=0;r<s.childNodes.length;r++){if(s.childNodes[r].nodeType==3){t+=s.childNodes[r].nodeValue.length}else{if(s.childNodes[r].nodeName=="BR"){t+=1}else{if(s.childNodes[r].nodeType==1){p.push({event:"start",offset:t,node:s.childNodes[r]});t=arguments.callee(s.childNodes[r],t);p.push({event:"stop",offset:t,node:s.childNodes[r]})}}}}return t})(q,0);return p}function k(y,w,x){var q=0;var z="";var s=[];function u(){if(y.length&&w.length){if(y[0].offset!=w[0].offset){return(y[0].offset<w[0].offset)?y:w}else{return w[0].event=="start"?y:w}}else{return y.length?y:w}}function t(D){var A="<"+D.nodeName.toLowerCase();for(var B=0;B<D.attributes.length;B++){var C=D.attributes[B];A+=" "+C.nodeName.toLowerCase();if(C.value!==undefined&&C.value!==false&&C.value!==null){A+='="'+m(C.value)+'"'}}return A+">"}while(y.length||w.length){var v=u().splice(0,1)[0];z+=m(x.substr(q,v.offset-q));q=v.offset;if(v.event=="start"){z+=t(v.node);s.push(v.node)}else{if(v.event=="stop"){var p,r=s.length;do{r--;p=s[r];z+=("</"+p.nodeName.toLowerCase()+">")}while(p!=v.node);s.splice(r,1);while(r<s.length){z+=t(s[r]);r++}}}}return z+m(x.substr(q))}function j(){function q(x,y,v){if(x.compiled){return}var u;var s=[];if(x.k){x.lR=f(y,x.l||hljs.IR,true);for(var w in x.k){if(!x.k.hasOwnProperty(w)){continue}if(x.k[w] instanceof Object){u=x.k[w]}else{u=x.k;w="keyword"}for(var r in u){if(!u.hasOwnProperty(r)){continue}x.k[r]=[w,u[r]];s.push(r)}}}if(!v){if(x.bWK){x.b="\\b("+s.join("|")+")\\s"}x.bR=f(y,x.b?x.b:"\\B|\\b");if(!x.e&&!x.eW){x.e="\\B|\\b"}if(x.e){x.eR=f(y,x.e)}}if(x.i){x.iR=f(y,x.i)}if(x.r===undefined){x.r=1}if(!x.c){x.c=[]}x.compiled=true;for(var t=0;t<x.c.length;t++){if(x.c[t]=="self"){x.c[t]=x}q(x.c[t],y,false)}if(x.starts){q(x.starts,y,false)}}for(var p in e){if(!e.hasOwnProperty(p)){continue}q(e[p].dM,e[p],true)}}function d(B,C){if(!j.called){j();j.called=true}function q(r,M){for(var L=0;L<M.c.length;L++){if((M.c[L].bR.exec(r)||[null])[0]==r){return M.c[L]}}}function v(L,r){if(D[L].e&&D[L].eR.test(r)){return 1}if(D[L].eW){var M=v(L-1,r);return M?M+1:0}return 0}function w(r,L){return L.i&&L.iR.test(r)}function K(N,O){var M=[];for(var L=0;L<N.c.length;L++){M.push(N.c[L].b)}var r=D.length-1;do{if(D[r].e){M.push(D[r].e)}r--}while(D[r+1].eW);if(N.i){M.push(N.i)}return f(O,M.join("|"),true)}function p(M,L){var N=D[D.length-1];if(!N.t){N.t=K(N,E)}N.t.lastIndex=L;var r=N.t.exec(M);return r?[M.substr(L,r.index-L),r[0],false]:[M.substr(L),"",true]}function z(N,r){var L=E.cI?r[0].toLowerCase():r[0];var M=N.k[L];if(M&&M instanceof Array){return M}return false}function F(L,P){L=m(L);if(!P.k){return L}var r="";var O=0;P.lR.lastIndex=0;var M=P.lR.exec(L);while(M){r+=L.substr(O,M.index-O);var N=z(P,M);if(N){x+=N[1];r+='<span class="'+N[0]+'">'+M[0]+"</span>"}else{r+=M[0]}O=P.lR.lastIndex;M=P.lR.exec(L)}return r+L.substr(O,L.length-O)}function J(L,M){if(M.sL&&e[M.sL]){var r=d(M.sL,L);x+=r.keyword_count;return r.value}else{return F(L,M)}}function I(M,r){var L=M.cN?'<span class="'+M.cN+'">':"";if(M.rB){y+=L;M.buffer=""}else{if(M.eB){y+=m(r)+L;M.buffer=""}else{y+=L;M.buffer=r}}D.push(M);A+=M.r}function G(N,M,Q){var R=D[D.length-1];if(Q){y+=J(R.buffer+N,R);return false}var P=q(M,R);if(P){y+=J(R.buffer+N,R);I(P,M);return P.rB}var L=v(D.length-1,M);if(L){var O=R.cN?"</span>":"";if(R.rE){y+=J(R.buffer+N,R)+O}else{if(R.eE){y+=J(R.buffer+N,R)+O+m(M)}else{y+=J(R.buffer+N+M,R)+O}}while(L>1){O=D[D.length-2].cN?"</span>":"";y+=O;L--;D.length--}var r=D[D.length-1];D.length--;D[D.length-1].buffer="";if(r.starts){I(r.starts,"")}return R.rE}if(w(M,R)){throw"Illegal"}}var E=e[B];var D=[E.dM];var A=0;var x=0;var y="";try{var s,u=0;E.dM.buffer="";do{s=p(C,u);var t=G(s[0],s[1],s[2]);u+=s[0].length;if(!t){u+=s[1].length}}while(!s[2]);if(D.length>1){throw"Illegal"}return{r:A,keyword_count:x,value:y}}catch(H){if(H=="Illegal"){return{r:0,keyword_count:0,value:m(C)}}else{throw H}}}function g(t){var p={keyword_count:0,r:0,value:m(t)};var r=p;for(var q in e){if(!e.hasOwnProperty(q)){continue}var s=d(q,t);s.language=q;if(s.keyword_count+s.r>r.keyword_count+r.r){r=s}if(s.keyword_count+s.r>p.keyword_count+p.r){r=p;p=s}}if(r.language){p.second_best=r}return p}function i(r,q,p){if(q){r=r.replace(/^((<[^>]+>|\t)+)/gm,function(t,w,v,u){return w.replace(/\t/g,q)})}if(p){r=r.replace(/\n/g,"<br>")}return r}function n(t,w,r){var x=h(t,r);var v=a(t);var y,s;if(v){y=d(v,x)}else{return}var q=c(t);if(q.length){s=document.createElement("pre");s.innerHTML=y.value;y.value=k(q,c(s),x)}y.value=i(y.value,w,r);var u=t.className;if(!u.match("(\\s|^)(language-)?"+v+"(\\s|$)")){u=u?(u+" "+v):v}if(/MSIE [678]/.test(navigator.userAgent)&&t.tagName=="CODE"&&t.parentNode.tagName=="PRE"){s=t.parentNode;var p=document.createElement("div");p.innerHTML="<pre><code>"+y.value+"</code></pre>";t=p.firstChild.firstChild;p.firstChild.cN=s.cN;s.parentNode.replaceChild(p.firstChild,s)}else{t.innerHTML=y.value}t.className=u;t.result={language:v,kw:y.keyword_count,re:y.r};if(y.second_best){t.second_best={language:y.second_best.language,kw:y.second_best.keyword_count,re:y.second_best.r}}}function o(){if(o.called){return}o.called=true;var r=document.getElementsByTagName("pre");for(var p=0;p<r.length;p++){var q=b(r[p]);if(q){n(q,hljs.tabReplace)}}}function l(){if(window.addEventListener){window.addEventListener("DOMContentLoaded",o,false);window.addEventListener("load",o,false)}else{if(window.attachEvent){window.attachEvent("onload",o)}else{window.onload=o}}}var e={};this.LANGUAGES=e;this.highlight=d;this.highlightAuto=g;this.fixMarkup=i;this.highlightBlock=n;this.initHighlighting=o;this.initHighlightingOnLoad=l;this.IR="[a-zA-Z][a-zA-Z0-9_]*";this.UIR="[a-zA-Z_][a-zA-Z0-9_]*";this.NR="\\b\\d+(\\.\\d+)?";this.CNR="\\b(0[xX][a-fA-F0-9]+|(\\d+(\\.\\d*)?|\\.\\d+)([eE][-+]?\\d+)?)";this.BNR="\\b(0b[01]+)";this.RSR="!|!=|!==|%|%=|&|&&|&=|\\*|\\*=|\\+|\\+=|,|\\.|-|-=|/|/=|:|;|<|<<|<<=|<=|=|==|===|>|>=|>>|>>=|>>>|>>>=|\\?|\\[|\\{|\\(|\\^|\\^=|\\||\\|=|\\|\\||~";this.ER="(?![\\s\\S])";this.BE={b:"\\\\.",r:0};this.ASM={cN:"string",b:"'",e:"'",i:"\\n",c:[this.BE],r:0};this.QSM={cN:"string",b:'"',e:'"',i:"\\n",c:[this.BE],r:0};this.CLCM={cN:"comment",b:"//",e:"$"};this.CBLCLM={cN:"comment",b:"/\\*",e:"\\*/"};this.HCM={cN:"comment",b:"#",e:"$"};this.NM={cN:"number",b:this.NR,r:0};this.CNM={cN:"number",b:this.CNR,r:0};this.BNM={cN:"number",b:this.BNR,r:0};this.inherit=function(r,s){var p={};for(var q in r){p[q]=r[q]}if(s){for(var q in s){p[q]=s[q]}}return p}}();hljs.LANGUAGES.cpp=function(){var a={keyword:{"false":1,"int":1,"float":1,"while":1,"private":1,"char":1,"catch":1,"export":1,virtual:1,operator:2,sizeof:2,dynamic_cast:2,typedef:2,const_cast:2,"const":1,struct:1,"for":1,static_cast:2,union:1,namespace:1,unsigned:1,"long":1,"throw":1,"volatile":2,"static":1,"protected":1,bool:1,template:1,mutable:1,"if":1,"public":1,friend:2,"do":1,"return":1,"goto":1,auto:1,"void":2,"enum":1,"else":1,"break":1,"new":1,extern:1,using:1,"true":1,"class":1,asm:1,"case":1,typeid:1,"short":1,reinterpret_cast:2,"default":1,"double":1,register:1,explicit:1,signed:1,typename:1,"try":1,"this":1,"switch":1,"continue":1,wchar_t:1,inline:1,"delete":1,alignof:1,char16_t:1,char32_t:1,constexpr:1,decltype:1,noexcept:1,nullptr:1,static_assert:1,thread_local:1,restrict:1,_Bool:1,complex:1},built_in:{std:1,string:1,cin:1,cout:1,cerr:1,clog:1,stringstream:1,istringstream:1,ostringstream:1,auto_ptr:1,deque:1,list:1,queue:1,stack:1,vector:1,map:1,set:1,bitset:1,multiset:1,multimap:1,unordered_set:1,unordered_map:1,unordered_multiset:1,unordered_multimap:1,array:1,shared_ptr:1}};return{dM:{k:a,i:"</",c:[hljs.CLCM,hljs.CBLCLM,hljs.QSM,{cN:"string",b:"'\\\\?.",e:"'",i:"."},{cN:"number",b:"\\b(\\d+(\\.\\d*)?|\\.\\d+)(u|U|l|L|ul|UL|f|F)"},hljs.CNM,{cN:"preprocessor",b:"#",e:"$"},{cN:"stl_container",b:"\\b(deque|list|queue|stack|vector|map|set|bitset|multiset|multimap|unordered_map|unordered_set|unordered_multiset|unordered_multimap|array)\\s*<",e:">",k:a,r:10,c:["self"]}]}}}();hljs.LANGUAGES.r={dM:{c:[hljs.HCM,{cN:"number",b:"\\b0[xX][0-9a-fA-F]+[Li]?\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\b\\d+(?:[eE][+\\-]?\\d*)?L\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\b\\d+\\.(?!\\d)(?:i\\b)?",e:hljs.IMMEDIATE_RE,r:1},{cN:"number",b:"\\b\\d+(?:\\.\\d*)?(?:[eE][+\\-]?\\d*)?i?\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\.\\d+(?:[eE][+\\-]?\\d*)?i?\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"keyword",b:"(?:tryCatch|library|setGeneric|setGroupGeneric)\\b",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\.\\.\\.",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\.\\.\\d+(?![\\w.])",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\b(?:function)",e:hljs.IMMEDIATE_RE,r:2},{cN:"keyword",b:"(?:if|in|break|next|repeat|else|for|return|switch|while|try|stop|warning|require|attach|detach|source|setMethod|setClass)\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"literal",b:"(?:NA|NA_integer_|NA_real_|NA_character_|NA_complex_)\\b",e:hljs.IMMEDIATE_RE,r:10},{cN:"literal",b:"(?:NULL|TRUE|FALSE|T|F|Inf|NaN)\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"identifier",b:"[a-zA-Z.][a-zA-Z0-9._]*\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"operator",b:"<\\-(?!\\s*\\d)",e:hljs.IMMEDIATE_RE,r:2},{cN:"operator",b:"\\->|<\\-",e:hljs.IMMEDIATE_RE,r:1},{cN:"operator",b:"%%|~",e:hljs.IMMEDIATE_RE},{cN:"operator",b:">=|<=|==|!=|\\|\\||&&|=|\\+|\\-|\\*|/|\\^|>|<|!|&|\\||\\$|:",e:hljs.IMMEDIATE_RE,r:0},{cN:"operator",b:"%",e:"%",i:"\\n",r:1},{cN:"identifier",b:"`",e:"`",r:0},{cN:"string",b:'"',e:'"',c:[hljs.BE],r:0},{cN:"string",b:"'",e:"'",c:[hljs.BE],r:0},{cN:"paren",b:"[[({\\])}]",e:hljs.IMMEDIATE_RE,r:0}]}};
hljs.initHighlightingOnLoad();
</script>

<!-- MathJax scripts -->
<script type="text/javascript" src="https://cdn.bootcss.com/mathjax/2.7.0/MathJax.js?config=TeX-MML-AM_CHTML">
</script>


<style type="text/css">
body, td {
   font-family: sans-serif;
   background-color: white;
   font-size: 13px;
}

body {
  max-width: 800px;
  margin: auto;
  padding: 1em;
  line-height: 20px;
}

tt, code, pre {
   font-family: 'DejaVu Sans Mono', 'Droid Sans Mono', 'Lucida Console', Consolas, Monaco, monospace;
}

h1 {
   font-size:2.2em;
}

h2 {
   font-size:1.8em;
}

h3 {
   font-size:1.4em;
}

h4 {
   font-size:1.0em;
}

h5 {
   font-size:0.9em;
}

h6 {
   font-size:0.8em;
}

a:visited {
   color: rgb(50%, 0%, 50%);
}

pre, img {
  max-width: 100%;
}
pre {
  overflow-x: auto;
}
pre code {
   display: block; padding: 0.5em;
}

code {
  font-size: 92%;
  border: 1px solid #ccc;
}

code[class] {
  background-color: #F8F8F8;
}

table, td, th {
  border: none;
}

blockquote {
   color:#666666;
   margin:0;
   padding-left: 1em;
   border-left: 0.5em #EEE solid;
}

hr {
   height: 0px;
   border-bottom: none;
   border-top-width: thin;
   border-top-style: dotted;
   border-top-color: #999999;
}

@media print {
   * {
      background: transparent !important;
      color: black !important;
      filter:none !important;
      -ms-filter: none !important;
   }

   body {
      font-size:12pt;
      max-width:100%;
   }

   a, a:visited {
      text-decoration: underline;
   }

   hr {
      visibility: hidden;
      page-break-before: always;
   }

   pre, blockquote {
      padding-right: 1em;
      page-break-inside: avoid;
   }

   tr, img {
      page-break-inside: avoid;
   }

   img {
      max-width: 100% !important;
   }

   @page :left {
      margin: 15mm 20mm 15mm 10mm;
   }

   @page :right {
      margin: 15mm 10mm 15mm 20mm;
   }

   p, h2, h3 {
      orphans: 3; widows: 3;
   }

   h2, h3 {
      page-break-after: avoid;
   }
}
</style>



</head>

<body>
<h2>Introduction</h2>

<p>Ravages was developped to simulate genetic data and to perform rare variant association tests (burden tests and the variance-component test SKAT) on more than two groups of individuals (Bocher et al., 2019, doi:10.1002/gepi.22210). Ravages relies on the package Gaston. Most functions are written in C++ thanks to the packages Rcpp, RcppParallel and RcppEigen.<br/>
Functions of Ravages use bed.matrix to manipulate genetic data as in the package Gaston (see documentation of this package for more details).<br/>
In this vignette, we illustrate how to perform genetic simulations and how to use it to compute power of rare variant association tests. See the main vignette for more details about rare variant association tests.<br/>
To learn more about all options of the functions, the reader is advised to look at the manual pages.<br/>
We develpped two main simulations procedures, one based on allelic frequencies and genetic relative risks (GRR), and the other one based on haplotypes. All the functions can be used to simulate more than two groups of individuals.</p>

<h2>Simulations based on allelic frequencies and GRR</h2>

<h3>Calculation of frequencies in each group of individuals</h3>

<p>The first step to simulate genetic data is to compute genotypic frequencies in each group of individuals based on frequencies in the general population and on genetic relative risk (GRR) values. GRR correspond to the increased risk of a disease for a given genotype compared to a reference genotype, here the homozygous genotype for the reference allele. More precisely, the GRR associated to the heterozygous genotype in the group \(c\) can be calculated as follow: \[GRR_{Aa} = \frac{P(Y=c | Aa)}{P(Y=c | AA)}\] With Y the phenotype (1 for the controls, and c going from 2 to C in the cases), \(A\) the reference allele, and \(a\) the alternate allele. The frequency of each genotype in each group of cases \(c\) can be calculated using Bayes theorem:
\[P(Aa | Y=c) = \frac{P(Y=c | Aa) P(Aa)}{\sum_{Geno=AA, Aa, aa} P(Y=c | Geno) P(Geno)} = \frac{GRR_{Aa} \times P(Aa)}{P(AA)+GRR_{Aa} \times P(Aa)+GRR_{aa} \times P(aa)}\]
P(AA), P(Aa) and P(aa) corresponding to the genotypic probabilities in the general population.
The three genotypic frequencies can then be calculated in the controls group using the rule of total probability:
\[P(Geno | Y=1)=P(Geno) - \sum{_{c=2}^{c=C} P(Geno | Y=c) P(Y=c)}\]
The function <strong>genotypic.freq()</strong> performs these calculations to obtain the three genotypic frequencies in the different groups of individuals. To do so, the user needs to give <em>P(Y=c)</em>, the prevalence of each group of cases (argument <em>prev</em>), and the GRR values. GRR values need to be in a matrix form with one row per cases group and one column per variant. If there is no supposed link between the GRR associated to the heterozygous genotype and the GRR associated to the homozygous alternative genotype (general model of the disease, \(genetic.model=&ldquo;general&rdquo;\)), the user needs to specify two GRR matrices: one for \(GRR_{Aa}\) (argument <em>GRR.het</em>) and one for \(GRR_{aa}\) (argument <em>GRR.homo.alt</em>). If <em>genetic.model=&ldquo;recessive&rdquo;</em>, <em>&ldquo;multiplicative&rdquo;</em> or <em>&ldquo;dominant&rdquo;</em>, only one GRR matrix is needed. <strong>genotypic.freq()</strong> will return a list with three matrices, one for each genotype containing the genotypic frequencies, with one row per group of individuals and one column per variant.<br/>
To help the user with the construction of the GRR matrix for <strong>genotypic.freq()</strong>, we implemented the function <strong>GRR.matrix()</strong>.<br/>
To use this function, the user needs to specify how the GRR should be calculated (argument <em>GRR</em>):  </p>

<ul>
<li>the same GRR is given to all the variants (<em>GRR=&ldquo;constant&rdquo;</em>), its value being specified to <em>GRR.value</em>;<br/></li>
<li>the GRR is computed using the same formula as in SKAT: \(w=-0.4*|log_{10}(MAF)|\) (<em>GRR=&ldquo;SKAT&rdquo;</em>);<br/></li>
<li>the GRR is computed using another function depending on MAF in the general population (<em>GRR=&ldquo;variable&rdquo;</em>), this function being specified to <em>GRR.function</em>.<br/></li>
</ul>

<p>In the two last situations, a file containing the MAF in the general population with at least a column &ldquo;maf&rdquo; and a column &ldquo;gene&rdquo; should be given to the argument <em>genes.maf</em>. Two such files are available in Ravages: the file <em>Kryukov</em> containing MAF simulated under the demographic model of Kryukov and the file <em>GnomADgenes</em> containing MAF from the NFE population in GnomAD. As these files contain MAF for multiple genes, the user needs to specify which gene to choose to simulate the data with the argument <em>select.gene</em>. If this argument is empty, only the first gene will be kept in the simulation procedure.<br/>
Finally, the multiplicative factor of the GRR between each group of cases compared to the first group of cases needs to be specified to the argument <em>GRR.multiplicative.factor</em> (number of values: number of cases groups - 1).<br/>
<strong>GRR.matrix()</strong> will return a GRR matrix in the appropriate format for the function <strong>genotypic.freq()</strong>.<br/>
Examples of these two functions are shown below:</p>

<pre><code class="r">#GRR calculated using the same formula as in SKAT,
#with values in the second group of cases being twice the values from the first one

GRR.del &lt;- GRR.matrix(GRR = &quot;SKAT&quot;, genes.maf = Kryukov, n.case.groups = 2,
                      GRR.multiplicative.factor=2, select.gene = &quot;R1&quot;)
GRR.del[,1:5]
</code></pre>

<pre><code>##           [,1]      [,2]     [,3]      [,4]      [,5]
## [1,]  5.037728  6.222656 12.34472  9.042877  8.133663
## [2,] 10.075455 12.445313 24.68944 18.085755 16.267327
</code></pre>

<pre><code class="r">#Calculation of genotype frequencies in the two groups of cases and the controls group
#The previous GRR matrix is used with a multiplicative model of the disease
#The prevalence in each group of cases is 0.001

geno.freq.groups &lt;- genotypic.freq(genes.maf = Kryukov, select.gene=&quot;R1&quot;, 
                                   GRR.het = GRR.del, prev = c(0.001, 0.001),
                                   genetic.model = &quot;multiplicative&quot;)
str(geno.freq.groups)
</code></pre>

<pre><code>## List of 3
##  $ freq.homo.ref: num [1:3, 1:383] 1 0.999 0.998 1 1 ...
##   ..- attr(*, &quot;dimnames&quot;)=List of 2
##   .. ..$ : chr [1:3] &quot;controls&quot; &quot;cases_1&quot; &quot;cases_2&quot;
##   .. ..$ : NULL
##  $ freq.het     : num [1:3, 1:383] 1.87e-04 9.56e-04 1.91e-03 5.57e-05 3.52e-04 ...
##   ..- attr(*, &quot;dimnames&quot;)=List of 2
##   .. ..$ : chr [1:3] &quot;controls&quot; &quot;cases_1&quot; &quot;cases_2&quot;
##   .. ..$ : NULL
##  $ freq.homo.alt: num [1:3, 1:383] 7.90e-09 2.29e-07 9.15e-07 6.49e-10 3.11e-08 ...
##   ..- attr(*, &quot;dimnames&quot;)=List of 2
##   .. ..$ : chr [1:3] &quot;controls&quot; &quot;cases_1&quot; &quot;cases_2&quot;
##   .. ..$ : NULL
</code></pre>

<pre><code class="r">#frequencies of the homozygous alternative genotype in the different groups
geno.freq.groups$freq.homo.alt[,1:5]
</code></pre>

<pre><code>##                  [,1]         [,2]         [,3]         [,4]         [,5]
## controls 7.897334e-09 6.485888e-10 7.480447e-14 6.568600e-12 2.503543e-11
## cases_1  2.288672e-07 3.106821e-08 4.778956e-11 9.067311e-10 2.469545e-09
## cases_2  9.145933e-07 1.242291e-07 1.911556e-10 3.626706e-09 9.877199e-09
</code></pre>

<p>It is also possible to calculate the MAF in each group of individuals as follow:</p>

<pre><code class="r">#MAF calculation for the first five variants
geno.freq.groups$freq.homo.alt[,1:5] + 0.5*geno.freq.groups$freq.het[,1:5]
</code></pre>

<pre><code>##                  [,1]         [,2]         [,3]         [,4]         [,5]
## controls 9.375276e-05 2.785699e-05 5.403418e-07 3.246158e-06 5.972867e-06
## cases_1  4.784006e-04 1.762618e-04 6.912999e-06 3.011198e-05 4.969452e-05
## cases_2  9.563437e-04 3.524614e-04 1.382590e-05 6.022214e-05 9.938410e-05
</code></pre>

<h3>Simulation of genotypes</h3>

<p>In addition to compute the genotypic frequencies in each group of individuals, it is possible to directly simulate genotypes for a group of controls and more than two groups of cases. This can be done using the function <strong>random.bed.matrix()</strong> which relies on the function <strong>genotypic.freq()</strong> explained previously.
The arguments <em>genes.maf</em>, <em>select.gene</em>, <em>prev</em> and <em>genetic.model</em> are the same as in <strong>genotypic.freq()</strong>.<br/>
In <strong>random.bed.matrix()</strong>, the proportion of causal variants and protective variants (among causal variants) simulated in the genomic region should be specified to <em>p.causal</em> and <em>p.protect</em> respectively. The argument <em>GRR.matrix.del</em> should contain a matrix with GRR values as if all the variants were deleterious. If <em>genetic.model=&ldquo;general&rdquo;</em>, two GRR matrices need to be given as a list to the argument <em>GRR.matrix.del</em> (one for the heterozygous genotype and the other for the homozygous alternative genotype).<br/>
If the user wants to simulate protective variants in addition to deleterious variants, a similar argument to <em>GRR.matrix</em> should be given to <em>GRR.matrix.pro</em> with GRR values as if all variants were protective. If the argument <em>GRR.matrix.pro</em> is empty and <em>p.protect</em>&gt;0, <em>GRR.matrix.pro</em> will correspond to 1/<em>GRR.matrix</em>. 
These deleterious and protective GRR values will then be assigned to the sampled deleterious and protective variants in the simulations, the non-causal variants having GRR values of 1.<br/>
The size of the different groups of individuals should be a vector specified to <em>size</em>, and the user has to choose whether the causal variants will be the same between the different groups of cases with the argument <em>same.variant</em>.<br/>
Finally, the number of simulations need to be specified to <em>replicates</em>.<br/>
<strong>random.bed.matrix()</strong> will return a bed matrix with the group of each individual in the field <em>@ped$pheno</em>, the first one being considered by default as the controls group, and the replicate number corresponding to the genomic region in the field <em>@snps$genomic.region</em>.<br/>
The example below shows how to simulate a group of 1 000 controls and two groups of 500 cases with 50% of deleterious variants having GRR values from the previous example. The deleterious variants are different between the two groups of cases and 5 genomic regions are simulated.</p>

<pre><code class="r">x &lt;- random.bed.matrix(genes.maf = Kryukov, size = c(1000, 500, 500), replicates = 5,
                       prev = c(0.001, 0.001), GRR.matrix.del = GRR.del, 
                       p.causal = 0.5, p.protect = 0, same.variant = FALSE,
                       genetic.model = &quot;multiplicative&quot;, select.gene = &quot;R1&quot;)
x
</code></pre>

<pre><code>## A bed.matrix with 2000 individuals and 1915 markers.
## snps stats are set
##   There are  1634  monomorphic SNPs
## ped stats are set
</code></pre>

<pre><code class="r">table(x@ped$pheno)
</code></pre>

<pre><code>## 
##    0    1    2 
## 1000  500  500
</code></pre>

<pre><code class="r">table(x@snps$genomic.region)
</code></pre>

<pre><code>## 
##  R1  R2  R3  R4  R5 
## 383 383 383 383 383
</code></pre>

<h2>Simulations based on haplotypes</h2>

<p><strong>Ravages</strong> also offers the possibility to perform genetic simulations based on real haplotypes data to mimic linkage desequlibrium pattern observed in these data.<br/>
Two options are available to do so: genetic data are simulated using either haplotypes and their frequencies in different groups of individuals (function <strong>rbm.haplos.freqs()</strong>), or using haplotypes and a liability model on their burdens (function <strong>rbm.haplos.thresholds()</strong>).<br/>
For the first situation, <strong>rbm.haplos.freqs()</strong> requires a matrix of haplotypes (argument <em>haplos</em> with one row per haplotype and one column per variant), and a matrix with each haplotype frequency in each simulated group of individuals (argument <em>freqs</em> with one row per haplotype and one column per group of individuals). Observed haplotypes are then sampled in each group of individuals in respect to their frequencies in each one of them.<br/>
For the second situation, <strong>rbm.haplos.thresholds()</strong> requires a matrix of haplotypes (argument <em>haplos</em> as well), and a number of other arguments to draw the liability model. The idea of these simulations is to draw a liability distribution of the burden of causal variants for each haplotype, to compute the probability of each haplotype in each group of individuals based on this distribution, and finally to sample haplotypes in each group based on those probabilities. <em>weights</em> is a vector whose length corresponds to the number of variants and is used to weight variant into burdens computation (by default, the same formula as in SKAT is used, which gives the higher weights to the rarest variants : \(w=-0.4*|log_{10}(MAF)|\)). <em>maf.threshold</em> corresponds to the maf threshold used to define a rare variants, i.e. all monomorphic variants and variants with a MAF upper this threshold will have a weight of 0. In addition, the number of causal variants needs to be given to <em>nb.causal</em>, and the proportion of protective variants among causal variants needs to be given to <em>p.protect</em>. Causal variants will be sampled among variants with a positive weight, and haplotypes&#39; burden will be computed on these variants.<br/>
<em>h2</em> corresponds to the phenotypic variance explained by the gene: haplotypes burdens will be adjusted on this value; and <em>prev</em> corresponds to the prevalence of each group of individuals and will be used to compute the liability threshold for each group of individuals (individuals in group \(i\) will fall between 1-\(prev_{i}\) and +Inf). This quantile is drawn from a standard normal distribution if <em>normal.approx</em>=TRUE (applicable for low <em>h2</em> value and pretty large <em>prev</em> values), otherwise, quantiles will be drawn from a distribution based on 1e6 observed sampled burdens. If one wants to simulate a group of unselected controls, <em>prev</em> needs to be set at 1, regardless of the other arguments: haplotypes are sampled uniformly in the whole liability distribution. <em>nb.causal</em>, <em>p.protect</em>, <em>h2</em> and <em>prev</em> should be vector of equal size as the number of groups. If they are of size 1, their value will be repeated.<br/>
As for the previous functions, the number of simulations to preform should be given to <em>replicates</em>. As haplotypes&#39; probabilies can take some time to run for a set of causal variants, it is possible to run several multiple replicates on this same set. Indeed, even if the same causal variants are taken, different haplotypes (with same probabilities) will be sampled in each individual for each replicate. To ensure a minimal variability, we yet recommend to resample the set of causal variants several times. For example, for 1000 replicates, we recommend to resample causal variants 20 times, and therefore to perform 50 replicates by set of causal variants. The number of replicates to perform by set of causal variants should be given to <em>rep.by.causal</em>.</p>

<p>As for the genetic simulations based on GRR with <strong>rbm.haplos.freqs()</strong> or <strong>rbm.haplos.thresholds()</strong>, the sizes of each group should be given to <em>size</em>, and the number of simulations to <em>replicates</em>. As before, the replicate number will be in the slot <em>@snps$genomic.region</em> and the group of individuals in <em>@ped$pheno</em>, both being factors. If <strong>rbm.haplos.freqs()</strong> is used, phenotype values will correspond to the colnames of <em>freqs</em>.  </p>

<p>Two examples to illustrate these functions are presented below. The first one enables the simulation of 5 groups of individuals, with haplotype frequencies from the five european populations for the LCT gene. The second one enables the simulation of one group of 100 controls and two groups of 50 cases based on the liability model with different genetic effects. Parameters for these simulations are obtained using <strong>get.rbm.thresholds()</strong></p>

<pre><code class="r">#Load LCT dataset for haplotype matrix
data(LCT.haplotypes)

#Simulation based on haplotypes frequencies
#for the variants in the LCT gene in the EUR population
LCT.gene.hap &lt;- LCT.hap[which(LCT.sample$super.population==&quot;EUR&quot;),
                        which(LCT.snps$pos&gt;=136545410 &amp; LCT.snps$pos&lt;=136594750)]
#Individuals from EUR
LCT.sample.EUR &lt;- subset(LCT.sample, super.population==&quot;EUR&quot;)
#Matrix of haplotype frequencies
LCT.freqs &lt;- sapply(unique(LCT.sample.EUR$population), function(z) 
                    ifelse(LCT.sample.EUR$population==z, 
                           1/table(LCT.sample.EUR$population)[z], 0))
#Simulation of genetic data for five groups of 50 individuals
x &lt;- rbm.haplos.freqs(haplos=LCT.gene.hap, freqs=LCT.freqs, size=rep(50,5), replicates=5)

#Simulation of 100 controls, and two groups of 50 cases with 30 causal variants
#and with the second group having half h2 and twice the prevalence
#compared to the first one
#5 replicates are performed and causal variants are sampled once
x &lt;- rbm.haplos.thresholds(haplos=LCT.gene.hap, nb.causal = 30, h2=c(0.01,0.01,0.02),
                           prev=c(1,0.01,0.005), size=c(100, 50, 50), replicates=5,
                           rep.by.causal = 5)
</code></pre>

<pre><code>## 3  groups of individuals will be simulated
</code></pre>

<h2>Power calculation</h2>

<p>Power calculations are not directly implemented into <strong>Ravages</strong>, but the functions <em>burden()</em> and <em>SKAT()</em> can be used to perform burden tests and SKAT. To have more informations about these two functions, please refer to the main vignette of the package. After the simulation of a bedmatrix using one of the two previously explained methods, statistical power can be computed as follows:</p>

<pre><code class="r">#Simulations using GRR values
#The two groups of cases are genetically heterogeneous
GRR.del &lt;- GRR.matrix(GRR = &quot;SKAT&quot;, genes.maf = Kryukov, n.case.groups = 2,
                      GRR.multiplicative.factor=2, select.gene = &quot;R1&quot;)

x.GRR &lt;- random.bed.matrix(genes.maf = Kryukov, size = c(200, 100, 100), 
                          prev = c(0.001, 0.001), GRR.matrix.del = GRR.del, 
                          p.causal = 0.3, p.protect = 0, same.variant = TRUE,
                          replicates = 100, genetic.model = &quot;multiplicative&quot;, 
                          select.gene = &quot;R1&quot;)
table(x.GRR@ped$pheno)
#Selection of rare variants: MAF&lt;1% in the entire sample
x.GRR.filter &lt;- filter.rare.variants(x.GRR, maf.threshold=0.01, filter=&quot;whole&quot;)

#Comparing each group of cases to the controls with WSS
x.GRR.1 &lt;- select.inds(x.GRR.filter, x.GRR.filter@ped$pheno %in% c(0,1))
x.GRR.2 &lt;- select.inds(x.GRR.filter, x.GRR.filter@ped$pheno %in% c(0,2))
#Construction of null models
H0.cas1 &lt;- NullObject.parameters(x.GRR.1@ped$pheno, ref.level = 0, 
                                 RVAT = &quot;burden&quot;, pheno.type = &quot;categorial&quot;)
H0.cas2 &lt;- NullObject.parameters(x.GRR.2@ped$pheno, ref.level = 0, 
                                 RVAT = &quot;burden&quot;, pheno.type = &quot;categorial&quot;)
#Run WSS
wss.cas1 &lt;- burden(x.GRR.1, H0.cas1, burden=&quot;WSS&quot;, cores = 1)
mean(wss.cas1$p.value&lt;0.05)
wss.cas2 &lt;- burden(x.GRR.2, H0.cas2, burden=&quot;WSS&quot;, cores = 1)
mean(wss.cas2$p.value&lt;0.05)


#Power of SKAT with each group considered separately
#Null model
skat.null.3gps &lt;- NullObject.parameters(x.GRR.filter@ped$pheno, 
                                        RVAT = &quot;SKAT&quot;, pheno.type = &quot;categorial&quot;)
skat.3gps &lt;- SKAT(x.GRR.filter, skat.null.3gps, 
                  params.sampling = list(perm.target = 100, perm.max = 5e4))
mean(skat.3gps$p.value&lt;0.05)
</code></pre>

</body>

</html>
