<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1">



<title>Binary Dosage Formats</title>






<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#header {
text-align: center;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; }  code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Binary Dosage Formats</h1>



<p>There are currently 4 formats for the binary dosage file.</p>
<p>The first three formats consist of three files, binary dosage, family, and map. The family and maps are data frames with information about the subjects and SNPs in the binary dosage file, respectively. These are data frames saved with the <span style="font-family:Courier">saveRDS</span> command.</p>
<div id="format-1" class="section level1">
<h1>Format 1</h1>
<p>Format 1 has a header that begins with a magic word and is followed by a number indicating whether it is in format 1.1 or 1.2. It is the followed by the genotype information. The total header length is 8 bytes.</p>
<div id="format-1.1" class="section level2">
<h2>Format 1.1</h2>
<p>In format 1.1 the only value stored is the dosage. The dosage values stored are multiplied by <span class="math inline">\(2^{15} - 1\)</span>, 0x7ffe, and stored as short integers. If a value is missing it is stored as <span class="math inline">\(2^{16}\)</span>, 0xffff. Each subject needs to 2 bytes per SNP. The total size of the data section is 2 times the number of subjects times the number of SNPs bytes.</p>
</div>
<div id="format-1.2" class="section level2">
<h2>Format 1.2</h2>
<p>In format 1.2 the only values stored are <span class="math inline">\(\Pr(g=1)\)</span> and <span class="math inline">\(\Pr(g=2)\)</span>. These values are multiplied by <span class="math inline">\(2^{16} - 1\)</span>, 0xfffe, and stored as short integers. A value of <span class="math inline">\(2^16\)</span>, 0xffff indicates a value is missing. The total size of the data section is 4 times the number of subjects times the number of SNPs bytes.</p>
</div>
</div>
<div id="format-2" class="section level1">
<h1>Format 2</h1>
<p>Format 2 has the same header as format 1.</p>
<div id="format-2.1" class="section level2">
<h2>Format 2.1</h2>
<p>The format of the data section is same as format 1.1 except the dosage values are multiplied by 20,000, 0x4e20. The missing value is still <span class="math inline">\(2^{16}\)</span>, 0xffff.</p>
</div>
<div id="format-2.2" class="section level2">
<h2>Format 2.2</h2>
<p>The format of the data section is same as format 1.1 except the dosage values are multiplied by 20,000, 0x4e20. The missing value is still <span class="math inline">\(2^{16}\)</span>, 0xffff.</p>
<p><strong>Note:</strong> Format 2 was adopted when it was discovered that the values returned from the imputation programs were limited to 3 or 4 digits passed the decimal point. When results from fitting models were compared between using the binary dosage file and the original VCF or GEN file, there were slight but unimportant differences. It was considered desirable to be able to return the values exactly as they appear in the original imputation file.</p>
</div>
</div>
<div id="format-3" class="section level1">
<h1>Format 3</h1>
<p>Format 3.1 and 3.2 has a header similar to formats 1 and 2 but the number of subjects and SNPs were added to the header to avoid problems associating the wrong family or map file to the binary dosage file.</p>
<p>Format 3.3 and 3.4 has a header similar to formats 1 and 2 but the md5 hash of the family and map data frames were added to the header to avoid problems associating the wrong family or map file to the binary dosage file.</p>
<div id="format-3.1-and-3.3" class="section level2">
<h2>Format 3.1 and 3.3</h2>
<p>The data section of formats 3.1 and 3.3 are the same as format 2.1</p>
</div>
<div id="format-3.2" class="section level2">
<h2>Format 3.2</h2>
<p>Each SNP in the data section begins with a integer value identifying how long the section is for that SNP. The data is then stored as described below in minimizing the data stored.</p>
</div>
<div id="format-3.4" class="section level2">
<h2>Format 3.4</h2>
<p>Format 3.4 stores the data is a similar format as 3.2 but the section begins with the lengths of all the sections for the SNPs and then is followed by the genotype information.</p>
</div>
</div>
<div id="format-4" class="section level1">
<h1>Format 4</h1>
<p>Format 4 takes the data that is in the family and map files and moves it into the header of the binary dosage file. The first section of the header has the magic word and the format. This is followed by information on where the family, map, and genotype data are stored in the file. After the header there is the family data, followed by the map data, and then the imputation data.</p>
<div id="format-4.1-and-4.3" class="section level2">
<h2>Format 4.1 and 4.3</h2>
<p>The data section of formats 4.1 and 4.3 are the same as format 2.1</p>
</div>
<div id="format-4.2-and-4.4" class="section level2">
<h2>Format 4.2 and 4.4</h2>
<p>The data section of formats 4.2 and 4.4 are the same as format 4.2 and 4.3 respectively.</p>
</div>
</div>
<div id="minimizing-the-data-stored" class="section level1">
<h1>Minimizing the data stored</h1>
<p>A lot is known about the imputation data. We know the following <span class="math display">\[\Pr(g=0) + \Pr(g=1) + \Pr(g=2) = 1 \]</span> <span class="math display">\[ d = \Pr(g=1) + 2\Pr(g=2)\]</span> where <span class="math inline">\(d\)</span> is the dosage. This means we only need to know two of the values to calculate the other two. In the <span style="font-family:Courier">BinaryDosage</span> package, the dosage and $(g=1) are used.</p>
<p>It is quite often the case that either <span class="math inline">\(\Pr(g=0)\)</span> or <span class="math inline">\(\Pr(g=2)\)</span> is 0. In this case, knowing the dosage is enough.</p>
<p><span class="math display">\[
\Pr(g = 1) = \left\{\begin{array}{ll}%
d &amp; \; \Pr(g=2)=0, d \leq 1\\%
2 - d &amp; \; \Pr(g=0) = 0, d &gt; 1 %
\end{array}\right.
\]</span> Once the dosage and <span class="math inline">\(\Pr(g=1)\)</span> is know the other values can be quickly calculated.</p>
<p><span class="math display">\[\Pr(g=2) = \frac{d - \Pr(g=1)}{2}\]</span></p>
<p><span class="math display">\[\Pr(g=0) = 1 - \Pr(g=1) - \Pr(g=2)\]</span></p>
<p>These formulae work well but sometimes there is round off error in the imputation values. In these cases the above equations canâ€™t be used to get the exact imputation values. In these situations all four imputations values, <span class="math inline">\(d\)</span>, <span class="math inline">\(\Pr(g=0)\)</span>,<span class="math inline">\(\Pr(g=1)\)</span>, and <span class="math inline">\(\Pr(g=2)\)</span> have to be saved. Fortunately this is not a common occurrence.</p>
<p>Since the values stored are short integers of 2 bytes in length, only the last 15 bits are used. This allows the 16th bit to be used as an indicator. For each SNP for each subject the first value saved is the dosage. If the 16th bit is 0, this indicates that either <span class="math inline">\(\Pr(g=0)\)</span> or <span class="math inline">\(\Pr(g=2)\)</span> is 0 and the other values can be calculated as described above. If the 16th bit is set to 1, this indicates that the value of <span class="math inline">\(\Pr(g=1)\)</span> follows. If the 16th bit is set to 0, this indicates that the above equations can be used to calculate <span class="math inline">\(\Pr(g=0)\)</span> and <span class="math inline">\(\Pr(g=2)\)</span>. If the 16th bit is set to one, this indicates the next two values are <span class="math inline">\(\Pr(g=0)\)</span> and <span class="math inline">\(\Pr(g=2)\)</span> respectively.</p>
<p><strong>Note:</strong> Usage of the this method general results in 2.2 to 2.4 bytes needed to store each SNP for each subject.</p>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
