<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1">



<title>Genetic File Information</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>



<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#header {
text-align: center;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; }  code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Genetic File Information</h1>



<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="kw">library</span>(BinaryDosage)</a></code></pre></div>
<p>The routines <span style="font-family:Courier">getbdinfo</span>, <span style="font-family:Courier">getvcfinfo</span>, and <span style="font-family:Courier">getgeninfo</span> return a list with information about the data in the files. The list returned by each of these routines a section common to them all and a list <span style="font-family:Courier">additionalinfo</span> that is specific to the file type.</p>
<div id="common-section" class="section level2">
<h2>Common section</h2>
<p>The common section has the following elements</p>
<ul>
<li>filename - Character value with the complete path and file name of the file with the genetic data</li>
<li>usesfid - Logical value indicating if the subject data has family IDs.</li>
<li>samples - Data frame containing the following information about the subjects
<ul>
<li>fid - Character value with family IDs</li>
<li>sid - Character value with the individual IDs</li>
</ul></li>
<li>onchr - Logical value indicating if all the SNPs are on the same chromosome</li>
<li>snpidformat - Integer indicating the format of the SNP IDs as follows
<ul>
<li>0 - Unknown for VCF and GEN files or user specified for binary dosage files</li>
<li>1 - chromosome:location</li>
<li>2 - chromosome:location:referenceallele:alternateallele</li>
<li>3 - chromosome:location_referenceallele_alternateallele</li>
</ul></li>
<li>snps - Data frame containing the following values
<ul>
<li>chromosome - Character value indicating what chromosome the SNP is on</li>
<li>location - Integer value with the location of the SNP on the chromosome</li>
<li>snpid - Character value with the ID of the SNP</li>
<li>reference - Character value of the reference allele</li>
<li>alternate - Character value of the alternate allele</li>
</ul></li>
<li>snpinfo - List that contain the following information
<ul>
<li>aaf - numeric vector with the alternate allele frequencies</li>
<li>maf - numeric vector with the minor allele frequencies</li>
<li>avgcall - Numeric vector with the imputation average call</li>
<li>rsq - Numeric vector with the imputation r squared value</li>
</ul></li>
<li>datasize - Numeric vector indicating the size of data in the file for each SNP</li>
<li>indices - Numeric vector indicating the starting location in the file for each SNP</li>
</ul>
<p>The list returned has its class value set to “genetic-info”.</p>
<p>The <span style="font-family:Courier">datasize</span> and <span style="font-family:Courier">indices</span> values are only returned if the parameter <span style="font-family:Courier">index</span> is set equal to <span style="font-family:Courier">TRUE</span></p>
</div>
<div id="binary-dosage-additional-information" class="section level2">
<h2>Binary Dosage Additional Information</h2>
<p>The additional information returned for binary dosage files contains the following information.</p>
<ul>
<li>format - numeric value with the format of the binary dosage file</li>
<li>subformat - numeric value with the subformat of the binary dosage file</li>
<li>headersize - integer value with the size of the header in the binary dosage file</li>
<li>numgroups - integer value of the number of groups of subjects in the binary dosage file. This is usually the number of binary dosage files merged together to form the file</li>
<li>groups - integer vector with size of each of the groups</li>
</ul>
<p>This list has its class value set to “bdose-info”.</p>
</div>
<div id="vcf-file-additional-information" class="section level2">
<h2>VCF File Additional Information</h2>
<p>The additional information returned for VCF files contains the following information.</p>
<ul>
<li>gzipped - Logical value indicating if the file has been compressed using gzip</li>
<li>headerlines - Integer value indicating the number of lines in the header</li>
<li>headersize - Numeric value indicating the size of the header in bytes</li>
<li>quality - Character vector containing the values in QUALITY column</li>
<li>filter - Character vector containing the values in the FILTER column</li>
<li>info - Character vector containing the values in the INFO column</li>
<li>format - Character vector containing the values in the FORMAT column</li>
<li>datacolumns - Data frame summarizing the entries in the FORMAT value containing the following information
<ul>
<li>numcolumns - Integer value indicating the number of values in the FORMAT value</li>
<li>dosage - Integer value indicating the column containing the dosage value</li>
<li>genotypeprob - Integer value indicating the column containing the genotype probabilities</li>
<li>genotype - Integer value indicating the column containing the genotype call</li>
</ul></li>
</ul>
<p>This list has its class value set to “vcf-info”.</p>
<p>The values for quality, filter, info, and format can have a length of 0 if all the values are missing. They will have a length of 1 if all the values are equal. The number of rows in the datacolumns data frame will be equal to the length of the format value.</p>
</div>
<div id="gen-file-additional-information" class="section level2">
<h2>GEN File Additional Information</h2>
<p>The additional information returned for GEN files contains the following information.</p>
<ul>
<li>gzipped - Logical value indicating if the GEN file is compressed using gz</li>
<li>headersize - Integer value indicating the size of the header in bytes</li>
<li>format - Integer value indicating the number of genotype probabilities for each subject with the following meanings
<ul>
<li>1 - Dosage only</li>
<li>2 - <span class="math inline">\(\Pr(g=0)\)</span> and <span class="math inline">\(\Pr(g=1)\)</span></li>
<li>3 - <span class="math inline">\(\Pr(g=0)\)</span>, <span class="math inline">\(\Pr(g=1)\)</span>, and <span class="math inline">\(\Pr(g=2)\)</span></li>
</ul></li>
<li>startcolumn - Integer value indicating in which column the genetic data starts</li>
<li>sep - Character value indicating what value separates the columns</li>
</ul>
<p><span class="math inline">\(g\)</span> indicates the number of alternate alleles the subject has.</p>
<p>This list has its class value set to “gen-info”.</p>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
