<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />



<title>Benchmarks: discretize()</title>

<script>// Hide empty <a> tag within highlighted CodeBlock for screen reader accessibility (see https://github.com/jgm/pandoc/issues/6352#issuecomment-626106786) -->
// v0.0.1
// Written by JooYoung Seo (jooyoung@psu.edu) and Atsushi Yasumoto on June 1st, 2020.

document.addEventListener('DOMContentLoaded', function() {
  const codeList = document.getElementsByClassName("sourceCode");
  for (var i = 0; i < codeList.length; i++) {
    var linkList = codeList[i].getElementsByTagName('a');
    for (var j = 0; j < linkList.length; j++) {
      if (linkList[j].innerHTML === "") {
        linkList[j].setAttribute('aria-hidden', 'true');
      }
    }
  }
});
</script>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
code.sourceCode > span { display: inline-block; line-height: 1.25; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>



<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 800px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Palatino Linotype", "Book Antiqua", Palatino, serif;
font-size: 16px;
line-height: 1.35;
}
#header {
text-align: center;
}
#TOC {
clear: both;
margin: 0 0 10px 0px;
padding: 15px;
width: 770px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 16px;
line-height: 1.5;
}
#TOC .toctitle {
font-weight: bold;
font-size: 16px;
margin-left: 5px;
}
#TOC ul {
padding-left: 50px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 20px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
text-align: justify;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 5px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 5px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: "Lucida Console", Monaco, Consolas, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 25px;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 1px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
margin-left: 8px;
font-size: 105%;
}
em {
font-style: normal;
}
emph {
font-style: oblique;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #123d79;
text-decoration: none;
}
a:hover {
color: #007CC3; }
a:visited {
color: #581858; }
a:visited:hover {
color: #007CC3; }

code > span.kw { color: #4271ae; } 
code > span.dt { color: #c82829; } 
code > span.dv { color: #f5871f; } 
code > span.bn { color: #718c00; } 
code > span.fl { color: #718c00; } 
code > span.ch { color: #718c00; } 
code > span.st { color: #718c00; } 
code > span.co { color: #8e908c; } 
code > span.ot { color: #4d4d4c; } 
code > span.al { color: #ff0000; } 
code > span.fu { color: #4271ae; } 
code > span.er { color: #a61717; } 

img {
display: block;
margin: 0 auto;
}
</style>




</head>

<body>




<h1 class="title toc-ignore">Benchmarks: discretize()</h1>
<h3 class="subtitle"><small><a href="http://r-addict.com/About.html">Marcin Kosinski</a>  <a href="https://stackoverflow.com/users/3857701"><i class="fa fa-stack-overflow"></i></a>  <a href="http://r-addict.com"><i class="fa fa-comment"></i></a>  <a href="https://github.com/MarcinKosinski"><i class="fa fa-github"></i></a>  <a href="mailto:m.p.kosinski@gmail.com"><i class="fa fa-envelope-o"></i></a><br>Krzysztof Slomczynski  <a href="https://github.com/krzyslom"><i class="fa fa-github"></i></a>  <a href="mailto:krzysztofslomczynski@gmail.com"><i class="fa fa-envelope-o"></i></a></small><br></h3>


<div id="TOC">
<ul>
<li><a href="#benchmark-is-not-available-here-we-dont-want-to-run-this-code-on-the-cran-server.-please-visit-the-httpsmi2-warsaw.github.iofselectorrcpparticlesbenchmarks_discretize.html">Benchmark is not available here (we don’t want to run this code on the CRAN server). Please visit the <span>https://mi2-warsaw.github.io/FSelectorRcpp/articles/benchmarks_discretize.html</span></a></li>
<li><a href="#about-discretization">About discretization</a></li>
<li><a href="#comparisons-on-rtcga-data">Comparisons on RTCGA data</a><ul>
<li><a href="#about-the-cancer-genome-atlas">About The Cancer Genome Atlas</a></li>
<li><a href="#dependent-variable">Dependent variable</a></li>
<li><a href="#independent-variables">Independent variables</a></li>
<li><a href="#more">More</a></li>
</ul></li>
<li><a href="#discretization-benchmarks">Discretization Benchmarks</a><ul>
<li><a href="#plots">Plots</a></li>
</ul></li>
<li><a href="#references">References</a></li>
</ul>
</div>

<div id="benchmark-is-not-available-here-we-dont-want-to-run-this-code-on-the-cran-server.-please-visit-the-httpsmi2-warsaw.github.iofselectorrcpparticlesbenchmarks_discretize.html" class="section level3">
<h3>Benchmark is not available here (we don’t want to run this code on the CRAN server). Please visit the <a href="https://mi2-warsaw.github.io/FSelectorRcpp/articles/benchmarks_discretize.html" class="uri">https://mi2-warsaw.github.io/FSelectorRcpp/articles/benchmarks_discretize.html</a></h3>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a><span class="kw">library</span>(microbenchmark)</span>
<span id="cb1-2"><a href="#cb1-2"></a><span class="kw">library</span>(FSelectorRcpp)</span>
<span id="cb1-3"><a href="#cb1-3"></a><span class="kw">library</span>(RWeka)</span></code></pre></div>
</div>
<div id="about-discretization" class="section level1">
<h1>About discretization</h1>
<p>In the statistical modelling, the <strong>discretization</strong> is the process of transferring continuous explanatory variables into discrete counterparts. Problems caused by categorization of continuous variables are known and widely spread <span class="citation">(Harrell 2015)</span>, but in some cases there appear an algorithmic requirement for the discretization. Moreover, there exist few algorithms, like <strong>Decision Trees</strong> <span class="citation">(Salzberg 1994)</span>, where continuous attributes are discretized during the learning process. Other reason for variable discretization include increasing the speed of induction algorithms <span class="citation">(Catlett 1991)</span>.</p>
<p>Even though many categorization algorithms have been developed <span class="citation">(Holte 1993, @Chan169942)</span>, in this article we focus on a recursive entropy minimization heuristic for categorization coupled with a <strong>Minimum Description Length (MDL)</strong> critetion <span class="citation">(Rissanen 1986)</span> that controls the number of intervals produced over the continuous space. This is motivated by the original usage of this algorithm in <strong>FSelector</strong> package, which is our baseline.</p>
<p><span class="citation">(Dougherty, Kohavi, and Sahami 1995)</span> showed better performance of classification for discretized feature set on real-world datasets and states that the described method was found promising by the authors not only for the local discretization but also for global discretization <span class="citation">(Ting 1994)</span>.</p>
</div>
<div id="comparisons-on-rtcga-data" class="section level1">
<h1>Comparisons on RTCGA data</h1>
<div id="about-the-cancer-genome-atlas" class="section level2">
<h2>About The Cancer Genome Atlas</h2>
<blockquote>
<p>The Cancer Genome Atlas (TCGA) is a comprehensive and coordinated effort to accelerate our understanding of the molecular basis of cancer through the application of genome analysis technologies, including large-scale genome sequencing - <a href="https://www.cancer.gov/about-nci/organization/ccg/research/structural-genomics/tcga">https://cancergenome.nih.gov/</a>.</p>
</blockquote>
<p><a href="https://github.com/orgs/RTCGA/people">RTCGA team</a> converted selected datasets from this study into few separate packages that are hosted on Bioconductor. These R packages make selected datasets easier to access and manage. Data sets in RTCGA packages are large and cover complex relations between clinical outcomes and genetic background.</p>
<p>To use RTCGA install package with instructions from it’s <a href="https://bioconductor.org/packages/release/bioc/html/RTCGA.html">Bioconductor home page</a></p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1"></a><span class="co">## try https:// if https:// URLs are not supported</span></span>
<span id="cb2-2"><a href="#cb2-2"></a><span class="kw">source</span>(<span class="st">&quot;https://bioconductor.org/biocLite.R&quot;</span>)</span>
<span id="cb2-3"><a href="#cb2-3"></a><span class="kw">biocLite</span>(<span class="st">&quot;RTCGA&quot;</span>)</span></code></pre></div>
</div>
<div id="dependent-variable" class="section level2">
<h2>Dependent variable</h2>
<p>Specifies whether a sample came from tumour or healthy tissue. It is denoted by 14th sign in <code>bcr_patiend_barcode</code> variable in <code>BRCA.rnaseq</code> dataset that containt RNA-Seq genes’ expressions for patients suffering from Breast Cancer.</p>
</div>
<div id="independent-variables" class="section level2">
<h2>Independent variables</h2>
<p>It is mandatory to get rid of special signs from their names.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1"></a><span class="kw">library</span>(RTCGA.rnaseq)</span>
<span id="cb3-2"><a href="#cb3-2"></a>BRCA.rnaseq &lt;-<span class="st"> </span>RTCGA.rnaseq<span class="op">::</span>BRCA.rnaseq</span>
<span id="cb3-3"><a href="#cb3-3"></a>BRCA.rnaseq<span class="op">$</span>bcr_patient_barcode &lt;-<span class="st"> </span></span>
<span id="cb3-4"><a href="#cb3-4"></a><span class="st">   </span><span class="kw">factor</span>(<span class="kw">substr</span>(BRCA.rnaseq<span class="op">$</span>bcr_patient_barcode, <span class="dv">14</span>, <span class="dv">14</span>))</span>
<span id="cb3-5"><a href="#cb3-5"></a><span class="kw">names</span>(BRCA.rnaseq) &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="dt">pattern =</span> <span class="st">&quot;?&quot;</span>, <span class="dt">replacement =</span> <span class="st">&quot;q&quot;</span>,</span>
<span id="cb3-6"><a href="#cb3-6"></a>                           <span class="dt">x =</span> <span class="kw">names</span>(BRCA.rnaseq), <span class="dt">fixed =</span> <span class="ot">TRUE</span>)</span>
<span id="cb3-7"><a href="#cb3-7"></a><span class="kw">names</span>(BRCA.rnaseq) &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="dt">pattern =</span> <span class="st">&quot;[[:punct:]]&quot;</span>, <span class="dt">replacement =</span> <span class="st">&quot;_&quot;</span>,</span>
<span id="cb3-8"><a href="#cb3-8"></a>                           <span class="dt">x =</span> <span class="kw">names</span>(BRCA.rnaseq))</span></code></pre></div>
</div>
<div id="more" class="section level2">
<h2>More</h2>
<ul>
<li><a href="http://r-addict.com/2016/05/04/RTCGA-Quick-Guide.html">RTCGA Quick Guide</a></li>
<li><a href="https://rtcga.github.io/RTCGA/">RTCGA Website</a></li>
</ul>
</div>
</div>
<div id="discretization-benchmarks" class="section level1">
<h1>Discretization Benchmarks</h1>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1"></a><span class="kw">library</span>(ggplot2)</span>
<span id="cb4-2"><a href="#cb4-2"></a><span class="kw">library</span>(pbapply)</span>
<span id="cb4-3"><a href="#cb4-3"></a><span class="kw">library</span>(tibble)</span>
<span id="cb4-4"><a href="#cb4-4"></a>names_by &lt;-<span class="st"> </span><span class="cf">function</span>(nameBy, vecBy) {</span>
<span id="cb4-5"><a href="#cb4-5"></a>  <span class="kw">c</span>(<span class="kw">paste0</span>(nameBy,</span>
<span id="cb4-6"><a href="#cb4-6"></a>           <span class="kw">sapply</span>(<span class="kw">max</span>(<span class="kw">nchar</span>(vecBy)) <span class="op">-</span><span class="st"> </span><span class="kw">nchar</span>(vecBy),</span>
<span id="cb4-7"><a href="#cb4-7"></a>                  <span class="cf">function</span>(zeros) {</span>
<span id="cb4-8"><a href="#cb4-8"></a>                    <span class="kw">paste0</span>(<span class="kw">rep</span>(<span class="dv">0</span>, zeros), <span class="dt">collapse =</span> <span class="st">&quot;&quot;</span>)</span>
<span id="cb4-9"><a href="#cb4-9"></a>                  }), </span>
<span id="cb4-10"><a href="#cb4-10"></a>                  vecBy))</span>
<span id="cb4-11"><a href="#cb4-11"></a>}</span>
<span id="cb4-12"><a href="#cb4-12"></a></span>
<span id="cb4-13"><a href="#cb4-13"></a>get_times &lt;-<span class="st"> </span><span class="cf">function</span>(fun, vecRows, vecCols, nTimes) {</span>
<span id="cb4-14"><a href="#cb4-14"></a>  forRows &lt;-<span class="st"> </span><span class="kw">pblapply</span>(vecRows, <span class="cf">function</span>(nRows) {</span>
<span id="cb4-15"><a href="#cb4-15"></a>    forCols &lt;-<span class="st"> </span><span class="kw">pblapply</span>(vecCols <span class="op">+</span><span class="st"> </span><span class="dv">1</span>, <span class="cf">function</span>(nCols) {</span>
<span id="cb4-16"><a href="#cb4-16"></a>      <span class="kw">suppressWarnings</span>(<span class="kw">microbenchmark</span>(<span class="kw">fun</span>(bcr_patient_barcode <span class="op">~</span><span class="st"> </span>.,</span>
<span id="cb4-17"><a href="#cb4-17"></a>                         BRCA.rnaseq[<span class="dv">1</span><span class="op">:</span>nRows, <span class="dv">1</span><span class="op">:</span>nCols]),</span>
<span id="cb4-18"><a href="#cb4-18"></a>                     <span class="dt">times =</span> nTimes))</span>
<span id="cb4-19"><a href="#cb4-19"></a>    })</span>
<span id="cb4-20"><a href="#cb4-20"></a>    colsNames &lt;-<span class="st"> </span><span class="kw">names_by</span>(<span class="dt">nameBy =</span> <span class="st">&quot;columns_&quot;</span>, <span class="dt">vecBy =</span> vecCols)</span>
<span id="cb4-21"><a href="#cb4-21"></a>    <span class="kw">setNames</span>(forCols, colsNames)</span>
<span id="cb4-22"><a href="#cb4-22"></a>  })</span>
<span id="cb4-23"><a href="#cb4-23"></a>  rowsNames &lt;-<span class="st"> </span><span class="kw">names_by</span>(<span class="dt">nameBy =</span> <span class="st">&quot;rows_&quot;</span>, <span class="dt">vecBy =</span> vecRows)</span>
<span id="cb4-24"><a href="#cb4-24"></a>  <span class="kw">setNames</span>(forRows, rowsNames)</span>
<span id="cb4-25"><a href="#cb4-25"></a>}</span>
<span id="cb4-26"><a href="#cb4-26"></a></span>
<span id="cb4-27"><a href="#cb4-27"></a>make_df &lt;-<span class="st"> </span><span class="cf">function</span>(bm_data, package) {</span>
<span id="cb4-28"><a href="#cb4-28"></a>  <span class="kw">data_frame</span>(</span>
<span id="cb4-29"><a href="#cb4-29"></a>    <span class="dt">rows =</span> <span class="kw">rep</span>(</span>
<span id="cb4-30"><a href="#cb4-30"></a>      <span class="dt">x =</span> <span class="kw">as.integer</span>(</span>
<span id="cb4-31"><a href="#cb4-31"></a>        <span class="kw">gsub</span>(</span>
<span id="cb4-32"><a href="#cb4-32"></a>          <span class="dt">pattern =</span> <span class="st">&quot;rows_0*&quot;</span>,</span>
<span id="cb4-33"><a href="#cb4-33"></a>             <span class="dt">replacement =</span> <span class="st">&quot;&quot;</span>,</span>
<span id="cb4-34"><a href="#cb4-34"></a>             <span class="dt">x =</span> <span class="kw">names</span>(bm_data)</span>
<span id="cb4-35"><a href="#cb4-35"></a>        )</span>
<span id="cb4-36"><a href="#cb4-36"></a>      ),</span>
<span id="cb4-37"><a href="#cb4-37"></a>      <span class="dt">each =</span>  <span class="kw">unique</span>(<span class="kw">sapply</span>(bm_data, length))</span>
<span id="cb4-38"><a href="#cb4-38"></a>    ),</span>
<span id="cb4-39"><a href="#cb4-39"></a>    <span class="dt">columns =</span> <span class="kw">as.integer</span>(</span>
<span id="cb4-40"><a href="#cb4-40"></a>      <span class="kw">gsub</span>(</span>
<span id="cb4-41"><a href="#cb4-41"></a>        <span class="dt">pattern =</span> <span class="st">&quot;columns_0*&quot;</span>,</span>
<span id="cb4-42"><a href="#cb4-42"></a>        <span class="dt">replacement =</span> <span class="st">&quot;&quot;</span>,</span>
<span id="cb4-43"><a href="#cb4-43"></a>        <span class="dt">x =</span> <span class="kw">as.vector</span>(<span class="kw">sapply</span>(bm_data, names))</span>
<span id="cb4-44"><a href="#cb4-44"></a>      )</span>
<span id="cb4-45"><a href="#cb4-45"></a>    ),</span>
<span id="cb4-46"><a href="#cb4-46"></a>    <span class="dt">time =</span> <span class="kw">as.vector</span>(</span>
<span id="cb4-47"><a href="#cb4-47"></a>      <span class="kw">sapply</span>(</span>
<span id="cb4-48"><a href="#cb4-48"></a>        bm_data,</span>
<span id="cb4-49"><a href="#cb4-49"></a>        <span class="cf">function</span>(setOfCols) {</span>
<span id="cb4-50"><a href="#cb4-50"></a>          <span class="kw">sapply</span>(</span>
<span id="cb4-51"><a href="#cb4-51"></a>            setOfCols,</span>
<span id="cb4-52"><a href="#cb4-52"></a>            <span class="cf">function</span>(set) {</span>
<span id="cb4-53"><a href="#cb4-53"></a>              <span class="kw">median</span>(set<span class="op">$</span>time) <span class="op">/</span><span class="st"> </span><span class="dv">10</span> <span class="op">^</span><span class="st"> </span><span class="dv">9</span></span>
<span id="cb4-54"><a href="#cb4-54"></a>            }</span>
<span id="cb4-55"><a href="#cb4-55"></a>          )</span>
<span id="cb4-56"><a href="#cb4-56"></a>        }</span>
<span id="cb4-57"><a href="#cb4-57"></a>      )</span>
<span id="cb4-58"><a href="#cb4-58"></a>    ),</span>
<span id="cb4-59"><a href="#cb4-59"></a>    <span class="dt">package =</span> package</span>
<span id="cb4-60"><a href="#cb4-60"></a>  )</span>
<span id="cb4-61"><a href="#cb4-61"></a>}</span>
<span id="cb4-62"><a href="#cb4-62"></a></span>
<span id="cb4-63"><a href="#cb4-63"></a>bm_fsr_discretize &lt;-<span class="st"> </span><span class="kw">get_times</span>(<span class="dt">fun =</span> discretize,</span>
<span id="cb4-64"><a href="#cb4-64"></a>                               <span class="dt">vecRows =</span> <span class="kw">seq</span>(<span class="dt">from =</span> <span class="dv">200</span>, <span class="dt">to =</span> <span class="dv">1000</span>, <span class="dt">by =</span> <span class="dv">200</span>),</span>
<span id="cb4-65"><a href="#cb4-65"></a>                               <span class="dt">vecCols =</span> <span class="kw">c</span>(<span class="dv">10</span>, <span class="dv">100</span>, <span class="dv">500</span>, <span class="dv">1000</span>),</span>
<span id="cb4-66"><a href="#cb4-66"></a>                               <span class="dt">nTimes =</span> 1L)</span>
<span id="cb4-67"><a href="#cb4-67"></a>bm_rwk_discretize &lt;-<span class="st"> </span><span class="kw">get_times</span>(<span class="dt">fun =</span> Discretize,</span>
<span id="cb4-68"><a href="#cb4-68"></a>                               <span class="dt">vecRows =</span> <span class="kw">seq</span>(<span class="dt">from =</span> <span class="dv">200</span>, <span class="dt">to =</span> <span class="dv">1000</span>, <span class="dt">by =</span> <span class="dv">200</span>),</span>
<span id="cb4-69"><a href="#cb4-69"></a>                               <span class="dt">vecCols =</span> <span class="kw">c</span>(<span class="dv">10</span>, <span class="dv">100</span>, <span class="dv">500</span>, <span class="dv">1000</span>),</span>
<span id="cb4-70"><a href="#cb4-70"></a>                               <span class="dt">nTimes =</span> 1L)</span>
<span id="cb4-71"><a href="#cb4-71"></a></span>
<span id="cb4-72"><a href="#cb4-72"></a>df_fsr &lt;-<span class="st"> </span><span class="kw">make_df</span>(<span class="dt">bm_data =</span> bm_fsr_discretize, <span class="dt">package =</span> <span class="st">&quot;FSelectorRcpp&quot;</span>)</span>
<span id="cb4-73"><a href="#cb4-73"></a>df_rwk &lt;-<span class="st"> </span><span class="kw">make_df</span>(<span class="dt">bm_data =</span> bm_rwk_discretize, <span class="dt">package =</span> <span class="st">&quot;RWeka&quot;</span>)</span>
<span id="cb4-74"><a href="#cb4-74"></a>bm_df &lt;-<span class="st"> </span><span class="kw">rbind</span>(df_fsr, df_rwk)</span>
<span id="cb4-75"><a href="#cb4-75"></a></span>
<span id="cb4-76"><a href="#cb4-76"></a>make_plot &lt;-<span class="st"> </span><span class="cf">function</span>(y) {</span>
<span id="cb4-77"><a href="#cb4-77"></a>  yUnit &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="dt">pattern =</span> <span class="st">&quot;time&quot;</span>,</span>
<span id="cb4-78"><a href="#cb4-78"></a>                <span class="dt">replacement =</span> <span class="st">&quot;s&quot;</span>,</span>
<span id="cb4-79"><a href="#cb4-79"></a>                <span class="dt">x =</span> y)</span>
<span id="cb4-80"><a href="#cb4-80"></a>  </span>
<span id="cb4-81"><a href="#cb4-81"></a>  <span class="kw">ggplot</span>(<span class="dt">data =</span> bm_df, <span class="kw">aes_string</span>(<span class="dt">x =</span> <span class="st">&quot;rows&quot;</span>, <span class="dt">y =</span> y, <span class="dt">color =</span> <span class="st">&quot;package&quot;</span>)) <span class="op">+</span></span>
<span id="cb4-82"><a href="#cb4-82"></a><span class="st">  </span><span class="kw">facet_grid</span>(columns <span class="op">~</span><span class="st"> </span>., <span class="dt">labeller =</span> label_both) <span class="op">+</span></span>
<span id="cb4-83"><a href="#cb4-83"></a><span class="st">  </span><span class="kw">geom_point</span>() <span class="op">+</span></span>
<span id="cb4-84"><a href="#cb4-84"></a><span class="st">  </span><span class="kw">xlab</span>(<span class="st">&quot;Number of rows [-]&quot;</span>) <span class="op">+</span></span>
<span id="cb4-85"><a href="#cb4-85"></a><span class="st">  </span><span class="kw">ylab</span>(<span class="kw">paste0</span>(<span class="st">&quot;Elapsed time [&quot;</span>, yUnit, <span class="st">&quot;]&quot;</span>)) <span class="op">+</span></span>
<span id="cb4-86"><a href="#cb4-86"></a><span class="st">  </span><span class="kw">ggtitle</span>(<span class="dt">label =</span> <span class="st">&quot;Microbenchmark: discretization function comparison&quot;</span>,</span>
<span id="cb4-87"><a href="#cb4-87"></a>          <span class="dt">subtitle =</span> <span class="st">&quot;for different sizes of data sets&quot;</span>) <span class="op">+</span></span>
<span id="cb4-88"><a href="#cb4-88"></a><span class="st">  </span><span class="kw">theme</span>(<span class="dt">plot.title =</span> <span class="kw">element_text</span>(<span class="dt">hjust =</span> <span class="fl">0.5</span>),</span>
<span id="cb4-89"><a href="#cb4-89"></a>        <span class="dt">plot.subtitle =</span> <span class="kw">element_text</span>(<span class="dt">hjust =</span> <span class="fl">0.5</span>),</span>
<span id="cb4-90"><a href="#cb4-90"></a>        <span class="dt">legend.position =</span> <span class="st">&quot;top&quot;</span>)</span>
<span id="cb4-91"><a href="#cb4-91"></a>}</span>
<span id="cb4-92"><a href="#cb4-92"></a></span>
<span id="cb4-93"><a href="#cb4-93"></a>bm_plot &lt;-<span class="st"> </span><span class="kw">make_plot</span>(<span class="st">&quot;time&quot;</span>)</span>
<span id="cb4-94"><a href="#cb4-94"></a>bm_plot_log &lt;-<span class="st"> </span><span class="kw">make_plot</span>(<span class="st">&quot;log(time)&quot;</span>)</span></code></pre></div>
<div id="plots" class="section level2">
<h2>Plots</h2>
</div>
</div>
<div id="references" class="section level1 unnumbered">
<h1>References</h1>
<div id="refs" class="references">
<div id="ref-Catlett1991">
<p>Catlett, J. 1991. “On Changing Continuous Attributes into Ordered Discrete Attributes.” In <em>Machine Learning — Ewsl-91: European Working Session on Learning Porto, Portugal, March 6–8, 1991 Proceedings</em>, edited by Yves Kodratoff, 164–78. Berlin, Heidelberg: Springer Berlin Heidelberg. <a href="https://doi.org/10.1007/BFb0017012">https://doi.org/10.1007/BFb0017012</a>.</p>
</div>
<div id="ref-Chan169942">
<p>Chan, C. C., C. Batur, and A. Srinivasan. 1991. “Determination of Quantization Intervals in Rule Based Model for Dynamic Systems.” In <em>Systems, Man, and Cybernetics, 1991. ’Decision Aiding for Complex Systems, Conference Proceedings., 1991 Ieee International Conference on</em>, 1719–23 vol.3. <a href="https://doi.org/10.1109/ICSMC.1991.169942">https://doi.org/10.1109/ICSMC.1991.169942</a>.</p>
</div>
<div id="ref-Dougherty95supervisedand">
<p>Dougherty, James, Ron Kohavi, and Mehran Sahami. 1995. “Supervised and Unsupervised Discretization of Continuous Features.” In <em>MACHINE Learning: PROCEEDINGS of the Twelfth International Conference</em>, 194–202. Morgan Kaufmann.</p>
</div>
<div id="ref-HarrellDISC">
<p>Harrell, Frank. 2015. “General Handouts and Reference Material on Biostatistical Modeling - Problems Caused by Categorizing Continuous Variables.” Vanderbilt Biostatistics Wiki. 2015. <a href="http://biostat.mc.vanderbilt.edu/wiki/Main/CatContinuous">http://biostat.mc.vanderbilt.edu/wiki/Main/CatContinuous</a>.</p>
</div>
<div id="ref-Holte1993">
<p>Holte, Robert C. 1993. “Very Simple Classification Rules Perform Well on Most Commonly Used Datasets.” <em>Machine Learning</em> 11 (1): 63–90. <a href="https://doi.org/10.1023/A:1022631118932">https://doi.org/10.1023/A:1022631118932</a>.</p>
</div>
<div id="ref-rissanen1986">
<p>Rissanen, Jorma. 1986. “Stochastic Complexity and Modeling.” <em>Ann. Statist.</em> 14 (3): 1080–1100. <a href="https://doi.org/10.1214/aos/1176350051">https://doi.org/10.1214/aos/1176350051</a>.</p>
</div>
<div id="ref-Salzberg1994">
<p>Salzberg, Steven L. 1994. “C4.5: Programs for Machine Learning by J. Ross Quinlan. Morgan Kaufmann Publishers, Inc., 1993.” <em>Machine Learning</em> 16 (3): 235–40. <a href="https://doi.org/10.1007/BF00993309">https://doi.org/10.1007/BF00993309</a>.</p>
</div>
<div id="ref-Ting94discretizationof">
<p>Ting, Kai Ming. 1994. “Discretization of Continuous-Valued Attributes and Instance-Based Learning.”</p>
</div>
</div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
