<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Legana Fingerhut" />

<meta name="date" content="2020-05-11" />

<title>How to train your model</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>



<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">How to train your model</h1>
<h4 class="author">Legana Fingerhut</h4>
<h4 class="date">2020-05-11</h4>



<p>By default <code>ampir</code> predicts the probability of a protein to be an antimicrobial peptide (AMP) or not based on a <em>trained SVM model</em> with as input known AMP sequences corresponding to a wide diversity of organisms. However, within the <code>predict_amps</code> function there is a <code>model</code> argument that allows users to pass their own trained model object. Using a different trained model might be useful when users wish to e.g. use a taxonomic specific model to predict AMPs in a restricted group of taxa.</p>
<p>This vignette will go through a mock example of how you can train your own model using the <code>caret</code> package. For more information on how to use <code>caret</code> and the functions used within this example, please see the <a href="http://topepo.github.io/caret/index.html">extensive documentation</a> made by the author, Max Kuhn.</p>
<div id="step-1-obtain-input-data" class="section level3">
<h3>Step 1: Obtain input data</h3>
<p>First, a positive and negative dataset have to be obtained. In this example, we want to predict AMPs in bats and decide to train a model using protein sequences found in bats. The positive dataset are AMPs and the negative dataset are random sequences. Both datasets were obtained from <a href="https://www.uniprot.org/">UniProt</a>:</p>
<ul>
<li>positive by using the search term keyword:antimicrobial taxonomy:“Chiroptera [9397]”</li>
<li>negative by using the search term taxonomy:“Chiroptera [9397]”</li>
</ul>
<p>For the positive dataset:</p>
<ul>
<li>read data</li>
<li>add “positive” label</li>
<li>remove non standard amino acids</li>
</ul>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="kw">library</span>(ampir)</a></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1">bat_pos &lt;-<span class="st"> </span><span class="kw">read_faa</span>(<span class="kw">system.file</span>(<span class="st">&quot;extdata/bat_positive.fasta.gz&quot;</span>, <span class="dt">package =</span> <span class="st">&quot;ampir&quot;</span>))</a>
<a class="sourceLine" id="cb2-2" data-line-number="2">bat_pos<span class="op">$</span>Label &lt;-<span class="st"> &quot;Positive&quot;</span></a>
<a class="sourceLine" id="cb2-3" data-line-number="3">bat_pos &lt;-<span class="st"> </span><span class="kw">remove_nonstandard_aa</span>(bat_pos)</a></code></pre></div>
<p>For the negative dataset:</p>
<ul>
<li>read data</li>
<li>add “negative” lavel</li>
<li>remove non standard amino acids</li>
<li>remove sequences (if any) that are also present in the positive dataset</li>
<li>randomly select the same number of sequences that are in the positive dataset</li>
</ul>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1">bat_neg &lt;-<span class="st"> </span><span class="kw">read_faa</span>(<span class="kw">system.file</span>(<span class="st">&quot;extdata/bat_negative.fasta.gz&quot;</span>, <span class="dt">package =</span> <span class="st">&quot;ampir&quot;</span>))</a>
<a class="sourceLine" id="cb3-2" data-line-number="2">bat_neg<span class="op">$</span>Label &lt;-<span class="st"> &quot;Negative&quot;</span></a>
<a class="sourceLine" id="cb3-3" data-line-number="3">bat_neg &lt;-<span class="st"> </span><span class="kw">remove_nonstandard_aa</span>(bat_neg)</a>
<a class="sourceLine" id="cb3-4" data-line-number="4">bat_neg &lt;-<span class="st"> </span>bat_neg[<span class="op">!</span>bat_neg<span class="op">$</span>seq_aa <span class="op">%in%</span><span class="st"> </span>bat_pos<span class="op">$</span>seq_aa,]</a>
<a class="sourceLine" id="cb3-5" data-line-number="5">bat_neg &lt;-<span class="st"> </span>bat_neg[<span class="kw">sample</span>(<span class="kw">nrow</span>(bat_neg),<span class="dv">78</span>),]</a></code></pre></div>
<p>Combine the positive and negative dataset</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1">bats &lt;-<span class="st"> </span><span class="kw">rbind</span>(bat_pos, bat_neg)</a></code></pre></div>
<p>Calculate features on the combined positive and negative dataset and add the label column</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1">bats_features &lt;-<span class="st"> </span><span class="kw">calculate_features</span>(bats)</a>
<a class="sourceLine" id="cb5-2" data-line-number="2">bats_features<span class="op">$</span>Label &lt;-<span class="st"> </span><span class="kw">as.factor</span>(bats<span class="op">$</span>Label)</a>
<a class="sourceLine" id="cb5-3" data-line-number="3"><span class="kw">rownames</span>(bats_features) &lt;-<span class="st"> </span><span class="ot">NULL</span></a></code></pre></div>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1"><span class="kw">library</span>(caret)</a></code></pre></div>
<p>Split feature set data and create train and test set with <code>caret</code></p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" data-line-number="1">trainIndex &lt;-<span class="kw">createDataPartition</span>(<span class="dt">y=</span>bats_features<span class="op">$</span>Label, <span class="dt">p=</span>.<span class="dv">7</span>, <span class="dt">list =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb7-2" data-line-number="2">bats_featuresTrain &lt;-bats_features[trainIndex,]</a>
<a class="sourceLine" id="cb7-3" data-line-number="3">bats_featuresTest &lt;-bats_features[<span class="op">-</span>trainIndex,]</a></code></pre></div>
<p>Resample method using repeated cross validation and adding in a probability calculation with <code>caret</code></p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" data-line-number="1">trctrl_prob &lt;-<span class="st"> </span><span class="kw">trainControl</span>(<span class="dt">method =</span> <span class="st">&quot;repeatedcv&quot;</span>, <span class="dt">number =</span> <span class="dv">10</span>, <span class="dt">repeats =</span> <span class="dv">3</span>,</a>
<a class="sourceLine" id="cb8-2" data-line-number="2">                            <span class="dt">classProbs =</span> <span class="ot">TRUE</span>)</a></code></pre></div>
<p>Train model using a support vector machine with radial kernel with <code>caret</code>. <em>Note: Other classification models are supported too. For example, to use a <a href="https://topepo.github.io/caret/train-models-by-tag.html#random-forest">random forest model in <code>caret</code></a>, <code>method</code> could be changed from “svmRadial” to “ranger”.</em></p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" data-line-number="1">my_bat_svm_model &lt;-<span class="st"> </span><span class="kw">train</span>(Label<span class="op">~</span>.,</a>
<a class="sourceLine" id="cb9-2" data-line-number="2">                       <span class="dt">data =</span> bats_featuresTrain[,<span class="op">-</span><span class="dv">1</span>], <span class="co"># excluding seq_name column</span></a>
<a class="sourceLine" id="cb9-3" data-line-number="3">                       <span class="dt">method=</span><span class="st">&quot;svmRadial&quot;</span>,</a>
<a class="sourceLine" id="cb9-4" data-line-number="4">                       <span class="dt">trControl =</span> trctrl_prob,</a>
<a class="sourceLine" id="cb9-5" data-line-number="5">                       <span class="dt">preProcess =</span> <span class="kw">c</span>(<span class="st">&quot;center&quot;</span>, <span class="st">&quot;scale&quot;</span>))</a></code></pre></div>
<p>Test model to get an indication of how well the model performs on test data with <code>caret</code></p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" data-line-number="1">my_bat_pred &lt;-<span class="st"> </span><span class="kw">predict</span>(my_bat_svm_model, bats_featuresTest)</a>
<a class="sourceLine" id="cb10-2" data-line-number="2">cm &lt;-<span class="st"> </span><span class="kw">confusionMatrix</span>(my_bat_pred, bats_featuresTest<span class="op">$</span>Label, <span class="dt">positive =</span> <span class="st">&quot;Positive&quot;</span>)</a></code></pre></div>
<p><em>Subset from <code>cm$byClass</code></em></p>
<table>
<thead>
<tr class="header">
<th align="center">Balanced Accuracy</th>
<th align="center">Precision</th>
<th align="center">Recall</th>
<th align="center">F1</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center">0.98</td>
<td align="center">1</td>
<td align="center">0.96</td>
<td align="center">0.98</td>
</tr>
</tbody>
</table>
<p>Convert the bat feature test data to the original FASTA type format containing just the sequence name and sequence as this is the required input data for <code>ampir</code></p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" data-line-number="1">bat_test_set &lt;-<span class="st"> </span>bats[bats<span class="op">$</span>seq_name <span class="op">%in%</span><span class="st"> </span>bats_featuresTest<span class="op">$</span>seq_name,][,<span class="op">-</span><span class="dv">3</span>]</a>
<a class="sourceLine" id="cb11-2" data-line-number="2"><span class="kw">rownames</span>(bat_test_set) &lt;-<span class="st"> </span><span class="ot">NULL</span></a></code></pre></div>
<p>Use the trained bat model in <code>ampir</code>’s <code>predict_amps</code> function on the bat test set</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" data-line-number="1">my_bat_AMPs &lt;-<span class="st"> </span><span class="kw">predict_amps</span>(bat_test_set, <span class="dt">min_len =</span> <span class="dv">5</span>, <span class="dt">model =</span> my_bat_svm_model)</a></code></pre></div>
<p><em><code>my_bat_AMPs</code> sample</em></p>
<table>
<thead>
<tr class="header">
<th></th>
<th align="left">seq_name</th>
<th align="left">seq_aa</th>
<th align="right">prob_AMP</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>1</td>
<td align="left">G1P1T7…</td>
<td align="left">MKALLTLGLL…</td>
<td align="right">0.999</td>
</tr>
<tr class="even">
<td>2</td>
<td align="left">G1NZQ6…</td>
<td align="left">MKLFLILIIL…</td>
<td align="right">0.986</td>
</tr>
<tr class="odd">
<td>3</td>
<td align="left">U3KZW3…</td>
<td align="left">MKSLLILGLL…</td>
<td align="right">0.998</td>
</tr>
<tr class="even">
<td>44</td>
<td align="left">P14392…</td>
<td align="left">VHLSGEEKAA…</td>
<td align="right">0.010</td>
</tr>
<tr class="odd">
<td>45</td>
<td align="left">Q7I5M1…</td>
<td align="left">MALIYTNTLL…</td>
<td align="right">0.014</td>
</tr>
<tr class="even">
<td>46</td>
<td align="left">B2KI57…</td>
<td align="left">MEPFSSKSLA…</td>
<td align="right">0.050</td>
</tr>
</tbody>
</table>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
