<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1">

<meta name="author" content="Paul Staab" />


<title>Using External Simulators</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>



<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#header {
text-align: center;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; }  code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Using External Simulators</h1>
<h4 class="author">Paul Staab</h4>
<h4 class="date">coala 0.6.0</h4>



<p>Coala can call the coalescent simulators <em>ms</em>[1], <em>msms</em>[2] and <em>scrm</em>[3] and can use <em>seq-gen</em>[4] for finite sites simulations. The R version of <em>scrm</em> should get installed automatically as a dependeny of coala. For the other programs, you need to have an executable binary available on your system.</p>
<div id="installation" class="section level1">
<h1>Installation</h1>
<p>Short instructions on obtaining and compiling the programs are given in the help pages of <code>activate_ms</code>, <code>activate_msms</code> and <code>activate_seqgen</code>. More detailed instructions are provided in <a href="https://github.com/statgenlmu/coala/wiki/Installation">the wiki</a>.</p>
</div>
<div id="activation" class="section level1">
<h1>Activation</h1>
<p>In addition to providing the binary for a simulator, you need to inform <code>coala</code> where the binary is. We refer to this process as <em>activation</em> of a binary. Afterwards, <code>coala</code> will use the simulator automatically where-ever appropriate.</p>
<p>There are three different ways to activate a binary:</p>
<ol style="list-style-type: decimal">
<li>Use the <code>activate_msms</code> and <code>activate_seqgen</code> functions to activate the simulators from within R. You should use the functions before creating a model.</li>
<li>Alternatively, you can place the binaries in your working directory or in a folder listed in your <em>PATH</em> environment variable using one of the names listed under “Expected Binary Names” below. If there is a match file, coala will automatically activate the simulator.</li>
<li>You can start the R session with an environment variable that hold the path to the binaries. In this case, the simulators should also be automatically be activated when the coala package is loaded.</li>
<li>Coala uses the R versions of <code>scrm</code> and <code>ms</code>. <code>scrm</code> should alawys be available. Install the CRAN package <code>phyclust</code> to use <code>ms</code>.</li>
</ol>
<table>
<thead>
<tr class="header">
<th>Simulator</th>
<th>Priority</th>
<th>Expected Binary Names</th>
<th>Environment Var</th>
<th>Function</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>seq-gen</td>
<td>100</td>
<td>seqgen, seq-gen, seqgen.exe, seq-gen.exe</td>
<td>SEQGEN</td>
<td>activate_seqgen</td>
</tr>
<tr class="even">
<td>msms</td>
<td>200</td>
<td>msms.jar / java, java.exe</td>
<td>MSMS / JAVA</td>
<td>activate_msms</td>
</tr>
<tr class="odd">
<td>ms</td>
<td>300</td>
<td></td>
<td></td>
<td>activate_ms</td>
</tr>
<tr class="even">
<td>scrm</td>
<td>400</td>
<td></td>
<td></td>
<td></td>
</tr>
</tbody>
</table>
<p>You can use the <code>list_simulators()</code> command to view which simulators are currently available:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="kw">library</span>(coala)</a>
<a class="sourceLine" id="cb1-2" data-line-number="2"><span class="kw">list_simulators</span>()</a></code></pre></div>
<pre><code>##   name priority                      info
## 2 scrm      400         version : 1.7.3-1
## 1   ms      300 version : phyclust_0.1.28</code></pre>
</div>
<div id="priority" class="section level1">
<h1>Priority</h1>
<p>The <code>check_model</code> function checks which simulators support a specific model, and states the problems which coala has detected with the simulators that do not support it. For example, a simple model with infinite-sites mutations (IFS) can be simulated with <code>scrm</code> or – if installed – with <code>ms</code> and <code>msms</code>, but not with <code>seq-gen</code> because the latter generates finite-sites mutations:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1">model &lt;-<span class="st"> </span><span class="kw">coal_model</span>(<span class="dv">10</span>, <span class="dv">1</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb3-2" data-line-number="2"><span class="st">  </span><span class="kw">feat_mutation</span>(<span class="dv">5</span>, <span class="dt">model =</span> <span class="st">&quot;IFS&quot;</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb3-3" data-line-number="3"><span class="st">  </span><span class="kw">sumstat_nucleotide_div</span>()</a>
<a class="sourceLine" id="cb3-4" data-line-number="4"><span class="kw">check_model</span>(model)</a></code></pre></div>
<pre><code>## ms : OK
## 
## scrm : OK</code></pre>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1">model</a></code></pre></div>
<pre><code>## Features: 
## * Sampling of 10 (pop 1) individuals with ploidy 1 at time `0` 
## * Mutations with rate `5` following a IFS mutation model
## * Generating Seg. Sites
## 
## Parameter: None
## 
## Loci: 1 locus of length 1000 
## 
## Summary Statistics: stat_pi 
## 
## Simulator: scrm 
## Command: scrm 10 1 -t 5</code></pre>
<p>If multiple simulators can simulate a model, the one with the highest priority is used. In our example, that is <code>scrm</code>. If we would like to use <code>ms</code> instead, we need to raise its priority:</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" data-line-number="1"><span class="kw">activate_ms</span>(<span class="dt">priority =</span> <span class="dv">500</span>)</a></code></pre></div>
</div>
<div id="references" class="section level1">
<h1>References</h1>
<ul>
<li><strong>[1]</strong>: Richard R. Hudson. <em>Generating samples under a Wright-Fisher neutral model of genetic variation.</em> Bioinformatics (2002) 18 (2): 337-338 <a href="http://doi.org/10.1093/bioinformatics/18.2.337">10.1093/bioinformatics/18.2.337</a></li>
<li><strong>[2]</strong>: Gregory Ewing and Joachim Hermisson. <em>MSMS: a coalescent simulation program including recombination, demographic structure and selection at a single locus.</em> Bioinformatics (2010) 26 (16): 2064-2065 <a href="http://doi.org/10.1093/bioinformatics/btq322">10.1093/bioinformatics/btq322</a></li>
<li><strong>[3]</strong>: Paul R. Staab, Sha Zhu, Dirk Metzler and Gerton Lunter. <em>scrm: efficiently simulating long sequences using the approximated coalescent with recombination.</em> Bioinformatics (2015) 31 (10): 1680-1682. <a href="http://doi.org/doi:10.1093/bioinformatics/btu861">10.1093/bioinformatics/btu861</a></li>
<li><strong>[4]</strong>: Andrew Rambaut and Nicholas C. Grassly. <em>Seq-Gen: an application for the Monte Carlo simulation of DNA sequence evolution along phylogenetic trees.</em> Comput Appl Biosci (1997) 13 (3): 235-238 <a href="http://doi.org/10.1093/bioinformatics/13.3.235">10.1093/bioinformatics/13.3.235</a></li>
</ul>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
