<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>5 All menu options | Genepop version 4.7.5</title>
  <meta name="description" content="5 All menu options | Genepop version 4.7.5" />
  <meta name="generator" content="bookdown 0.10 and GitBook 2.6.7" />

  <meta property="og:title" content="5 All menu options | Genepop version 4.7.5" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="5 All menu options | Genepop version 4.7.5" />
  
  
  

<meta name="author" content="F. Rousset" />



  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="sec-settings.html">
<link rel="next" href="evaluating-the-performance-of-inferences-for-isolation-by-distance.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />










</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="1" data-path="introduction.html"><a href="introduction.html"><i class="fa fa-check"></i><b>1</b> Introduction</a><ul>
<li class="chapter" data-level="1.1" data-path="introduction.html"><a href="introduction.html#purpose"><i class="fa fa-check"></i><b>1.1</b> Purpose</a></li>
<li class="chapter" data-level="1.2" data-path="introduction.html"><a href="introduction.html#the-two-genepop-distributions"><i class="fa fa-check"></i><b>1.2</b> The two Genepop distributions</a></li>
<li class="chapter" data-level="1.3" data-path="introduction.html"><a href="introduction.html#changes-since-version-4.0"><i class="fa fa-check"></i><b>1.3</b> Changes since version 4.0</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="installing-genepop-and-session-examples.html"><a href="installing-genepop-and-session-examples.html"><i class="fa fa-check"></i><b>2</b> Installing Genepop and session examples</a><ul>
<li class="chapter" data-level="2.1" data-path="installing-genepop-and-session-examples.html"><a href="installing-genepop-and-session-examples.html#installation"><i class="fa fa-check"></i><b>2.1</b> Installation</a><ul>
<li class="chapter" data-level="2.1.1" data-path="installing-genepop-and-session-examples.html"><a href="installing-genepop-and-session-examples.html#r-package"><i class="fa fa-check"></i><b>2.1.1</b> R package</a></li>
<li class="chapter" data-level="2.1.2" data-path="installing-genepop-and-session-examples.html"><a href="installing-genepop-and-session-examples.html#stand-alone-executable"><i class="fa fa-check"></i><b>2.1.2</b> Stand-alone executable</a></li>
</ul></li>
<li class="chapter" data-level="2.2" data-path="installing-genepop-and-session-examples.html"><a href="installing-genepop-and-session-examples.html#example-sessions"><i class="fa fa-check"></i><b>2.2</b> Example sessions</a><ul>
<li class="chapter" data-level="2.2.1" data-path="installing-genepop-and-session-examples.html"><a href="installing-genepop-and-session-examples.html#example-1-basic-session"><i class="fa fa-check"></i><b>2.2.1</b> Example 1: basic session</a></li>
<li class="chapter" data-level="2.2.2" data-path="installing-genepop-and-session-examples.html"><a href="installing-genepop-and-session-examples.html#example-2-using-the-settings-file"><i class="fa fa-check"></i><b>2.2.2</b> Example 2: using the settings file</a></li>
<li class="chapter" data-level="2.2.3" data-path="installing-genepop-and-session-examples.html"><a href="installing-genepop-and-session-examples.html#example-3-batch-processing"><i class="fa fa-check"></i><b>2.2.3</b> Example 3: Batch processing</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="the-input-file.html"><a href="the-input-file.html"><i class="fa fa-check"></i><b>3</b> The input file</a></li>
<li class="chapter" data-level="4" data-path="sec-settings.html"><a href="sec-settings.html"><i class="fa fa-check"></i><b>4</b> The settings file and command line arguments</a></li>
<li class="chapter" data-level="5" data-path="all-menu-options.html"><a href="all-menu-options.html"><i class="fa fa-check"></i><b>5</b> All menu options</a><ul>
<li class="chapter" data-level="5.1" data-path="all-menu-options.html"><a href="all-menu-options.html#option-1-hardy-weinberg-hw-exact-tests"><i class="fa fa-check"></i><b>5.1</b> Option 1: Hardy-Weinberg (HW) exact tests</a><ul>
<li class="chapter" data-level="5.1.1" data-path="all-menu-options.html"><a href="all-menu-options.html#sub-options-13-tests-for-each-locus-in-each-population"><i class="fa fa-check"></i><b>5.1.1</b> Sub-options 1–3: Tests for each locus in each population</a></li>
<li class="chapter" data-level="5.1.2" data-path="all-menu-options.html"><a href="all-menu-options.html#output"><i class="fa fa-check"></i><b>5.1.2</b> Output</a></li>
<li class="chapter" data-level="5.1.3" data-path="all-menu-options.html"><a href="all-menu-options.html#sub-options-45-global-tests-across-loci-or-across-samples"><i class="fa fa-check"></i><b>5.1.3</b> Sub-options 4,5: Global tests across loci or across samples</a></li>
<li class="chapter" data-level="5.1.4" data-path="all-menu-options.html"><a href="all-menu-options.html#analyzing-a-single-genotypic-matrix"><i class="fa fa-check"></i><b>5.1.4</b> Analyzing a single genotypic matrix</a></li>
<li class="chapter" data-level="5.1.5" data-path="all-menu-options.html"><a href="all-menu-options.html#code-checks"><i class="fa fa-check"></i><b>5.1.5</b> Code checks</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="all-menu-options.html"><a href="all-menu-options.html#option-2-tests-and-tables-for-linkage-disequilibrium"><i class="fa fa-check"></i><b>5.2</b> Option 2: Tests and tables for linkage disequilibrium</a><ul>
<li class="chapter" data-level="5.2.1" data-path="all-menu-options.html"><a href="all-menu-options.html#sub-option-1-tests"><i class="fa fa-check"></i><b>5.2.1</b> Sub-option 1: Tests</a></li>
<li class="chapter" data-level="5.2.2" data-path="all-menu-options.html"><a href="all-menu-options.html#output-1"><i class="fa fa-check"></i><b>5.2.2</b> Output</a></li>
<li class="chapter" data-level="5.2.3" data-path="all-menu-options.html"><a href="all-menu-options.html#sub-option-2-create-tables"><i class="fa fa-check"></i><b>5.2.3</b> Sub-option 2: create tables</a></li>
<li class="chapter" data-level="5.2.4" data-path="all-menu-options.html"><a href="all-menu-options.html#code-checks-1"><i class="fa fa-check"></i><b>5.2.4</b> Code checks</a></li>
</ul></li>
<li class="chapter" data-level="5.3" data-path="all-menu-options.html"><a href="all-menu-options.html#option-3-population-differentiation"><i class="fa fa-check"></i><b>5.3</b> Option 3: population differentiation</a><ul>
<li class="chapter" data-level="5.3.1" data-path="all-menu-options.html"><a href="all-menu-options.html#sub-options-1-or-2-genic-differentiation"><i class="fa fa-check"></i><b>5.3.1</b> Sub-options 1 or 2 (genic differentiation)</a></li>
<li class="chapter" data-level="5.3.2" data-path="all-menu-options.html"><a href="all-menu-options.html#sub-options-3-or-4-genotypic-differentiation"><i class="fa fa-check"></i><b>5.3.2</b> Sub-options 3 or 4 (genotypic differentiation)</a></li>
<li class="chapter" data-level="5.3.3" data-path="all-menu-options.html"><a href="all-menu-options.html#output-2"><i class="fa fa-check"></i><b>5.3.3</b> Output</a></li>
<li class="chapter" data-level="5.3.4" data-path="all-menu-options.html"><a href="all-menu-options.html#gene-diversity-as-a-test-statistic"><i class="fa fa-check"></i><b>5.3.4</b> Gene diversity as a test statistic</a></li>
<li class="chapter" data-level="5.3.5" data-path="all-menu-options.html"><a href="all-menu-options.html#analyzing-a-single-contingency-table"><i class="fa fa-check"></i><b>5.3.5</b> Analyzing a single contingency table</a></li>
<li class="chapter" data-level="5.3.6" data-path="all-menu-options.html"><a href="all-menu-options.html#code-checks-2"><i class="fa fa-check"></i><b>5.3.6</b> Code checks</a></li>
</ul></li>
<li class="chapter" data-level="5.4" data-path="all-menu-options.html"><a href="all-menu-options.html#option-4-private-alleles"><i class="fa fa-check"></i><b>5.4</b> Option 4: private alleles</a></li>
<li class="chapter" data-level="5.5" data-path="all-menu-options.html"><a href="all-menu-options.html#option-5-basic-information-f_mathrmis-and-gene-diversities"><i class="fa fa-check"></i><b>5.5</b> Option 5: Basic information, <span class="math inline">\(F_\mathrm{IS}\)</span>, and gene diversities</a><ul>
<li class="chapter" data-level="5.5.1" data-path="all-menu-options.html"><a href="all-menu-options.html#sub-option-1-allele-and-genotype-frequencies"><i class="fa fa-check"></i><b>5.5.1</b> Sub-option 1: Allele and genotype frequencies</a></li>
<li class="chapter" data-level="5.5.2" data-path="all-menu-options.html"><a href="all-menu-options.html#sub-option-2-identity-based-gene-diversities-and-f_mathrmis"><i class="fa fa-check"></i><b>5.5.2</b> Sub-option 2: Identity-based gene diversities and <span class="math inline">\(F_\mathrm{IS}\)</span></a></li>
<li class="chapter" data-level="5.5.3" data-path="all-menu-options.html"><a href="all-menu-options.html#sub-option-3-allele-size-based-gene-diversities-and-rho_mathrmis"><i class="fa fa-check"></i><b>5.5.3</b> Sub-option 3: Allele size-based gene diversities and <span class="math inline">\(\rho_{\mathrm{IS}}\)</span></a></li>
</ul></li>
<li class="chapter" data-level="5.6" data-path="all-menu-options.html"><a href="all-menu-options.html#option-6-fst-and-other-correlations-isolation-by-distance"><i class="fa fa-check"></i><b>5.6</b> Option 6: Fst and other correlations, isolation by distance</a><ul>
<li class="chapter" data-level="5.6.1" data-path="all-menu-options.html"><a href="all-menu-options.html#sub-options-14-f-statistics-and-rho-statistics"><i class="fa fa-check"></i><b>5.6.1</b> Sub-options 1–4: <span class="math inline">\(F\)</span>-statistics and <span class="math inline">\(\rho\)</span>-statistics</a></li>
<li class="chapter" data-level="5.6.2" data-path="all-menu-options.html"><a href="all-menu-options.html#sub-option-5-isolation-by-distance-between-individuals"><i class="fa fa-check"></i><b>5.6.2</b> Sub-option 5: isolation by distance between individuals</a></li>
<li class="chapter" data-level="5.6.3" data-path="all-menu-options.html"><a href="all-menu-options.html#sub-option-6-isolation-by-distance-between-groups"><i class="fa fa-check"></i><b>5.6.3</b> Sub-option 6: isolation by distance between groups</a></li>
<li class="chapter" data-level="5.6.4" data-path="all-menu-options.html"><a href="all-menu-options.html#former-sub-option-5-of-genepop-analysis-of-isolation-by-distance-from-a-genetic-distance-matrix"><i class="fa fa-check"></i><b>5.6.4</b> Former sub-option 5 of Genepop: analysis of isolation by distance from a genetic distance matrix</a></li>
<li class="chapter" data-level="5.6.5" data-path="all-menu-options.html"><a href="all-menu-options.html#user-provided-geographic-distance-matrices"><i class="fa fa-check"></i><b>5.6.5</b> User-provided geographic distance matrices</a></li>
<li class="chapter" data-level="5.6.6" data-path="all-menu-options.html"><a href="all-menu-options.html#analysis-of-isolation-by-distance-from-multiple-genetic-distance-matrices"><i class="fa fa-check"></i><b>5.6.6</b> Analysis of isolation by distance from multiple genetic distance matrices</a></li>
<li class="chapter" data-level="5.6.7" data-path="all-menu-options.html"><a href="all-menu-options.html#analysis-of-mean-differentiation"><i class="fa fa-check"></i><b>5.6.7</b> Analysis of mean differentiation</a></li>
</ul></li>
<li class="chapter" data-level="5.7" data-path="all-menu-options.html"><a href="all-menu-options.html#data-selection-for-analyses-of-isolation-by-distance"><i class="fa fa-check"></i><b>5.7</b> Data selection for analyses of isolation by distance</a><ul>
<li class="chapter" data-level="5.7.1" data-path="all-menu-options.html"><a href="all-menu-options.html#selecting-a-subset-of-samples"><i class="fa fa-check"></i><b>5.7.1</b> Selecting a subset of samples</a></li>
</ul></li>
<li class="chapter" data-level="5.8" data-path="all-menu-options.html"><a href="all-menu-options.html#option-7-file-conversions"><i class="fa fa-check"></i><b>5.8</b> Option 7: File conversions</a></li>
<li class="chapter" data-level="5.9" data-path="all-menu-options.html"><a href="all-menu-options.html#option-8-null-alleles-and-some-input-file-utilities"><i class="fa fa-check"></i><b>5.9</b> Option 8: Null alleles and some input file utilities</a><ul>
<li class="chapter" data-level="5.9.1" data-path="all-menu-options.html"><a href="all-menu-options.html#sub-option-1-null-alleles"><i class="fa fa-check"></i><b>5.9.1</b> Sub-option 1: null alleles</a></li>
<li class="chapter" data-level="5.9.2" data-path="all-menu-options.html"><a href="all-menu-options.html#sub-option-2-diploidisation-of-haploid-data"><i class="fa fa-check"></i><b>5.9.2</b> Sub-option 2: Diploidisation of haploid data</a></li>
<li class="chapter" data-level="5.9.3" data-path="all-menu-options.html"><a href="all-menu-options.html#sub-option-3-relabeling-alleles-names"><i class="fa fa-check"></i><b>5.9.3</b> Sub-option 3: Relabeling alleles names</a></li>
<li class="chapter" data-level="5.9.4" data-path="all-menu-options.html"><a href="all-menu-options.html#sub-options-4-and-5-conversion-of-population-data-to-individual-data"><i class="fa fa-check"></i><b>5.9.4</b> Sub-options 4 and 5: Conversion of population data to individual data</a></li>
<li class="chapter" data-level="5.9.5" data-path="all-menu-options.html"><a href="all-menu-options.html#sub-option-6-random-sampling-of-haploid-genotypes-from-diploid-ones"><i class="fa fa-check"></i><b>5.9.5</b> Sub-option 6: Random sampling of haploid genotypes from diploid ones</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="evaluating-the-performance-of-inferences-for-isolation-by-distance.html"><a href="evaluating-the-performance-of-inferences-for-isolation-by-distance.html"><i class="fa fa-check"></i><b>6</b> Evaluating the performance of inferences for Isolation by distance</a></li>
<li class="chapter" data-level="7" data-path="methods.html"><a href="methods.html"><i class="fa fa-check"></i><b>7</b> Methods</a><ul>
<li class="chapter" data-level="7.1" data-path="methods.html"><a href="methods.html#null-alleles"><i class="fa fa-check"></i><b>7.1</b> Null alleles</a></li>
<li class="chapter" data-level="7.2" data-path="methods.html"><a href="methods.html#exact-tests"><i class="fa fa-check"></i><b>7.2</b> Exact tests</a></li>
<li class="chapter" data-level="7.3" data-path="methods.html"><a href="methods.html#algorithms-for-exact-tests"><i class="fa fa-check"></i><b>7.3</b> Algorithms for exact tests</a></li>
<li class="chapter" data-level="7.4" data-path="methods.html"><a href="methods.html#accuracy-of-p-values-estimated-by-the-markov-chain-algorithms"><i class="fa fa-check"></i><b>7.4</b> Accuracy of P values estimated by the Markov chain algorithms</a></li>
<li class="chapter" data-level="7.5" data-path="methods.html"><a href="methods.html#test-statistics"><i class="fa fa-check"></i><b>7.5</b> Test statistics</a></li>
<li class="chapter" data-level="7.6" data-path="methods.html"><a href="methods.html#estimating-f-statistics-and-related-quantities"><i class="fa fa-check"></i><b>7.6</b> Estimating <span class="math inline">\(F\)</span>-statistics and related quantities</a><ul>
<li class="chapter" data-level="7.6.1" data-path="methods.html"><a href="methods.html#Fmulti"><i class="fa fa-check"></i><b>7.6.1</b> ANOVA estimators: single- and multilocus definitions</a></li>
<li class="chapter" data-level="7.6.2" data-path="methods.html"><a href="methods.html#rho-stats"><i class="fa fa-check"></i><b>7.6.2</b> Microsatellite allele sizes, <span class="math inline">\(R_\mathrm{ST}\)</span>, and <span class="math inline">\(\rho_\mathrm{ST}\)</span></a></li>
<li class="chapter" data-level="7.6.3" data-path="methods.html"><a href="methods.html#robertson-and-hills-estimator-of-f_mathrmis"><i class="fa fa-check"></i><b>7.6.3</b> Robertson and Hill’s estimator of <span class="math inline">\(F_\mathrm{IS}\)</span></a></li>
</ul></li>
<li class="chapter" data-level="7.7" data-path="methods.html"><a href="methods.html#bootstraps"><i class="fa fa-check"></i><b>7.7</b> Bootstraps</a></li>
<li class="chapter" data-level="7.8" data-path="methods.html"><a href="methods.html#mantel-test"><i class="fa fa-check"></i><b>7.8</b> Mantel test</a><ul>
<li class="chapter" data-level="7.8.1" data-path="methods.html"><a href="methods.html#misuse-1-tests-of-correlation-at-different-distance"><i class="fa fa-check"></i><b>7.8.1</b> Misuse 1: tests of correlation at different distance</a></li>
<li class="chapter" data-level="7.8.2" data-path="methods.html"><a href="methods.html#misuse-2-partial-mantel-tests"><i class="fa fa-check"></i><b>7.8.2</b> Misuse 2: partial Mantel tests</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="8" data-path="code-maintenance-credits-contact-etc-.html"><a href="code-maintenance-credits-contact-etc-.html"><i class="fa fa-check"></i><b>8</b> Code maintenance, credits, contact, etc.</a><ul>
<li class="chapter" data-level="8.1" data-path="code-maintenance-credits-contact-etc-.html"><a href="code-maintenance-credits-contact-etc-.html#code-maintenance"><i class="fa fa-check"></i><b>8.1</b> Code maintenance</a></li>
<li class="chapter" data-level="8.2" data-path="code-maintenance-credits-contact-etc-.html"><a href="code-maintenance-credits-contact-etc-.html#credits-for-the-current-version"><i class="fa fa-check"></i><b>8.2</b> Credits for the current version</a></li>
<li class="chapter" data-level="8.3" data-path="code-maintenance-credits-contact-etc-.html"><a href="code-maintenance-credits-contact-etc-.html#previous-history"><i class="fa fa-check"></i><b>8.3</b> Previous history</a></li>
<li class="chapter" data-level="8.4" data-path="code-maintenance-credits-contact-etc-.html"><a href="code-maintenance-credits-contact-etc-.html#contact"><i class="fa fa-check"></i><b>8.4</b> Contact</a><ul>
<li class="chapter" data-level="8.4.1" data-path="code-maintenance-credits-contact-etc-.html"><a href="code-maintenance-credits-contact-etc-.html#bug-fixes-since-release-of-genepop-version-3.4-in-may-2003-until-first-release-of-genepop-4.0"><i class="fa fa-check"></i><b>8.4.1</b> Bug fixes since release of Genepop version 3.4 in May 2003 until first release of Genepop 4.0:</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="9" data-path="copyright.html"><a href="copyright.html"><i class="fa fa-check"></i><b>9</b> Copyright</a></li>
<li class="chapter" data-level="" data-path="bibliography.html"><a href="bibliography.html"><i class="fa fa-check"></i>Bibliography</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Genepop version 4.7.5</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="all-menu-options" class="section level1">
<h1><span class="header-section-number">5</span> All menu options</h1>
<div id="option-1-hardy-weinberg-hw-exact-tests" class="section level2">
<h2><span class="header-section-number">5.1</span> Option 1: Hardy-Weinberg (HW) exact tests</h2>
<p>The following menu appears:</p>
<pre><code>Hardy Weinberg tests:

HW test for each locus in each population:
   H1 = Heterozygote deficiency.......1
   H1 = Heterozygote excess...........2
   Probability test...................3

Global test:
   H1 = Heterozygote deficiency.......4
   H1 = Heterozygote excess...........5

Main menu.............................6</code></pre>
<div id="sub-options-13-tests-for-each-locus-in-each-population" class="section level3">
<h3><span class="header-section-number">5.1.1</span> Sub-options 1–3: Tests for each locus in each population</h3>
<p>Three distinct tests are available, all concerned with the same null hypothesis (random union of gametes). The difference between them is the construction of the rejection zone. For the Probability test (sub-option 3), the probability of the observed sample is used to define the rejection zone, and the <span class="math inline">\(P\)</span>-value of the test corresponds to the sum of the probabilities of all tables (with the same allelic counts) with the same or lower probability. This is the “exact HW test” of <span class="citation">Haldane (1954)</span>, <span class="citation">Weir (1996)</span>, <span class="citation">Guo and Thompson (1992)</span> and others. When the alternative hypothesis of interest is heterozygote excess or deficiency, more powerful tests than the probability test can be used <span class="citation">(Rousset and Raymond 1995)</span>. One of them, the score test or <span class="math inline">\(U\)</span> test, is available here, either for heterozygote deficiency (sub-option 1) or heterozygote excess (sub-option 2). The multi-samples versions of these two tests are accessible through sub-options 4 or 5.</p>
<p>Two distinct algorithms are available: first, the complete enumeration method, as described by <span class="citation">Louis and Dempster (1987)</span>. This algorithm works for less than five alleles. As an exact <span class="math inline">\(P\)</span>-value is calculated by complete enumeration, no standard error is computed. Second, a Markov chain (MC) algorithm to estimate without bias the exact <span class="math inline">\(P\)</span>-value of this test <span class="citation">(Guo and Thompson 1992)</span>, and three parameters are needed to control this algorithm (see Section <a href="methods.html#algorithms-for-exact-tests">7.3</a>). These different values may be provided either at Genepop’s request, or through the <code>Dememorisation</code>, <code>BatchLength</code> and <code>BatchNumber</code> settings. Two results are provided for each test by the MC algorithm: the estimated <span class="math inline">\(P\)</span>-value associated with the null hypothesis of HW equilibrium, and the standard error (S.E.) of this estimate.</p>
<p>For all tests concerned with sub-options 1-3, there are three possible cases. The number of distinct alleles at each locus in each sample is</p>
<ul>
<li><p><strong>no more than 4:</strong> Genepop will give you the choice between the complete enumeration and the MC method. If you have less than 1000 individuals per sample, the complete enumeration is recommended. Otherwise, the MC method could be much faster. But there are no general rules, results are highly variable, depending also on allele frequencies.</p></li>
<li><p><strong>always 5 or more:</strong> Genepop will automatically perform only the MC method.</p></li>
<li><p><strong>sometimes higher than 4, sometimes not:</strong> For cases where the number of alleles is 4 or lower, Genepop will give you the choice between both methods. For the other situations (5 alleles or more in some samples), the MC method will be automatically performed.</p></li>
</ul>
<p>Whether one wants enumeration or MC methods to be performed can be specified at runtime, or otherwise by the <code>HWtests</code> setting, with options <code>HWtests=enumeration</code> and <code>HWtests=MCMC</code>. The default in the batch mode is <code>enumeration</code>.</p>
</div>
<div id="output" class="section level3">
<h3><span class="header-section-number">5.1.2</span> Output</h3>
<p>Results are stored in a file named as follows</p>
<table>
<thead>
<tr class="header">
<th>sub-option</th>
<th>Extension</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>1</td>
<td><em>yourdata</em><code>.D</code></td>
</tr>
<tr class="even">
<td>2</td>
<td><em>yourdata</em><code>.E</code></td>
</tr>
<tr class="odd">
<td>3</td>
<td><em>yourdata</em><code>.P</code></td>
</tr>
<tr class="even">
<td>4</td>
<td><em>yourdata</em><code>.DG</code></td>
</tr>
<tr class="odd">
<td>5</td>
<td><em>yourdata</em><code>.EG</code></td>
</tr>
</tbody>
</table>
<p>where <em>yourdata</em> is (throughout this document) the name of the input file.</p>
<p>For each test, several values are indicated on the same line: (i) the <span class="math inline">\(P\)</span>-value of the test (or “-” if no data were available, or only one allele was present, or two alleles were detected but one was represented by only one copy); (ii) the standard error (only if a MC method was used); (iii) two estimates of <span class="math inline">\(F_\mathrm{IS}\)</span>, <span class="citation">Weir and Cockerham (1984)</span>’s (1984) estimate (W&amp;C), and <span class="citation">Robertson and Hill (1984)</span>’s (1984) estimate (R&amp;H). The latter has a lower variance under the null hypothesis. Finally, the number of “steps” is given: for the complete enumeration algorithm this is the number of different genotypic matrices considered, and for the Markov chain algorithm the number of switches (change of genotypic matrice) performed.<a href="#fn11" class="footnoteRef" id="fnref11"><sup>11</sup></a></p>
</div>
<div id="sub-options-45-global-tests-across-loci-or-across-samples" class="section level3">
<h3><span class="header-section-number">5.1.3</span> Sub-options 4,5: Global tests across loci or across samples</h3>
<p>For sub-option 3, a global test across loci or across sample is constructed using Fisher’s method. This method (sometimes conservative because discrete probabilities are analyzed), is only performed for convenience and its relevance should be first established (e.g. statistical independence of loci).</p>
<p>General statistical theory shows that there is no uniformly better way to combine <span class="math inline">\(P\)</span>-values of different tests. When an alternative model is specified, it is possible to find a better way of combining results from different data sets than Fisher’s method, and usually not by combining <span class="math inline">\(P\)</span>-values. In the present context one such method is the multisample score test of <span class="citation">Rousset and Raymond (1995)</span>, which defines a global test across loci and/or across samples generalizing the tests of sub-options 1 and 2. The global tests are performed by sub-options 4 and 5, only by the MC algorithm. Independence of loci is also assumed for these global tests.</p>
<p>The output file reports global P value estimates and standard errors per population, per locus, and over all loci and populations. For each global P value, the average number of switches per test combined is also reported. Since it is tempting to reduce the chain length parameters in this option, special care is needed in checking this accuracy diagnostic (see p.41).<a href="#fn12" class="footnoteRef" id="fnref12"><sup>12</sup></a></p>
<p>This option generates several large temporary files. The space used temporarily by Genepop can be estimated as: (#of Loci+#of pop+1)*batches*(iterations per batch)*8 octets. For example it will require about 240 Mo of temporary hard disk space if you have 10 loci, 50 samples and if you use a chain of 500,000 steps (100 batches of 5000 iterations).</p>
</div>
<div id="analyzing-a-single-genotypic-matrix" class="section level3">
<h3><span class="header-section-number">5.1.4</span> Analyzing a single genotypic matrix</h3>
<p> It is possible to perform a single HW test independently of the Genepop input file. This option is not presented in the Genepop menu. You should have an input file with a genotypic matrix (which can be taken from the output file of option 5 and edited), and use the <code>HWfile</code> setting.<a href="#fn13" class="footnoteRef" id="fnref13"><sup>13</sup></a> When Genepop is launched in this way, the following menu will appear:</p>
<pre><code> HW test for each locus in each population:
    H1 = Heterozygote deficiency .................1
    H1 = Heterozygote excess .....................2
    Probability test .............................3

 Allele frequencies, expected genotypes, Fis .... 4
 Quit ........................................... 5</code></pre>
<p>All HW tests corresponding to options 1.1–3 of “regular” Genepop are available through options 1–3, and basic information similar to that given by regular option 5.1 is available through the present option 4. Results are stored at the end of your input file. The exact format of the input file is:</p>
<p><strong>First line:</strong> anything. Use this line to store information about your data.</p>
<p><strong>Second line:</strong> The number of alleles <span class="math inline">\(n\)</span>.</p>
<p><strong>Line three through</strong> <span class="math inline">\(n+2\)</span><strong>:</strong> the genotypic matrix (see example).</p>
<p><strong>Beyond line</strong> <span class="math inline">\(n+2\)</span> <strong>:</strong> anything (this is not read by the program).</p>
<p>An example with four alleles is:</p>
<pre><code> Human Monoamine Oxidase (MOAO) Data
 4
 2
 12 24
 30 34 54
 22 21 20 10</code></pre>
<p>If this file is named <code>MOAO</code>, you can analyze it by setting <code>HWfile=MOAO</code> in the settings; you can also set <code>HWfileOptions=1</code> to run option 1 without making your way through the menus. All this can be done through the console command line. For example</p>
<pre><code>Genepop HWFile=MOAO HWfileOptions=1,2,3,4</code></pre>
<p>will perform all four analyses available through the above menu. General settings <code>Dememorisation</code>, <code>BatchLength</code>, <code>BatchNumber</code>, and <code>Mode</code> all affect these analyses in the same way as they affect analyses of regular input files.</p>
</div>
<div id="code-checks" class="section level3">
<h3><span class="header-section-number">5.1.5</span> Code checks</h3>
<p> Code for HW tests has a now venerable history of testing. Early versions of Genepop were compared with the Exactp step in Biosys <span class="citation">(Swofford and Selander 1989)</span> for two allele cases, and with data published in <span class="citation">Louis and Dempster (1987)</span> and <span class="citation">Guo and Thompson (1992)</span> for more alleles. The sample files <code>LouisD87.txt</code> and <code>GuoT92.txt</code> contain two such test samples, in single-matrix format.</p>
</div>
</div>
<div id="option-2-tests-and-tables-for-linkage-disequilibrium" class="section level2">
<h2><span class="header-section-number">5.2</span> Option 2: Tests and tables for linkage disequilibrium</h2>
<p>The following menu appears:<a href="#fn14" class="footnoteRef" id="fnref14"><sup>14</sup></a></p>
<pre><code>Pairwise associations (haploid and genotypic disequilibrium):
      Test for each pair of loci in each population ......... 1
      Only create genotypic contingency tables .............. 2

Menu  ....................................................... 3</code></pre>
<div id="sub-option-1-tests" class="section level3">
<h3><span class="header-section-number">5.2.1</span> Sub-option 1: Tests</h3>
<p>For this option the null hypothesis is: “Genotypes at one locus are independent from genotypes at the other locus”. For a pair of diploid loci, no assumption is made about the gametic phase in double heterozygotes. In particular, it is not inferred assuming one-locus HW equilibrium, as such equilibrium is not assumed anywhere in the formulation of the test. The test is thus one of association between diploid genotypes at both loci, sometimes described as a test of the composite linkage disequilibrium <span class="citation">(Weir 1996, 126–28)</span>. For a haploid locus and a diploid one, a test of association between the haploid and diploid genotypes is computed (there is no concern about gametic phase in this case). This makes it easy to test for cyto-nuclear disequilibria. For a pair of loci with haploid information, a straightforward test of association of alleles at the two loci is computed.</p>
<p>The default test statistic is now the log likelihood ratio statistic (<span class="math inline">\(G\)</span>-test). However one can still perform probability tests (as implemented in earlier versions of Genepop) by using the <code>GameticDiseqTest=Proba</code> setting.</p>
<p>For a given pair of loci within one sample, the relevant information is represented by a contingency table looking e.g. like</p>
<pre><code>       GOT2
       1.1  1.3  3.3  1.7  3.7
EST    _________________________
 1.1   1    1    0    0    1      3
 1.2   16   6    1    3    2      28
       _________________________
       17   7    1    3    3      31</code></pre>
<p>for two diploid loci (<code>1.1</code>, etc., are the diploid genotypes at each locus). Contingency tables are created for all pairs of loci in each sample, then a <span class="math inline">\(G\)</span> test or a probability test for each table is computed for each table using the Markov chain algorithm of <span class="citation">Raymond and Rousset (1995a)</span>. The number of switches of the algorithm is given for each table analyzed.<a href="#fn15" class="footnoteRef" id="fnref15"><sup>15</sup></a></p>
</div>
<div id="output-1" class="section level3">
<h3><span class="header-section-number">5.2.2</span> Output</h3>
<p>Results are stored in the file <em>yourdata</em><code>.DIS</code>. Three intractable situations are indicated: empty tables (“No data”), table with one row or one column only (“No contingency table”), and tables for which all rows or all columns marginal sums are 1 (“No information”). For each locus pair within each sample, the unbiased estimate of the P-value is indicated, as well as the standard error. Next, a global test (Fisher’s method) for each pair of loci is performed across samples.</p>
<p>See also the next section for analysis of a single table.</p>
</div>
<div id="sub-option-2-create-tables" class="section level3">
<h3><span class="header-section-number">5.2.3</span> Sub-option 2: create tables</h3>
<p>Suboption 2 only generates the above contingency tables and stores them in the file <em>yourdata</em><code>.TAB</code></p>
</div>
<div id="code-checks-1" class="section level3">
<h3><span class="header-section-number">5.2.4</span> Code checks</h3>
<p>See code checks for Option 3.</p>
</div>
</div>
<div id="option-3-population-differentiation" class="section level2">
<h2><span class="header-section-number">5.3</span> Option 3: population differentiation</h2>
<p>The following menu appears:</p>
<pre><code> Testing population differentiation :

      Genic differentiation:
           for all populations ........................ 1
           for all pairs of populations ............... 2

      Genotypic differentiation:
           for all populations ........................ 3
           for all pairs of populations ............... 4

      Main menu  ...................................... 5</code></pre>
<p>All tests are based on Markov chain algorithms. The Markov chain parameters are controlled exactly as in option 1.</p>
<div id="sub-options-1-or-2-genic-differentiation" class="section level3">
<h3><span class="header-section-number">5.3.1</span> Sub-options 1 or 2 (genic differentiation)</h3>
<p> They are concerned with the distribution of alleles is the various samples. The null hypothesis tested is “alleles are drawn from the same distribution in all populations”. For each locus, the test is performed on a contingency table like this one:</p>
<pre><code>          Sub-Pop.  Alleles
                    1    2   Total
                    _______
           1        14   46   60
           2        6    76   82
           3        10   74   84
           4        4    58   62
                    _______
          Total     34   254  288</code></pre>
<p>For each locus, an unbiased estimate of the P-value is computed. The test statistic is either the probability of the sample conditional on marginal values, the <span class="math inline">\(G\)</span> log likelihood ratio, or the level of gene diversity. In the first case, the test is Fisher’s exact probability test, and the algorithm is described in <span class="citation">Raymond and Rousset (1995a)</span>. A simple modification of this algorithm is used for the exact <span class="math inline">\(G\)</span> test.<a href="#fn16" class="footnoteRef" id="fnref16"><sup>16</sup></a> Genepop’s default is the <span class="math inline">\(G\)</span> test. You can revert to Fisher’s test by using the <code>DifferentiationTest=Proba</code> setting. Finally, the level of gene diversity can be used as a test statistic when coupled with the <code>GeneDivRanks</code> setting (this was new to version 4.1; see Section <a href="all-menu-options.html#gene-diversity-as-a-test-statistic">5.3.4</a>).</p>
<p>For sub-option 2, the tests are the same, but they are performed for all pairs of samples for all loci.</p>
</div>
<div id="sub-options-3-or-4-genotypic-differentiation" class="section level3">
<h3><span class="header-section-number">5.3.2</span> Sub-options 3 or 4 (genotypic differentiation)</h3>
<p> are concerned with the distribution of diploid genotypes in the various populations. The null hypothesis tested is “genotypes are drawn from the same distribution in all populations”. For each locus, the test is performed on a contingency table like this one:</p>
<pre><code>                  Genotypes:
                  -------------------------
                  1    1   2   1   2   3
         Pop:     1    2   2   3   3   3   All
         ----
         Pop1     142  27  0   13  1   0   183
         Pop2     149  20  0   11  0   4   184
         Pop3     131  12  0   9   0   1   153
         Pop4     119  22  1   10  0   0   152
         Pop5     120  17  1   10  1   0   149
         Pop6     134  18  2   15  0   0   169
         Pop7     116  15  1   10  1   1   144
         Pop8     214  41  3   14  2   1   275
         Pop9     84   17  0   7   2   0   110
         Pop10    107  18  0   15  3   0   143
         Pop11    134  32  1   21  4   0   192
         Pop12    105  26  1   11  1   4   148
         Pop13    97   19  2   23  4   0   145
         Pop14    95   28  3   19  3   1   149

         All:     1747 312 15  188 22  12  2296</code></pre>
<p>An unbiased estimate of the P-value of a log-likelihood ratio (<span class="math inline">\(G\)</span>) based exact test is performed. For this test, the statistics defining the rejection zone is the <span class="math inline">\(G\)</span> value computed on the genic table derived from the genotypic one <span class="citation">(see Goudet et al. 1996 for the choice of this statistic)</span>, so that the rejection zone is defined as the sum of the probabilities of all tables (with the same marginal genotypic values as the observed one) having a <span class="math inline">\(G\)</span> value computed on the derived genic table higher than or equal to the observed <span class="math inline">\(G\)</span> value.</p>
<p>For sub-option 4, the test is the same but is performed for all pairs of samples for all loci.</p>
</div>
<div id="output-2" class="section level3">
<h3><span class="header-section-number">5.3.3</span> Output</h3>
<p>For the four sub-options, results are stored in a file named as follows:<a href="#fn17" class="footnoteRef" id="fnref17"><sup>17</sup></a></p>
<table>
<thead>
<tr class="header">
<th>sub-option</th>
<th>test</th>
<th>output file name</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>1</td>
<td>Probability test</td>
<td><em>yourdata</em><code>.PR</code></td>
</tr>
<tr class="even">
<td>1</td>
<td><span class="math inline">\(G\)</span></td>
<td><em>yourdata</em><code>.GE</code></td>
</tr>
<tr class="odd">
<td>2</td>
<td>Probability test</td>
<td><em>yourdata</em><code>.PR2</code></td>
</tr>
<tr class="even">
<td>2</td>
<td><span class="math inline">\(G\)</span></td>
<td><em>yourdata</em><code>.GE2</code></td>
</tr>
<tr class="odd">
<td>3</td>
<td><span class="math inline">\(G\)</span></td>
<td><em>yourdata</em><code>.G</code></td>
</tr>
<tr class="even">
<td>4</td>
<td><span class="math inline">\(G\)</span></td>
<td><em>yourdata</em><code>.2G2</code></td>
</tr>
</tbody>
</table>
<p>All contingency tables are saved in the output file. Two intractable situations are indicated: empty tables or tables with one row or one column only (“No table”), and tables for which all rows or all columns marginal sums are 1 (“No information”). Estimates of P-values are given, as well as (for sub-options 1 and 3) a combination of all test results (Fisher’s method), which assumes a statistical independence across loci. For sub-options 2 and 4, this combination of all tests across loci (Fisher’s method) is performed for each sample pair. The result <code>Highly sign.</code>[ificant] is reported when at least one of the individual tests being combined yielded a zero <span class="math inline">\(P\)</span>-value estimate.</p>
</div>
<div id="gene-diversity-as-a-test-statistic" class="section level3">
<h3><span class="header-section-number">5.3.4</span> Gene diversity as a test statistic</h3>

<pre><code> DifferentiationTest=GeneDiv
 GeneDivRanks=2,1,3,3,3</code></pre>
<p>  <code>DifferentiationTest=GeneDiv</code> makes Genepop use gene diversity as test statistic in tests of genetic differentiation (option 3). The test will look for a decrease in gene diversity from populations ranked first (value <code>1</code> in <code>GeneDivRanks</code>) to populations ranked last. This should work for both genic and genotypic tables, and for pairwise comparisons as well as for all populations, i.e. for all sub-options 3.1 to 3.4. The test statistic is <span class="math display">\[\sum_{\textrm{all subsamples $i$}}\sum_{j&gt;i} (Q_j-Q_i)(R_j-R_i)\]</span> where <span class="math inline">\(Q_i\)</span> is gene identity in subsample <span class="math inline">\(i\)</span> and <span class="math inline">\(R_i\)</span> is the <code>GeneDivRanks</code> value for this subsample.</p>

<p>This option also works on input files in contingency table format (<code>strucfile</code> setting). In that case each <em>row</em> of the table is interpreted as a new population.</p>
</div>
<div id="analyzing-a-single-contingency-table" class="section level3">
<h3><span class="header-section-number">5.3.5</span> Analyzing a single contingency table</h3>
<p> It is possible to analyse any contingency table independently of the Genepop input file. You should have an input file with a contingency table, and use the <code>strucFile</code> setting.<a href="#fn18" class="footnoteRef" id="fnref18"><sup>18</sup></a> This option is not presented in the Genepop menu. Both the <span class="math inline">\(G\)</span> and probability tests are available and performed as in option 3.1. Results are stored at the end of your input file. An example of input file is:</p>
<pre><code> Dull example
 6 5
 1   2  5 10 11
 2   0  8 11 15
 0   0  1  5  6
 10 15 20 51 55
 0   0  0  2  1
 4   5  6 11 10</code></pre>
<p>If this file is named <code>structest</code>, you can analyze it by writing <code>StrucFile=structest</code> in the settings file, or by the console command line</p>
<pre><code>Genepop StrucFile=structest</code></pre>
<p>The exact format of the input file is:</p>
<p><strong>First line:</strong> anything. Use this line to store information about your data.</p>
<p><strong>Second line:</strong> The numbers of rows (<span class="math inline">\(n\)</span>) and columns.</p>
<p><strong>Line three through</strong> <span class="math inline">\(n+2\)</span> <strong>:</strong> the contingency table (see example).</p>
<p><strong>Beyond line</strong> <span class="math inline">\(n+2\)</span> <strong>:</strong> anything (this is not read by the program).</p>
<p>The default is to perform a <span class="math inline">\(G\)</span> test, but as in options 3.1 and 3.2 you can revert to Fisher’s exact test by the setting <code>DifferentiationTest=Proba</code>.</p>
</div>
<div id="code-checks-2" class="section level3">
<h3><span class="header-section-number">5.3.6</span> Code checks</h3>
<p> Code for contingency tables also has a venerable history of testing. Early versions of Genepop were tested by comparison with published data <span class="citation">(e.g. Mehta and Patel 1983)</span> or by hand calculations. The example file <code>MehtaP83.txt</code> contains one such test sample.</p>
</div>
</div>
<div id="option-4-private-alleles" class="section level2">
<h2><span class="header-section-number">5.4</span> Option 4: private alleles</h2>
<p>This option provides a multilocus estimate of the effective number of migrants (<span class="math inline">\(Nm\)</span>)by Barton and Slatkin’s (1986) method. Three estimates of <span class="math inline">\(Nm\)</span> are provided, using the three regression lines published in that reference, and a corrected estimate is provided using the values from the closest regression line. Results are stored in the file <em>yourdata</em><code>.PRI</code>.</p>
</div>
<div id="option-5-basic-information-f_mathrmis-and-gene-diversities" class="section level2">
<h2><span class="header-section-number">5.5</span> Option 5: Basic information, <span class="math inline">\(F_\mathrm{IS}\)</span>, and gene diversities</h2>
<p>The following menu appears:</p>
<pre><code>      Allele and genotype frequencies per locus and per sample .. 1

      Gene diversities &amp; Fis :
                                  Using allele identity ......... 2
                                  Using allele size ............. 3

      Main menu ................................................. 4</code></pre>
<div id="sub-option-1-allele-and-genotype-frequencies" class="section level3">
<h3><span class="header-section-number">5.5.1</span> Sub-option 1: Allele and genotype frequencies</h3>
<p>This option provides basic information on the data set. The output file is saved in the file <em>yourdata</em><code>.INF</code>. For each locus in each sample, several variables are calculated:</p>
<ul>
<li><p>allele frequencies.</p></li>
<li><p>observed and expected genotype proportions.</p></li>
<li><p><span class="math inline">\(F_\mathrm{IS}\)</span> estimates for each allele following <span class="citation">Weir and Cockerham (1984)</span>.</p></li>
<li><p>global estimate of <span class="math inline">\(F_\mathrm{IS}\)</span> over alleles according to <span class="citation">Weir and Cockerham (1984)</span> (W&amp;C) and <span class="citation">Robertson and Hill (1984)</span> (R&amp;H).</p></li>
<li><p>observed and “expected” numbers of homozygotes and heterozygotes. “Expected” here means the expected numbers, conditional on observed allelic counts, under HW equilibrium; the difference from naive products of observed allele frequencies is sometimes called Levene’s correction, after <span class="citation">Levene (1949)</span>.</p></li>
<li><p>the genotypic matrix.</p></li>
</ul>
<p>A table of allele frequencies for each locus and for each sample is also computed.</p>
</div>
<div id="sub-option-2-identity-based-gene-diversities-and-f_mathrmis" class="section level3">
<h3><span class="header-section-number">5.5.2</span> Sub-option 2: Identity-based gene diversities and <span class="math inline">\(F_\mathrm{IS}\)</span></h3>
<p>This option takes the observed frequencies of identical pairs of genes as estimates (<span class="math inline">\(Q\)</span>) of corresponding probabilities of identity (<span class="math inline">\(Q\)</span>) and then simply computes diversities as <span class="math inline">\(1-Q\)</span>: gene diversity within individuals (<code>1-Qintra</code>), and among individuals within samples (<code>1-Qinter</code>), per locus per sample, and averaged over samples or over loci. One-locus <span class="math inline">\(F_\mathrm{IS}\)</span> estimates are also computed in a way consistent with <span class="citation">Weir and Cockerham (1984)</span>. No estimate is given when no information is available (e.g. no estimate of diversity between individuals within a sample when only one individual has been genotyped).</p>
<p>For haploid data, only the gene diversity among individuals is computed. Multilocus estimates ignore haploid loci, or on the contrary ignore diploid loci if the setting <code>EstimationPloidy=Haploid</code> is used. Single-locus estimates are computed for both haploid and diploid loci irrespective of this setting.</p>
<p>The output is saved in the file <em>yourdata</em><code>.DIV</code>.</p>
</div>
<div id="sub-option-3-allele-size-based-gene-diversities-and-rho_mathrmis" class="section level3">
<h3><span class="header-section-number">5.5.3</span> Sub-option 3: Allele size-based gene diversities and <span class="math inline">\(\rho_{\mathrm{IS}}\)</span></h3>
<p>Option 5.3 is analogous to option 5.2. It computes measures of diversity based on allele size, namely mean squared allele size differences within individuals (<code>MSDintra</code>), and among individuals within samples (<code>MSDinter</code>), per locus per sample, and averaged over samples or over loci. Corresponding <span class="math inline">\(\rho_\mathrm{IS}\)</span> (the <span class="math inline">\(F_\mathrm{IS}\)</span> analogue, see Section <a href="methods.html#rho-stats">7.6.2</a>) estimates are also computed. Allele size is the allele name unless it has been given through the <code>AlleleSizes</code> setting.</p>
<p>For haploid data, only the mean squared difference <code>MSDinter</code> among individuals is computed. Multilocus estimates ignore haploid loci, or on the contrary ignore diploid loci if the setting <code>EstimationPloidy=Haploid</code> is used. Single-locus estimates are computed for both haploid and diploid loci irrespective of this setting.</p>
<p>The output is saved in the file <em>yourdata</em><code>.MSD</code>.</p>
</div>
</div>
<div id="option-6-fst-and-other-correlations-isolation-by-distance" class="section level2">
<h2><span class="header-section-number">5.6</span> Option 6: Fst and other correlations, isolation by distance</h2>
<p>The following menu appears:</p>
<pre><code> Estimating spatial structure:

 The information considered is :
      --&gt; Allele identity (F-statistics)
                For all populations ............ 1
                For all population pairs ....... 2
      --&gt; Allele size (Rho-statistics)
                For all populations ............ 3
                For all population pairs ....... 4

 Isolation by distance
                between individuals ............ 5
                between groups.................. 6

    Main menu  ................................. 7</code></pre>
<table>
<caption><span id="tab:isolstats">Table 5.1: </span> Genetic distance statistics available in options 6.5 and 6.6</caption>
<colgroup>
<col width="9%" />
<col width="18%" />
<col width="21%" />
<col width="50%" />
</colgroup>
<thead>
<tr class="header">
<th>Data ploidy</th>
<th>pop = individual?</th>
<th><code>isolationStatistic</code> setting</th>
<th>Estimator used</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Diploid</td>
<td>Yes (option 6.5)</td>
<td><code>=a</code></td>
<td><span class="math inline">\(\hat{a}\)</span></td>
</tr>
<tr class="even">
<td>Diploid</td>
<td>Yes (option 6.5)</td>
<td><code>=e</code></td>
<td><span class="math inline">\(\hat{e}\)</span></td>
</tr>
<tr class="odd">
<td>Diploid</td>
<td>No (option 6.6)</td>
<td>none (default)</td>
<td><span class="math inline">\(F_\mathrm{ST}\)</span>/(1-<span class="math inline">\(F_\mathrm{ST}\)</span>)</td>
</tr>
<tr class="even">
<td>Diploid</td>
<td>No (option 6.6)</td>
<td><code>=singleGeneDiv</code></td>
<td><span class="math inline">\(F/(1-F)\)</span> variant with denominator common to all pairs</td>
</tr>
<tr class="odd">
<td>Haploid</td>
<td>Yes (option 6.5)</td>
<td>none (default)</td>
<td><span class="math inline">\(\hat{a}\)</span>-like statistic with stand-in for within-deme gene diversity</td>
</tr>
<tr class="even">
<td>Haploid</td>
<td>No (option 6.6)</td>
<td>none (default)</td>
<td><span class="math inline">\(F_\mathrm{ST}\)</span>/(1-<span class="math inline">\(F_\mathrm{ST}\)</span>)</td>
</tr>
<tr class="odd">
<td>Haploid</td>
<td>No (option 6.6)</td>
<td><code>=singleGeneDiv</code></td>
<td><span class="math inline">\(F/(1-F)\)</span> variant with denominator common to all pairs</td>
</tr>
</tbody>
</table>
<p>Suboptions 5 and 6 provide a variety of analyses of isolation by distance patterns, including bootstrap confidence intervals of the slope of spatial regression (or equivalently, for “neighborhood” size estimates). Starting with version 4.1, it is even possible to test given values of the slope, through the <code>testPoint</code> setting; and additional estimators (merely minor variation on a common logic) have been implemented, in particular for haploid data. Table <a href="all-menu-options.html#tab:isolstats">5.1</a> summarizes the choice of methods, each of which will now be detailed.</p>
<div id="sub-options-14-f-statistics-and-rho-statistics" class="section level3">
<h3><span class="header-section-number">5.6.1</span> Sub-options 1–4: <span class="math inline">\(F\)</span>-statistics and <span class="math inline">\(\rho\)</span>-statistics</h3>
<p>These options compute estimates of <span class="math inline">\(F_\mathrm{IS}\)</span>, <span class="math inline">\(F_\mathrm{IT}\)</span> and <span class="math inline">\(F_\mathrm{ST}\)</span> or analogous correlations for allele size, either for each pair of population (sub-options 2 and 4) or a single measure for all populations (sub-options 1 and 3). <span class="math inline">\(F_\mathrm{ST}\)</span> is estimated by a “weighted” analysis of variance <span class="citation">Cockerham (1973; Weir and Cockerham 1984)</span>, and the analogous measure of correlation in allele size (<span class="math inline">\(\rho_\mathrm{ST}\)</span>) is estimated by the same technique (see Section <a href="methods.html#rho-stats">7.6.2</a>). Multilocus estimates are computed as detailed in Section <a href="methods.html#Fmulti">7.6.1</a>). For haploid data, remember to use the <code>EstimationPloidy=Haploid</code> setting.</p>
<p>In sub-option 1, the output is saved in the file <em>yourdata</em><code>.FST</code>. Beyond <span class="math inline">\(F_\mathrm{IS}\)</span>, <span class="math inline">\(F_\mathrm{IT}\)</span> and <span class="math inline">\(F_\mathrm{ST}\)</span> estimates, estimation of within-individual gene diversity and within-population among-individual gene diversity are reported as in option 5.2.</p>
<p>In sub-option 2 (pairs of populations), single locus and multilocus estimates are written in the <em>yourdata</em><code>.ST2</code> file and multilocus estimates are also written in the <em>yourdata</em><code>.MIG</code> file in a format suitable for analysis of isolation by distance (see option 6.6 for further details).</p>
<p>Sub-option 3 is analogous to sub-option 1, but for allele-size based estimates. the output is saved in the file <em>yourdata</em><code>.RHO</code>. Beyond <span class="math inline">\(\rho_\mathrm{IS}\)</span>, <span class="math inline">\(\rho_\mathrm{IT}\)</span> and <span class="math inline">\(\rho_\mathrm{ST}\)</span> estimates, estimation of within-individual gene diversity and within-population among-individual gene diversity are reported as in option 5.3.</p>
<p>Sub-option 4 is analogous to sub-option 2, but for allele-size based estimates. Output file names are as in sub-option 2.</p>
</div>
<div id="sub-option-5-isolation-by-distance-between-individuals" class="section level3">
<h3><span class="header-section-number">5.6.2</span> Sub-option 5: isolation by distance between individuals</h3>
<p>This option allows analysis of isolation by distance between pairs of individuals. It provides estimates of “neighborhood size”, or more precisely of <span class="math inline">\(D\sigma^2\)</span>, the product of population density and axial mean square parent-offspring distance, derived from the slope of the regression of pairwise genetic statistics against geographical distance or log(distance) in linear or two-dimensional habitats, respectively. More details are described in <span class="citation">Rousset (2000)</span> (<span class="math inline">\(\hat{a}\)</span> statistic), <span class="citation">Leblois, Estoup, and Rousset (2003)</span> (bootstrap confidence intervals) and <span class="citation">Watts et al. (2007)</span> (<span class="math inline">\(\hat{e}\)</span> statistic). For haploid data, a proxy for the <span class="math inline">\(\hat{a}\)</span> statistic has been introduced in version 4.1.</p>
<p>The position of individuals must be specified as two coordinates standing for their name (i.e. before the comma on the line for each individual), and since each individual is considered as a sample, it must be separated by a <code>Pop</code>. An example of such input file is given below: The first individual is located at the point <span class="math inline">\(x = 0.0\)</span>, <span class="math inline">\(y = 15.0\)</span> (showing that the decimal separator is a period), the second at the point <span class="math inline">\(x = 0\)</span>, <span class="math inline">\(y =30\)</span>, etc. This example also shows that <em>individual identifiers can be added after these coordinates</em>.</p>
<pre><code> Title line: A really too small data set
 ADH Locus 1
 ADH #2
 ADH three
 ADH-4
 ADH-5
 Pop
 0.0 15.0,  0201 0303 0102 0302 1011
 Pop
 0 30 Second indiv,  0202 0301 0102 0303 1111
 Pop
 0 45,  0102 0401 0202 0102 1010
 Pop
 0 60,  0103 0202 0101 0202 1011
 Pop
 0 75,  0203 0204 0101 0102 1010
 POP
 15 15,      0102 0202 0201 0405 0807
 Pop
 15 30,      0102 0201 0201 0405 0307
 Pop
 15 45,      0201 0203 0101 0505 0402
 Pop
 15 60,      0201 0303 0301 0303 0603
 Pop
 15 75,      0101 0201 0301 0505 0807</code></pre>
<p><strong>Missing information</strong> arises when there is no genetic estimate (if a pair of individuals has no genotypes for the same locus, for example), or when geographic distance is zero and log(distance) is used. Genepop will correctly handle such missing information until it comes to the point where regression cannot be computed or there are not several loci to bootstrap over.</p>
<p>Options to be described within option 6.5 are: <span class="math inline">\(\hat{a}\)</span> or <span class="math inline">\(\hat{e}\)</span> pairwise statistics (for diploid data); log transformation for geographic distances; minimal geographic distance; coverage probability of confidence interval; testing a given value of the slope; Mantel test settings; conversion to genetic distance matrix in Phylip format. Allele-size based analogues of <span class="math inline">\(\hat{a}\)</span> or <span class="math inline">\(\hat{e}\)</span> can be defined, but they should perform very poorly <span class="citation">(Leblois, Estoup, and Rousset 2003; Rousset 2007)</span>, so such an analysis has been purposely disabled.</p>
<p><strong>Pairwise statistics for diploid data</strong>: They are selected by the setting <code>IsolationStatistic=a</code> or <code>=e</code>, or at runtime (in batch mode, the default is <span class="math inline">\(\hat{a}\)</span>). The <span class="math inline">\(\hat{e}\)</span> statistic is asymptotically biased in contrast to <span class="math inline">\(\hat{a}\)</span>, but has lower variance. The bias of the <span class="math inline">\(\hat{e}\)</span>-based slope is higher the more limited dispersal is, so it performs less well in the lower range of observed dispersal among various species. Confidence intervals are also biased <span class="citation">(Leblois, Estoup, and Rousset 2003; Watts et al. 2007)</span>, being too short in the direction of low <span class="math inline">\(D\sigma^2\)</span> values, and on the contrary conservative in the direction of low <span class="math inline">\(D\sigma^2\)</span> values. Based on the simulation results of <span class="citation">Watts et al. (2007)</span>, a provisional advice is to run analyses with both statistics, and to derive an upper bound for the <span class="math inline">\(D\sigma^2\)</span> confidence interval (CI), hence the lower bound for the regression slope, from <span class="math inline">\(\hat{e}\)</span> (which has CI shorter than <span class="math inline">\(\hat{a}\)</span>, though still conservative) and the other <span class="math inline">\(D\sigma^2\)</span> bound, hence the upper bound for the regression slope, from <span class="math inline">\(\hat{a}\)</span> (which has too short CI, but less biased than the <span class="math inline">\(\hat{e}\)</span> CI). When the <span class="math inline">\(\hat{e}\)</span>-based <span class="math inline">\(D\sigma^2\)</span> estimate is below 2500 (linear habitat) or 4 (two-dimensional habitat) it is suggested to derive both bounds from <span class="math inline">\(\hat{a}\)</span>.</p>
<p>Note that <span class="math inline">\(\hat{e}\)</span> is essentially Loiselle’s statistic <span class="citation">(Loiselle et al. 1995)</span>, which use in this context has previously been advocated by e.g. <span class="citation">Vekemans and Hardy (2004)</span>.</p>
<p>For <strong>haploid data</strong> (i.e. <code>EstimationPloidy=Haploid</code>) the denominators of the <span class="math inline">\(\hat{a}\)</span> and <span class="math inline">\(\hat{e}\)</span> statistics cannot be computed. Ideally the denominator should be the gene diversity among individuals that would compete for the same position, as could be estimated from “group” data. As a reasonable first substitute, Genepop uses a single estimate of gene diversity (from the total sample and for each locus) to compute the denominators for all pairs of individuals. This amount to assume that overall differentiation in the population is weak.</p>
<p><strong>Log transformation for geographic distances</strong>: This transformation is required for estimation of <span class="math inline">\(D\sigma^2\)</span> when dispersal occurs over a surface rather than over a linear habitat. It is the default option in batch mode. It can be turned on and off by the setting <code>GeographicScale=Log</code> or <code>=Linear</code> or equivalently by <code>Geometry=2D</code> or <code>=1D</code>.</p>
<p><strong>Coverage probability of confidence interval</strong> This is the target probability that the confidence interval contains the parameter value. The usage is to compute intervals with 95% coverage and equal 2.5% tails, and this is the default coverage in Genepop. This can be changed by the setting <code>CIcoverage</code>, e.g. <code>CIcoverage=0.99</code> will compute interval with target probabilities 0.5% that either the confidence interval is too low or too high (an unrealistically large number of loci may be necessary to achieve the latter precision).</p>
<p><strong>Minimal and maximal geographic distances:</strong> As discussed in <span class="citation">Rousset (1997)</span>, samples at small geographic distances are not expected to follow the simple theory of the regression method, so the program asks for a minimum geographical distance. Only pairwise comparisons of samples at larger distances are used to estimate the regression coefficient (all pairs are used for the Mantel test). The minimal distance may be specified by the setting <code>MinimalDistance=</code><em>value</em> or at runtime. This being said, it is wise to include all pairs in the estimation as no substantial bias is expected, and this avoids uncontrolled hacking of the data. Thus, the suggested minimal distance here is any distance large enough to exclude only pairs at zero geographical distance. Only non-negative values are accepted, and the default in batch mode is 0.0001.</p>
<p>There is also a setting <code>MaximalDistance=</code><em>value</em>. This should not be abused, and is (therefore) available only through the settings file, not as a runtime option.</p>
<p><strong>Testing a given value of the slope</strong> The setting <code>testPoint=0.00123</code> (say) returns the unidirectional P-value for a specific value of the slope, using the ABC bootstrap method. This is the reciprocal of a confidence interval computation: confidence intervals evaluate parameter values corresponding to given error levels, say the 0.025 and 0.975 unidirectional levels for a 95% bidirectional CI, while this option evaluates the unidirectional P-value associated with a given parameter value.</p>
<p><strong>Mantel test:</strong> The Mantel test is implemented. See Section <a href="methods.html#mantel-test">7.8</a> for limitations of this test. In the present context this is an exact test of the null hypothesis that there in no spatial correlation between genetic samples.</p>
<p>Up to version 4.3 Genepop implemented only a Mantel test based on the rank correlation. It now also implements, and performs by default, Mantel tests based on the regression coefficient for the “genetic distance” statistic used to quantify isolation by distance. The latter tests should generally be more congruent with the confidence intervals based on the same distances than the rank-based tests are. The rank test can now be performed by using the setting <code>MantelRankTest=</code> (no <em>value</em> needed).</p>
<p>Ideally the confidence interval for the slope should contain zero if and only if the Mantel test is non-significant. Some exceptions may occur as the bootstrap method is only approximate, but such exceptions appear to be rare. Exceptions may more commonly occur when the bootstrap is based on the regression of genetic “distance” and geographic distance over a selected range of the latter.</p>
<p>The number of permutations may be specified by the setting <code>MantelPermutations=</code><em>value</em>, or else at runtime. In batch mode, if no such value has been given the default behaviour is not to perform the test.</p>
<p><strong>Export genetic distance matrix in Phylip format</strong>. This option is activated by the setting <code>PhylipMatrix=</code> (no <em>value</em> needed). It may be useful, if you wish to use Phylip, to draw a tree based on genetic distances. A constant is added to all values if necessary so that all resulting distances are positive. Output is written in the file <em>yourdata</em><code>.PMA</code>. No further estimation or testing is done, so the name of the groups/individuals does not need to be their spatial coordinates.</p>
<p>Except for this export option, output files are:</p>
<ul>
<li><p>the <em>yourdata</em><code>.ISO</code> output file, containing (i) a genetic distance (<span class="math inline">\(\hat{a}\)</span> or <span class="math inline">\(\hat{e}\)</span>) half-matrix and a geographic (log-)distance half-matrix; missing information is reported as ‘<code>-</code>’; (ii) regression estimates and bootstrap confidence intervals; (iii) the result of testing a slope value (using <code>testPoint</code>); (iv) results of a Mantel test for evidence of isolation by distance, if requested; (v) a bootstrap interval for the intercept. The order of elements in the half-matrices is:</p>
<pre><code>       1     2     3
 2     x
 3     x     x
 4     x     x     x</code></pre></li>
<li><p>a <em>yourdata</em><code>.MIG</code> output file, containing the same genetic and geographic distances as in the <code>ISO</code> file, but with more digits, and without estimation or test results. This file was formerly useful as input for the Isolde program (see “Former option 5 of Genepop”, below), and is a bit redundant now.</p></li>
<li><p>a <em>yourdata</em><code>.GRA</code> output file, where again the genetic and geographic distances are reported, now as <span class="math inline">\((x,y)\)</span> coordinates for each pair of individuals (one per line). This is useful e.g. for importing the output into programs with good graphics. Pairs with missing values (either <span class="math inline">\(x\)</span> or <span class="math inline">\(y\)</span>) are not reported in this file.</p></li>
</ul>
</div>
<div id="sub-option-6-isolation-by-distance-between-groups" class="section level3">
<h3><span class="header-section-number">5.6.3</span> Sub-option 6: isolation by distance between groups</h3>
<p>This option is analogous to the previous one, but derives <span class="math inline">\(D\sigma^2\)</span> estimates from a regression of <span class="math inline">\(F_{\mathrm{ST}}\)</span>/(1-<span class="math inline">\(F_{\mathrm{ST}}\)</span>)* estimates to geographic distance in a linear habitat, or log(distance) in a two-dimensional habitat <span class="citation">(Rousset 1997)</span>.</p>
<p>Both diploid and haploid data (through <code>EstimationPloidy=Haploid</code>) are handled. Missing information is handled as in option 6.5. Input format is the same, except that some samples must contain several individuals. The coordinates of each sample are still contained in the name of each sample, that is in the name of the last individual in each sample.</p>
<p>In addition some allele-size based analyses are possible (by the setting <code>AllelicDistance=Size</code>) but again they are not advised in general. Further options within option 6.6 are: <code>isolationStatistic</code>; <code>SingleGeneDiv</code>; minimal geographic distance; log transformation for geographic distances; testing a given value of the slope; Mantel test settings; conversion to genetic distance matrix in Phylip format. They operate as described above for analyses between individuals, the only difference being the genetic distance used (see Table <a href="all-menu-options.html#tab:isolstats">5.1</a>). In particular, a minor variant of the <span class="math inline">\(F/(1-F)\)</span> estimator is introduced in version 4.1, by analogy to the “between individuals” estimators. Recall that <span class="math inline">\(F/(1-F)=(Q_r-Q_0)/(1-Q_0)\)</span> where <span class="math inline">\(1-Q_0\)</span> is the within-deme gene diversity. The <span class="math inline">\(F/(1-F)\)</span> method uses per-pair estimates of this within-deme gene diversity, which may not be best. With <code>IsolationStatistic=SingleGeneDiv</code> a single estimate is used for all pairwise statistics. In principe this should be better when small per-group samples are considered, but the generic <span class="math inline">\(F/(1-F)\)</span> method is still available as the default method. Limited testing so far suggests little effect of the choice of the statistic on inferences from samples with 10 haploid individuals per group and high overall diversity.</p>
<p>Output is written in three files <em>yourdata</em><code>.ISO</code>, <em>yourdata</em><code>.MIG</code>, and <em>yourdata</em><code>.GRA</code> with the same contents as in option 6.5, except for the nature of the genetic distances.</p>
</div>
<div id="former-sub-option-5-of-genepop-analysis-of-isolation-by-distance-from-a-genetic-distance-matrix" class="section level3">
<h3><span class="header-section-number">5.6.4</span> Former sub-option 5 of Genepop: analysis of isolation by distance from a genetic distance matrix</h3>
<p>That option (using the Isolde program) allowed one to perform the analyses of sub-options 5 and 6 from a file with two semi-matrices, one for genetic “distances” <span class="math inline">\(F_{\mathrm{ST}}\)</span> or whatever), the other for Euclidian distances. These analyses are now available through the <code>IsolationFile</code> setting. Most choices within options 6.5 and 6.6 are available through this option, and missing data are handled<a href="#fn19" class="footnoteRef" id="fnref19"><sup>19</sup></a> (see example below). However, it is not possible to compute nonparametric confidence intervals for the regression slope since per-locus information is not provided (remarkably, some software pretends to compute nonparametric intervals in this case). This option may serve as a general purpose program for Mantel tests. Of course, some settings (minimal geographic distance, the <span class="math inline">\(F/(1-F)\)</span> transformation, and the interpretation of one one-tailed <span class="math inline">\(P\)</span> value as a test of isolation by distance) make sense in the narrower inference context of options 6.5 and 6.6.</p>
<p>The option is called by <code>IsolationFile=</code><em>input file name</em> where the input file follows the format of the <em>yourdata</em><code>.MIG</code> file written by options 6.5 and 6.6, which may be used as models. An example is</p>
<pre><code> Lousy data                   &lt;------anything (comments)
 8 (an example)                      &lt;---# of samples (comments ignored)
 Fst estimates:                              &lt;---anything (comments)
  0.003
  0.18 0.107
  0.19 0.068  0.011
  0.20 0.664  0.665 0.009
  0.21 0.098    -   0.673  0.675
  0.22 0.048  0.682  0.683  0.017  0.001
  0.23 0.715  0.721  0.666  0.666  0.037 0.006
 distances:                          &lt;---anything (comments)
  158.0
  158.0 1215.0
  158.1 1213.0 2300.0
  158.2 2300.0    2.0 1057.0
  158.3 1055.0 2525.0 2525.0 1000.0
  158.4 1057.0 1055.0 2525.0 2525.0 1000.0
   - 3582.0 3582.0 3582.0 3582.0    1.0 2.222
 Anything after the second half matrix       &lt;----as it says
 is ignored</code></pre>
<p>The order of elements in the half-matrices is again</p>
<pre><code>       1     2      3
 2     x
 3     x     x
 4     x     x     x</code></pre>
<p>Again as in options 6.5 and 6.6, both missing genetic and geographic information (‘<code>-</code>’) are handled.</p>
<p>Output is written at the end of the input file, and as in options 6.5 and 6.6, <span class="math inline">\((x,y)\)</span> data points are also written in the file <em>yourdata</em><code>.GRA</code>.</p>
<p><code>Genepop IsolationFile=</code><em>input file name</em> <code>MantelRankTest=</code> will further replicate the rank test of the old Isolde program.</p>
</div>
<div id="user-provided-geographic-distance-matrices" class="section level3">
<h3><span class="header-section-number">5.6.5</span> User-provided geographic distance matrices</h3>
<p>The setting <code>geoDistFile=</code><em>file name</em><a href="#fn20" class="footnoteRef" id="fnref20"><sup>20</sup></a> can be used to provide a geographic distance matrix. Its format is that of other geographic distances matrices, with one required line of comment:</p>
<pre><code> Geographic distances:                 &lt;---anything (comments)
  21
  31 32
  41 42 43
  ...</code></pre>
<p>The number of samples does not need to be given.</p>
</div>
<div id="analysis-of-isolation-by-distance-from-multiple-genetic-distance-matrices" class="section level3">
<h3><span class="header-section-number">5.6.6</span> Analysis of isolation by distance from multiple genetic distance matrices</h3>
<p>If another program has generated <span class="math inline">\(F_{\mathrm{ST}}\)</span> or <span class="math inline">\(F_{\mathrm{ST}}\)</span>/(1 - <span class="math inline">\(F_{\mathrm{ST}}\)</span>) matrices for a number of loci, the computation of bootstrap confidence intervals is possible. Analysis of such data sets is allowed by the <code>MultiMigFile=</code><em>input file name</em> setting. The format of the input file is the same as for a single genetic matrix, except that it contains multiple matrices and that the number of genetic matrices must be given (third line of input):</p>
<pre><code> More lousy data
 8
 16 loci (for example)                 &lt;---# of samples (comments ignored)
 locus 1:                              &lt;---anything (comments)
...                                    &lt;-half matrix (not shown here)
 locus 2:                              &lt;---anything (comments)
...
...                                    &lt;-more loci and half matrices (not shown here)
...
 locus 16:                             &lt;---anything (comments)
...
 Geographic distances:                 &lt;---anything (comments)
  158.0
  158.0 1215.0
  158.1 1213.0 2300.0
  158.2 2300.0    2.0 1057.0
  158.3 1055.0 2525.0 2525.0 1000.0
  158.4 1057.0 1055.0 2525.0 2525.0 1000.0
   - 3582.0 3582.0 3582.0 3582.0    1.0 2.222
 Anything after the second half matrix       &lt;----as it says
 is ignored</code></pre>
<p>The main use of this option is to allow analyses based on genetic distances not considered in Genepop. If the same estimates are input as would be computed by Genepop, the results should be similar to those from options 6.5 and 6.6, but not identical in general, because Genepop’s bootstrap estimates are computed as ratio of weighted average numerators and denominators of genetic estimates, while <code>MultiMigFile</code> can only use weighted averages of the ratios, i.e. of the input genetic values.</p>
</div>
<div id="analysis-of-mean-differentiation" class="section level3">
<h3><span class="header-section-number">5.6.7</span> Analysis of mean differentiation</h3>
<p> It is possible to perform a bootstrap analysis of the mean pairwise differentiation, through all menu options that lead to bootstrap analyses of isolation by distance, when additionally using the setting <code>MeanDifferentiationTest=TRUE</code>. It takes into account selection of data by both <code>PopTypes</code> and range of geographical distances.</p>
</div>
</div>
<div id="data-selection-for-analyses-of-isolation-by-distance" class="section level2">
<h2><span class="header-section-number">5.7</span> Data selection for analyses of isolation by distance</h2>
<div id="selecting-a-subset-of-samples" class="section level3">
<h3><span class="header-section-number">5.7.1</span> Selecting a subset of samples</h3>
<p>  The settings <code>PopTypes</code> and <code>PopTypeSelection</code> have been developed to facilitate comparison of differentiation patterns within and among different ecotypes or host races. They are used as follows:</p>
<pre><code> PopTypes= 1 1 2 1 2 1 1 2 3 4
 PopTypeSelection=only 1
 // PopTypeSelection=inter 1 2
 // PopTypeSelection=all</code></pre>
<p><code>PopTypes</code> allows to distinguish different types of samples (e.g. different ecotypes) by integer indices. The number of indices must match the number of samples in the data file.</p>
<p><code>PopTypeSelection</code> allows performing analyses (genetic distance regressions, confidence intervals, Mantel tests) only on pairs of populations belonging to the types specified. That is, the genetic differentiation statistic among excluded pairs is not used in any of these analyses. The different choices are shown above: <code>all</code> excludes no pairs (this is the default value); <code>inter</code> <span class="math inline">\(a\)</span> <span class="math inline">\(b\)</span> will exclude all pairs that do not involve both types <span class="math inline">\(a\)</span> and <span class="math inline">\(b\)</span> (only two types can be specified); and <code>only</code> <span class="math inline">\(a\)</span> will exclude all pairs that involve a type different from <span class="math inline">\(a\)</span> (only one type can be specified). For the latter two choices, permutations are made only among samples from a given type. <code>inter_all_types</code> excludes all pairs within types; no Mantel test is performed in that case. <code>intra_all_types</code> keeps all pairs within types, and performs a single regression for all types; again, no Mantel test is performed in that case.</p>
<p>You have to perform the “<code>only</code>” and “<code>inter</code>” analyses in distinct Genepop runs if you wish to compare their results. <span class="citation">Rousset (1999)</span> explains how inferences can be made from such comparisons. Note that in this perspective, some comparison of the intercept may be useful and that Genepop also provides confidence intervals on the intercept at zero distance [or log(distance)].</p>
<p><em>The inter-type Mantel test may be misleading</em>. The null hypothesis implied by the permutation procedure is that there is no isolation by distance among populations within each type, rather than the often more relevant hypothesis that spatial processes within each type of populations are independent from each other. For this reason, a more appropriate test of the latter hypothesis is whether the bootstrap confidence interval for the inter-types regression slope includes zero or not.</p>
</div>
</div>
<div id="option-7-file-conversions" class="section level2">
<h2><span class="header-section-number">5.8</span> Option 7: File conversions</h2>
<p>This option allows the conversion of the Genepop input file toward other formats required by some other programs (the “ecumenical” function of Genepop). Given the limited interest in some of these conversions, little effort has been made to update them. In particular, data including haploid loci or in three-digits format may not be converted into valid input for the other programs.</p>
<p>The following menu appears:</p>
<pre><code> File conversion (diploid data, 2-digits coding only):

      GENEPOP --&gt; FSTAT (F statistics) ........................ 1
      GENEPOP --&gt; BIOSYS (letter code) ........................ 2
      GENEPOP --&gt; BIOSYS (number code) ........................ 3
      GENEPOP --&gt; LINKDOS (D statistics) ...................... 4

      Main menu  .............................................. 5</code></pre>
<p>Sub-option 1 converts the Genepop input file into the format required by the Fstat program of <span class="citation">Goudet (1995)</span>. The new format is saved in the file <em>yourdata</em><code>.DAT</code>.</p>
<p>Sub-options 2 and 3 converts the Genepop input file into the format required by Biosys <span class="citation">(Swofford and Selander 1989)</span>, either the letter or the number code. The new format is saved in the file <em>yourdata</em><code>.BIO</code>. You should add the STEP procedures at the end of this new file before running Biosys. Refer to the Biosys manual for details.</p>
<p>Sub-option 4 converts the Genepop input file into the format required by Linkdos, a program described by <span class="citation">Garnier-Géré and Dillmann (1992)</span> and based on <span class="citation">Black and Krafsur (1985)</span>. This program performs pairwise linkage disequilibria analyses in subdivided populations and <span class="citation">Ohta (1982)</span>’s (1982) <span class="math inline">\(D\)</span> statistics. The new format is saved in the file <em>yourdata</em><code>.LKD</code>. The source Linkdos program (LINKDOS.PAS) and an executable (LINKDOS.EXE) have been distributed with previous versions of Genepop with permission of their authors, and are still available on the <a href="http://kimura.univ-montp2.fr/~rousset/Genepop.htm">Genepop distribution page</a>. The executable distributed with Genepop has been compiled for 40 samples, 20 loci and 99 alleles per locus. It may be wise to relabel alleles (option 8.3) before the conversion. <span class="citation">Garnier-Géré and Dillmann (1992)</span> should be cited whenever this program is used.</p>
</div>
<div id="option-8-null-alleles-and-some-input-file-utilities" class="section level2">
<h2><span class="header-section-number">5.9</span> Option 8: Null alleles and some input file utilities</h2>
<p>The following menu appears<a href="#fn21" class="footnoteRef" id="fnref21"><sup>21</sup></a></p>
<pre><code> Miscellaneous :
    Null allele: estimates of allele frequencies .......... 1
    Diploidisation of haploid data ........................ 2
    Relabeling alleles .................................... 3
    Conversion to individual data with population names ... 4
    Conversion to individual data with individual names ... 5
    Random sampling of haploid genotypes from diploid ones  6

    Main Menu   ........................................... 7</code></pre>
<div id="sub-option-1-null-alleles" class="section level3">
<h3><span class="header-section-number">5.9.1</span> Sub-option 1: null alleles</h3>
<p> This sub-option allows estimation of gene frequencies when a null allele is present. Different methods are available: maximum likelihood, maximum likelihood with genotyping failure, and Brookfield’s (1996) estimator, which differences are explained in Section <a href="methods.html#null-alleles">7.1</a>.<a href="#fn22" class="footnoteRef" id="fnref22"><sup>22</sup></a></p>
<p>Genepop takes the allele with the highest number for a given locus <strong>across all populations</strong> as the null allele.<a href="#fn23" class="footnoteRef" id="fnref23"><sup>23</sup></a> For example, if you have 4 alleles plus a null allele, a null homozygote individual should be indicated as e.g. <code>0505</code> or <code>9999</code> in the input file.</p>
<p>The default estimation method is maximum likelihood, using the EM algorithm of <span class="citation">Dempster, Laird, and Rubin (1977)</span>. Apparent null genotypes may also be due to nonspecific genotyping failures. Joint maximum likelihood estimation of such failure rate (“<span class="math inline">\(\beta\)</span>”) and of allele frequencies is available through the setting <code>NullAlleleMethod=ApparentNulls</code>. Finally, the estimator of <span class="citation">Brookfield (1996)</span> is also available through the setting <code>NullAlleleMethod=B96</code>. Confidence intervals for null allele frequencies are computed for each locus in each population. Their coverage probability can be modified by the same setting <code>CIcoverage</code> as in options 6.5 and 6.6.</p>
<p>The output file is saved in the file <em>yourdata</em><code>.NUL</code>. This file may contain</p>
<ul>
<li><p>For the maximum likelihood methods, estimated allelic frequencies and predicted numbers of homozygotes and of heterozygotes with a null allele. For example, in an output such as</p>
<pre><code> Allele   EM freq.  Homoz.    Null Heter.
  1      0.2762    2.7046     4.2954
  2      0.2576    1.8500     3.1500
  3      0.2251    1.3567     2.6433
  4      0.0217    0.0000     0.0000
 Null    0.2193</code></pre>
<p>of the seven (<code>2.7046+4.2954</code>) apparent homozygotes for allele 1, it is predicted that 4.2954 are actually heterozygotes for allele 1 and for the null allele. This predicted value is the expected, or average, number of such heterozygotes over different samples with the same number of apparent genotypes, under the assumptions of the model.</p></li>
<li><p>a summary locus-by-population table of estimates of null allele frequencies.</p></li>
<li><p>a summary locus-by-population table of estimates of genotyping failure frequencies (“<code>beta</code>”), if applicable.</p></li>
<li><p>A table of bootstrap confidence intervals for estimates of null allele frequencies.</p></li>
</ul>
<p>Note that there may be insufficient information to compute estimates and/or confidence intervals: not enough alleles in the sample, for example. These are indicated by the message <code>No information</code>. Sometimes the point estimate can formally be computed but the computed CI is not meaningful. This happens for example in case of heterozygote excess, and generates a <code>(No info for CI)</code> warning (if all pseudo-samples generated by some resampling technique show an heterozygote excess, all pseudo-estimates of null allele frequency will be zero and there is no information to construct a non-null CI from this distribution).</p>
<p>The confidence intervals for null allele frequencies are obtained by a bootstrap method, and are <strong>not suitable</strong> for testing for the presence of null alleles, because the null hypothesis is at the boundary of the parameter space <span class="citation">(Andrews 2000)</span>. Instead, the exact score test for Hardy-Weinberg proportions can be used.</p>
</div>
<div id="sub-option-2-diploidisation-of-haploid-data" class="section level3">
<h3><span class="header-section-number">5.9.2</span> Sub-option 2: Diploidisation of haploid data</h3>
<p> This sub-option “diploidizes” haploid loci. For example, the line<br />
<code>popul 1, 01 02 10 00</code><br />
of an haploid dataset with 4 loci, will become<br />
<code>popul 1, 0101 0202 1010 0000</code>.<br />
Only haploid data are thus modified in a mixed haploid/diploid data file. The new file is named <code>D</code><em>yourdata</em>.<a href="#fn24" class="footnoteRef" id="fnref24"><sup>24</sup></a></p>
<p>Note that there may no longer be any need for this option for further analyses with Genepop (except perhaps as a preliminary to file conversions, option 7), since Genepop 4.0 now perform analyses on haploid data without such prior “diploidization” (don’t forget the <code>EstimationPloidy=Haploid</code> setting).</p>
</div>
<div id="sub-option-3-relabeling-alleles-names" class="section level3">
<h3><span class="header-section-number">5.9.3</span> Sub-option 3: Relabeling alleles names</h3>
<p> This sub-option relabels all alleles starting from 1 up to <span class="math inline">\(x\)</span>, <span class="math inline">\(x\)</span> being the true number of distinct alleles for each locus. The new file is named <code>N</code><em>yourdata</em>. The correspondence between the old and the new numbering is indicated in the file <em>new_file_name</em>.NUM. This option was originally introduced in Genepop because for some options, the memory space required depends on the highest allele number. I don’t expect this to be a cause of concern now.</p>
</div>
<div id="sub-options-4-and-5-conversion-of-population-data-to-individual-data" class="section level3">
<h3><span class="header-section-number">5.9.4</span> Sub-options 4 and 5: Conversion of population data to individual data</h3>
<p> These sub-options convert “population” data (with several individuals per <code>Pop</code> to “individual” data where each individual is put in a distinct <code>Pop</code>. This is useful for individual-based analyses of isolation by distance and, in this perspective, the name of each individual is replaced by what should be its coordinates, that is, either the name of the last individual in the original population (sub-option 4), or the name of each individual if their locations are distinguished (sub-option 5)<a href="#fn25" class="footnoteRef" id="fnref25"><sup>25</sup></a>.</p>
</div>
<div id="sub-option-6-random-sampling-of-haploid-genotypes-from-diploid-ones" class="section level3">
<h3><span class="header-section-number">5.9.5</span> Sub-option 6: Random sampling of haploid genotypes from diploid ones</h3>
<p>This sub-option randomly samples haploid genotypes at diploid loci.<a href="#fn26" class="footnoteRef" id="fnref26"><sup>26</sup></a> This may be useful for external analyses that require haploid data or that would be biased by Hardy-Weinberg disequilibria.</p>
</div>
</div>
</div>
<div class="footnotes">
<hr />
<ol start="11">
<li id="fn11"><p>New to Genepop 4.0.<a href="all-menu-options.html#fnref11">↩</a></p></li>
<li id="fn12"><p>Again new to Genepop 4.0.<a href="all-menu-options.html#fnref12">↩</a></p></li>
<li id="fn13"><p>In earlier versions of Genepop, this analysis was done through the HW.BAT batch file.<a href="all-menu-options.html#fnref13">↩</a></p></li>
<li id="fn14"><p>The distinct option 2.3 of Genepop 3.4 is no longer necessary as option 2.1 of Genepop 4.0 more gracefully handles haploid data.<a href="all-menu-options.html#fnref14">↩</a></p></li>
<li id="fn15"><p>This was not the case in earlier versions of Genepop<a href="all-menu-options.html#fnref15">↩</a></p></li>
<li id="fn16"><p>Up to version 3.4, Genepop only computed Fisher’s exact test in these sub-options.<a href="all-menu-options.html#fnref16">↩</a></p></li>
<li id="fn17"><p>slightly modified in comparison to earlier versions of Genepop<a href="all-menu-options.html#fnref17">↩</a></p></li>
<li id="fn18"><p>In previous versions of Genepop, this analysis was done by the Struc program called through the <code>Struc.BAT</code> batch file.<a href="all-menu-options.html#fnref18">↩</a></p></li>
<li id="fn19"><p>more extensively than in earlier versions of Genepop.<a href="all-menu-options.html#fnref19">↩</a></p></li>
<li id="fn20"><p>New to Genepop 4.2<a href="all-menu-options.html#fnref20">↩</a></p></li>
<li id="fn21"><p>Former sub-option 3 (erasing all temporary files) has been discarded.<a href="all-menu-options.html#fnref21">↩</a></p></li>
<li id="fn22"><p>The last two methods are new to Genepop 4.0.<a href="all-menu-options.html#fnref22">↩</a></p></li>
<li id="fn23"><p>This is a notable difference from Genepop 3.4, where the allele with the highest number in each population was taken as the null allele in this population. Consequently, null allele estimation is now meaningful even if no null homozygote is observed in a given population. The output format has also been improved, compared to earlier versions of Genepop, with a more logical ordering of results (samples within loci) and a final locus by population table of estimated null allele frequencies.<a href="all-menu-options.html#fnref23">↩</a></p></li>
<li id="fn24"><p>No longer truncated to 8 letters as it was in earlier versions of Genepop<a href="all-menu-options.html#fnref24">↩</a></p></li>
<li id="fn25"><p>New to Genepop 4.3<a href="all-menu-options.html#fnref25">↩</a></p></li>
<li id="fn26"><p>New to Genepop 4.3<a href="all-menu-options.html#fnref26">↩</a></p></li>
</ol>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="sec-settings.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="evaluating-the-performance-of-inferences-for-isolation-by-distance.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
},
"search": false
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:" && /^https?:/.test(src))
      src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
