<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Bibliography | Genepop version 4.7.5</title>
  <meta name="description" content="Bibliography | Genepop version 4.7.5" />
  <meta name="generator" content="bookdown 0.10 and GitBook 2.6.7" />

  <meta property="og:title" content="Bibliography | Genepop version 4.7.5" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Bibliography | Genepop version 4.7.5" />
  
  
  

<meta name="author" content="F. Rousset" />



  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="copyright.html">

<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />










</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="1" data-path="introduction.html"><a href="introduction.html"><i class="fa fa-check"></i><b>1</b> Introduction</a><ul>
<li class="chapter" data-level="1.1" data-path="introduction.html"><a href="introduction.html#purpose"><i class="fa fa-check"></i><b>1.1</b> Purpose</a></li>
<li class="chapter" data-level="1.2" data-path="introduction.html"><a href="introduction.html#the-two-genepop-distributions"><i class="fa fa-check"></i><b>1.2</b> The two Genepop distributions</a></li>
<li class="chapter" data-level="1.3" data-path="introduction.html"><a href="introduction.html#changes-since-version-4.0"><i class="fa fa-check"></i><b>1.3</b> Changes since version 4.0</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="installing-genepop-and-session-examples.html"><a href="installing-genepop-and-session-examples.html"><i class="fa fa-check"></i><b>2</b> Installing Genepop and session examples</a><ul>
<li class="chapter" data-level="2.1" data-path="installing-genepop-and-session-examples.html"><a href="installing-genepop-and-session-examples.html#installation"><i class="fa fa-check"></i><b>2.1</b> Installation</a><ul>
<li class="chapter" data-level="2.1.1" data-path="installing-genepop-and-session-examples.html"><a href="installing-genepop-and-session-examples.html#r-package"><i class="fa fa-check"></i><b>2.1.1</b> R package</a></li>
<li class="chapter" data-level="2.1.2" data-path="installing-genepop-and-session-examples.html"><a href="installing-genepop-and-session-examples.html#stand-alone-executable"><i class="fa fa-check"></i><b>2.1.2</b> Stand-alone executable</a></li>
</ul></li>
<li class="chapter" data-level="2.2" data-path="installing-genepop-and-session-examples.html"><a href="installing-genepop-and-session-examples.html#example-sessions"><i class="fa fa-check"></i><b>2.2</b> Example sessions</a><ul>
<li class="chapter" data-level="2.2.1" data-path="installing-genepop-and-session-examples.html"><a href="installing-genepop-and-session-examples.html#example-1-basic-session"><i class="fa fa-check"></i><b>2.2.1</b> Example 1: basic session</a></li>
<li class="chapter" data-level="2.2.2" data-path="installing-genepop-and-session-examples.html"><a href="installing-genepop-and-session-examples.html#example-2-using-the-settings-file"><i class="fa fa-check"></i><b>2.2.2</b> Example 2: using the settings file</a></li>
<li class="chapter" data-level="2.2.3" data-path="installing-genepop-and-session-examples.html"><a href="installing-genepop-and-session-examples.html#example-3-batch-processing"><i class="fa fa-check"></i><b>2.2.3</b> Example 3: Batch processing</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="the-input-file.html"><a href="the-input-file.html"><i class="fa fa-check"></i><b>3</b> The input file</a></li>
<li class="chapter" data-level="4" data-path="sec-settings.html"><a href="sec-settings.html"><i class="fa fa-check"></i><b>4</b> The settings file and command line arguments</a></li>
<li class="chapter" data-level="5" data-path="all-menu-options.html"><a href="all-menu-options.html"><i class="fa fa-check"></i><b>5</b> All menu options</a><ul>
<li class="chapter" data-level="5.1" data-path="all-menu-options.html"><a href="all-menu-options.html#option-1-hardy-weinberg-hw-exact-tests"><i class="fa fa-check"></i><b>5.1</b> Option 1: Hardy-Weinberg (HW) exact tests</a><ul>
<li class="chapter" data-level="5.1.1" data-path="all-menu-options.html"><a href="all-menu-options.html#sub-options-13-tests-for-each-locus-in-each-population"><i class="fa fa-check"></i><b>5.1.1</b> Sub-options 1–3: Tests for each locus in each population</a></li>
<li class="chapter" data-level="5.1.2" data-path="all-menu-options.html"><a href="all-menu-options.html#output"><i class="fa fa-check"></i><b>5.1.2</b> Output</a></li>
<li class="chapter" data-level="5.1.3" data-path="all-menu-options.html"><a href="all-menu-options.html#sub-options-45-global-tests-across-loci-or-across-samples"><i class="fa fa-check"></i><b>5.1.3</b> Sub-options 4,5: Global tests across loci or across samples</a></li>
<li class="chapter" data-level="5.1.4" data-path="all-menu-options.html"><a href="all-menu-options.html#analyzing-a-single-genotypic-matrix"><i class="fa fa-check"></i><b>5.1.4</b> Analyzing a single genotypic matrix</a></li>
<li class="chapter" data-level="5.1.5" data-path="all-menu-options.html"><a href="all-menu-options.html#code-checks"><i class="fa fa-check"></i><b>5.1.5</b> Code checks</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="all-menu-options.html"><a href="all-menu-options.html#option-2-tests-and-tables-for-linkage-disequilibrium"><i class="fa fa-check"></i><b>5.2</b> Option 2: Tests and tables for linkage disequilibrium</a><ul>
<li class="chapter" data-level="5.2.1" data-path="all-menu-options.html"><a href="all-menu-options.html#sub-option-1-tests"><i class="fa fa-check"></i><b>5.2.1</b> Sub-option 1: Tests</a></li>
<li class="chapter" data-level="5.2.2" data-path="all-menu-options.html"><a href="all-menu-options.html#output-1"><i class="fa fa-check"></i><b>5.2.2</b> Output</a></li>
<li class="chapter" data-level="5.2.3" data-path="all-menu-options.html"><a href="all-menu-options.html#sub-option-2-create-tables"><i class="fa fa-check"></i><b>5.2.3</b> Sub-option 2: create tables</a></li>
<li class="chapter" data-level="5.2.4" data-path="all-menu-options.html"><a href="all-menu-options.html#code-checks-1"><i class="fa fa-check"></i><b>5.2.4</b> Code checks</a></li>
</ul></li>
<li class="chapter" data-level="5.3" data-path="all-menu-options.html"><a href="all-menu-options.html#option-3-population-differentiation"><i class="fa fa-check"></i><b>5.3</b> Option 3: population differentiation</a><ul>
<li class="chapter" data-level="5.3.1" data-path="all-menu-options.html"><a href="all-menu-options.html#sub-options-1-or-2-genic-differentiation"><i class="fa fa-check"></i><b>5.3.1</b> Sub-options 1 or 2 (genic differentiation)</a></li>
<li class="chapter" data-level="5.3.2" data-path="all-menu-options.html"><a href="all-menu-options.html#sub-options-3-or-4-genotypic-differentiation"><i class="fa fa-check"></i><b>5.3.2</b> Sub-options 3 or 4 (genotypic differentiation)</a></li>
<li class="chapter" data-level="5.3.3" data-path="all-menu-options.html"><a href="all-menu-options.html#output-2"><i class="fa fa-check"></i><b>5.3.3</b> Output</a></li>
<li class="chapter" data-level="5.3.4" data-path="all-menu-options.html"><a href="all-menu-options.html#gene-diversity-as-a-test-statistic"><i class="fa fa-check"></i><b>5.3.4</b> Gene diversity as a test statistic</a></li>
<li class="chapter" data-level="5.3.5" data-path="all-menu-options.html"><a href="all-menu-options.html#analyzing-a-single-contingency-table"><i class="fa fa-check"></i><b>5.3.5</b> Analyzing a single contingency table</a></li>
<li class="chapter" data-level="5.3.6" data-path="all-menu-options.html"><a href="all-menu-options.html#code-checks-2"><i class="fa fa-check"></i><b>5.3.6</b> Code checks</a></li>
</ul></li>
<li class="chapter" data-level="5.4" data-path="all-menu-options.html"><a href="all-menu-options.html#option-4-private-alleles"><i class="fa fa-check"></i><b>5.4</b> Option 4: private alleles</a></li>
<li class="chapter" data-level="5.5" data-path="all-menu-options.html"><a href="all-menu-options.html#option-5-basic-information-f_mathrmis-and-gene-diversities"><i class="fa fa-check"></i><b>5.5</b> Option 5: Basic information, <span class="math inline">\(F_\mathrm{IS}\)</span>, and gene diversities</a><ul>
<li class="chapter" data-level="5.5.1" data-path="all-menu-options.html"><a href="all-menu-options.html#sub-option-1-allele-and-genotype-frequencies"><i class="fa fa-check"></i><b>5.5.1</b> Sub-option 1: Allele and genotype frequencies</a></li>
<li class="chapter" data-level="5.5.2" data-path="all-menu-options.html"><a href="all-menu-options.html#sub-option-2-identity-based-gene-diversities-and-f_mathrmis"><i class="fa fa-check"></i><b>5.5.2</b> Sub-option 2: Identity-based gene diversities and <span class="math inline">\(F_\mathrm{IS}\)</span></a></li>
<li class="chapter" data-level="5.5.3" data-path="all-menu-options.html"><a href="all-menu-options.html#sub-option-3-allele-size-based-gene-diversities-and-rho_mathrmis"><i class="fa fa-check"></i><b>5.5.3</b> Sub-option 3: Allele size-based gene diversities and <span class="math inline">\(\rho_{\mathrm{IS}}\)</span></a></li>
</ul></li>
<li class="chapter" data-level="5.6" data-path="all-menu-options.html"><a href="all-menu-options.html#option-6-fst-and-other-correlations-isolation-by-distance"><i class="fa fa-check"></i><b>5.6</b> Option 6: Fst and other correlations, isolation by distance</a><ul>
<li class="chapter" data-level="5.6.1" data-path="all-menu-options.html"><a href="all-menu-options.html#sub-options-14-f-statistics-and-rho-statistics"><i class="fa fa-check"></i><b>5.6.1</b> Sub-options 1–4: <span class="math inline">\(F\)</span>-statistics and <span class="math inline">\(\rho\)</span>-statistics</a></li>
<li class="chapter" data-level="5.6.2" data-path="all-menu-options.html"><a href="all-menu-options.html#sub-option-5-isolation-by-distance-between-individuals"><i class="fa fa-check"></i><b>5.6.2</b> Sub-option 5: isolation by distance between individuals</a></li>
<li class="chapter" data-level="5.6.3" data-path="all-menu-options.html"><a href="all-menu-options.html#sub-option-6-isolation-by-distance-between-groups"><i class="fa fa-check"></i><b>5.6.3</b> Sub-option 6: isolation by distance between groups</a></li>
<li class="chapter" data-level="5.6.4" data-path="all-menu-options.html"><a href="all-menu-options.html#former-sub-option-5-of-genepop-analysis-of-isolation-by-distance-from-a-genetic-distance-matrix"><i class="fa fa-check"></i><b>5.6.4</b> Former sub-option 5 of Genepop: analysis of isolation by distance from a genetic distance matrix</a></li>
<li class="chapter" data-level="5.6.5" data-path="all-menu-options.html"><a href="all-menu-options.html#user-provided-geographic-distance-matrices"><i class="fa fa-check"></i><b>5.6.5</b> User-provided geographic distance matrices</a></li>
<li class="chapter" data-level="5.6.6" data-path="all-menu-options.html"><a href="all-menu-options.html#analysis-of-isolation-by-distance-from-multiple-genetic-distance-matrices"><i class="fa fa-check"></i><b>5.6.6</b> Analysis of isolation by distance from multiple genetic distance matrices</a></li>
<li class="chapter" data-level="5.6.7" data-path="all-menu-options.html"><a href="all-menu-options.html#analysis-of-mean-differentiation"><i class="fa fa-check"></i><b>5.6.7</b> Analysis of mean differentiation</a></li>
</ul></li>
<li class="chapter" data-level="5.7" data-path="all-menu-options.html"><a href="all-menu-options.html#data-selection-for-analyses-of-isolation-by-distance"><i class="fa fa-check"></i><b>5.7</b> Data selection for analyses of isolation by distance</a><ul>
<li class="chapter" data-level="5.7.1" data-path="all-menu-options.html"><a href="all-menu-options.html#selecting-a-subset-of-samples"><i class="fa fa-check"></i><b>5.7.1</b> Selecting a subset of samples</a></li>
</ul></li>
<li class="chapter" data-level="5.8" data-path="all-menu-options.html"><a href="all-menu-options.html#option-7-file-conversions"><i class="fa fa-check"></i><b>5.8</b> Option 7: File conversions</a></li>
<li class="chapter" data-level="5.9" data-path="all-menu-options.html"><a href="all-menu-options.html#option-8-null-alleles-and-some-input-file-utilities"><i class="fa fa-check"></i><b>5.9</b> Option 8: Null alleles and some input file utilities</a><ul>
<li class="chapter" data-level="5.9.1" data-path="all-menu-options.html"><a href="all-menu-options.html#sub-option-1-null-alleles"><i class="fa fa-check"></i><b>5.9.1</b> Sub-option 1: null alleles</a></li>
<li class="chapter" data-level="5.9.2" data-path="all-menu-options.html"><a href="all-menu-options.html#sub-option-2-diploidisation-of-haploid-data"><i class="fa fa-check"></i><b>5.9.2</b> Sub-option 2: Diploidisation of haploid data</a></li>
<li class="chapter" data-level="5.9.3" data-path="all-menu-options.html"><a href="all-menu-options.html#sub-option-3-relabeling-alleles-names"><i class="fa fa-check"></i><b>5.9.3</b> Sub-option 3: Relabeling alleles names</a></li>
<li class="chapter" data-level="5.9.4" data-path="all-menu-options.html"><a href="all-menu-options.html#sub-options-4-and-5-conversion-of-population-data-to-individual-data"><i class="fa fa-check"></i><b>5.9.4</b> Sub-options 4 and 5: Conversion of population data to individual data</a></li>
<li class="chapter" data-level="5.9.5" data-path="all-menu-options.html"><a href="all-menu-options.html#sub-option-6-random-sampling-of-haploid-genotypes-from-diploid-ones"><i class="fa fa-check"></i><b>5.9.5</b> Sub-option 6: Random sampling of haploid genotypes from diploid ones</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="evaluating-the-performance-of-inferences-for-isolation-by-distance.html"><a href="evaluating-the-performance-of-inferences-for-isolation-by-distance.html"><i class="fa fa-check"></i><b>6</b> Evaluating the performance of inferences for Isolation by distance</a></li>
<li class="chapter" data-level="7" data-path="methods.html"><a href="methods.html"><i class="fa fa-check"></i><b>7</b> Methods</a><ul>
<li class="chapter" data-level="7.1" data-path="methods.html"><a href="methods.html#null-alleles"><i class="fa fa-check"></i><b>7.1</b> Null alleles</a></li>
<li class="chapter" data-level="7.2" data-path="methods.html"><a href="methods.html#exact-tests"><i class="fa fa-check"></i><b>7.2</b> Exact tests</a></li>
<li class="chapter" data-level="7.3" data-path="methods.html"><a href="methods.html#algorithms-for-exact-tests"><i class="fa fa-check"></i><b>7.3</b> Algorithms for exact tests</a></li>
<li class="chapter" data-level="7.4" data-path="methods.html"><a href="methods.html#accuracy-of-p-values-estimated-by-the-markov-chain-algorithms"><i class="fa fa-check"></i><b>7.4</b> Accuracy of P values estimated by the Markov chain algorithms</a></li>
<li class="chapter" data-level="7.5" data-path="methods.html"><a href="methods.html#test-statistics"><i class="fa fa-check"></i><b>7.5</b> Test statistics</a></li>
<li class="chapter" data-level="7.6" data-path="methods.html"><a href="methods.html#estimating-f-statistics-and-related-quantities"><i class="fa fa-check"></i><b>7.6</b> Estimating <span class="math inline">\(F\)</span>-statistics and related quantities</a><ul>
<li class="chapter" data-level="7.6.1" data-path="methods.html"><a href="methods.html#Fmulti"><i class="fa fa-check"></i><b>7.6.1</b> ANOVA estimators: single- and multilocus definitions</a></li>
<li class="chapter" data-level="7.6.2" data-path="methods.html"><a href="methods.html#rho-stats"><i class="fa fa-check"></i><b>7.6.2</b> Microsatellite allele sizes, <span class="math inline">\(R_\mathrm{ST}\)</span>, and <span class="math inline">\(\rho_\mathrm{ST}\)</span></a></li>
<li class="chapter" data-level="7.6.3" data-path="methods.html"><a href="methods.html#robertson-and-hills-estimator-of-f_mathrmis"><i class="fa fa-check"></i><b>7.6.3</b> Robertson and Hill’s estimator of <span class="math inline">\(F_\mathrm{IS}\)</span></a></li>
</ul></li>
<li class="chapter" data-level="7.7" data-path="methods.html"><a href="methods.html#bootstraps"><i class="fa fa-check"></i><b>7.7</b> Bootstraps</a></li>
<li class="chapter" data-level="7.8" data-path="methods.html"><a href="methods.html#mantel-test"><i class="fa fa-check"></i><b>7.8</b> Mantel test</a><ul>
<li class="chapter" data-level="7.8.1" data-path="methods.html"><a href="methods.html#misuse-1-tests-of-correlation-at-different-distance"><i class="fa fa-check"></i><b>7.8.1</b> Misuse 1: tests of correlation at different distance</a></li>
<li class="chapter" data-level="7.8.2" data-path="methods.html"><a href="methods.html#misuse-2-partial-mantel-tests"><i class="fa fa-check"></i><b>7.8.2</b> Misuse 2: partial Mantel tests</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="8" data-path="code-maintenance-credits-contact-etc-.html"><a href="code-maintenance-credits-contact-etc-.html"><i class="fa fa-check"></i><b>8</b> Code maintenance, credits, contact, etc.</a><ul>
<li class="chapter" data-level="8.1" data-path="code-maintenance-credits-contact-etc-.html"><a href="code-maintenance-credits-contact-etc-.html#code-maintenance"><i class="fa fa-check"></i><b>8.1</b> Code maintenance</a></li>
<li class="chapter" data-level="8.2" data-path="code-maintenance-credits-contact-etc-.html"><a href="code-maintenance-credits-contact-etc-.html#credits-for-the-current-version"><i class="fa fa-check"></i><b>8.2</b> Credits for the current version</a></li>
<li class="chapter" data-level="8.3" data-path="code-maintenance-credits-contact-etc-.html"><a href="code-maintenance-credits-contact-etc-.html#previous-history"><i class="fa fa-check"></i><b>8.3</b> Previous history</a></li>
<li class="chapter" data-level="8.4" data-path="code-maintenance-credits-contact-etc-.html"><a href="code-maintenance-credits-contact-etc-.html#contact"><i class="fa fa-check"></i><b>8.4</b> Contact</a><ul>
<li class="chapter" data-level="8.4.1" data-path="code-maintenance-credits-contact-etc-.html"><a href="code-maintenance-credits-contact-etc-.html#bug-fixes-since-release-of-genepop-version-3.4-in-may-2003-until-first-release-of-genepop-4.0"><i class="fa fa-check"></i><b>8.4.1</b> Bug fixes since release of Genepop version 3.4 in May 2003 until first release of Genepop 4.0:</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="9" data-path="copyright.html"><a href="copyright.html"><i class="fa fa-check"></i><b>9</b> Copyright</a></li>
<li class="chapter" data-level="" data-path="bibliography.html"><a href="bibliography.html"><i class="fa fa-check"></i>Bibliography</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Genepop version 4.7.5</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="bibliography" class="section level1 unnumbered">
<h1>Bibliography</h1>
<div id="refs" class="references">
<div>
<p>Anderson, Marti J. 2001. “Permutation Tests for Univariate or Multivariate Analysis of Variance and Regression.” <em>Can. J. Fish Aquatic Scis.</em> 58: 626–39.</p>
</div>
<div>
<p>Andrews, Donald W. K. 2000. “Inconsistency of the Bootstrap When a Parameter Is on the Boundary of the Parameter Space.” <em>Econometrica</em> 68 (2): 399–405.</p>
</div>
<div>
<p>Black, W. C., IV, and E. S. Krafsur. 1985. “A FORTRAN Program for the Calculation and Analysis of Two-Locus Linkage Disequilibrium Coefficients.” <em>Theor. Appl. Genetics</em> 70: 491–96.</p>
</div>
<div>
<p>Brookfield, J. F. Y. 1996. “A Simple New Method for Estimating Null Allele Frequency from Heterozygote Deficiency.” <em>Mol. Ecol</em> 5: 453–55.</p>
</div>
<div>
<p>Brooks, Stephen, and Andrew Gelman. 1998. “General Methods for Monitoring Convergence of Iterative Simulations.” <em>J. Comput. Graphical Statistics</em> 7: 434–55.</p>
</div>
<div>
<p>Chakraborty, R., M. de Andrade, S.P. Daiger, and B. Budowle. 1992. “Apparent Heterozygote Deficiencies Observed in DNA Typing Data and Their Implications in Forensic Applications.” <em>Ann. Hum. Genetics</em> 56: 45–57.</p>
</div>
<div>
<p>Cockerham, C. Clark. 1973. “Analyses of Gene Frequencies.” <em>Genetics</em> 74: 679–700.</p>
</div>
<div>
<p>Cockerham, C. Clark, and Bruce S. Weir. 1987. “Correlations, Descent Measures: Drift with Migration and Mutation.” <em>PNAS</em> 84: 8512–4.</p>
</div>
<div>
<p>Cox, D. R., and D. V. Hinkley. 1974. <em>Theoretical Statistics</em>. London: Chapman &amp; Hall.</p>
</div>
<div>
<p>Dempster, A. P., N. M. Laird, and D. B. Rubin. 1977. “Maximum Likelihood from Incomplete Data via the <em>EM</em> Algorithm (with Discussion).” <em>JRSSB</em> 39: 1–38.</p>
</div>
<div>
<p>DiCiccio, Thomas J., and Bradley Efron. 1996. “Bootstrap Confidence Intervals (with Discussion).” <em>Stat. Sci.</em> 11: 189–228.</p>
</div>
<div>
<p>Felsenstein, J. 2005. “PHYLIP (Phylogeny Inference Package) Version 3.6.”</p>
</div>
<div>
<p>Fisher, Ronald Aylmer. 1935. “The Logic of Inductive Inference (with Discussion).” <em>JRSS</em> 98: 39–82.</p>
</div>
<div>
<p>Garnier-Géré, P., and C. Dillmann. 1992. “A Computer Program for Testing Pairwise Linkage Disequilibria in Subdivided Populations.” <em>J. Hered.</em> 83: 239.</p>
</div>
<div>
<p>Goudet, J. 1995. “FSTAT (Version 1.2): A Computer Program to Calculate F-Statistics.” <em>J. Hered.</em> 86: 485–86.</p>
</div>
<div>
<p>Goudet, Jérome, Michel Raymond, Thierry de Meeüs, and François Rousset. 1996. “Testing Differentiation in Diploid Populations.” <em>Genetics</em> 144: 1931–8.</p>
</div>
<div>
<p>Guillot, G., and François Rousset. 2013. “Dismantling the Mantel Tests.” <em>Methods Ecol. Evol.</em> 4: 336–44.</p>
</div>
<div>
<p>Guo, Sun Wei, and Elizabeth A. Thompson. 1992. “Performing the Exact Test of Hardy-Weinberg Proportion for Multiple Alleles.” <em>Biometrics</em> 48: 361–72.</p>
</div>
<div>
<p>Haldane, John Burdon Sanderson. 1954. “An Exact Test for Randomness of Mating.” <em>Journal of Genetics</em> 52: 631–35.</p>
</div>
<div>
<p>Hartl, Daniel L., and Andrew G. Clark. 1989. <em>Principles of Population Genetics</em>. Second. Sunderland, Mass.: Sinauer.</p>
</div>
<div>
<p>Hastings, W. K. 1970. “Monte Carlo Sampling Methods Using Markov Chains and Their Applications.” <em>Biometrika</em> 57: 97–109.</p>
</div>
<div>
<p>Kalinowski, Steven T., and Mark L. Taper. 2006. “Maximum Likelihood Estimation of the Frequency of Null Alleles at Microsatellite Loci.” <em>Conserv. Genetics</em> 7: 991–95.</p>
</div>
<div>
<p>Leblois, R., P. Pudlo, J. Néron, F. Bertaux, C. R. Beeravolu, R. Vitalis, and F Rousset. 2014. “Maximum Likelihood Inference of Population Size Contractions from Microsatellite Data.” <em>Mol. Biol. Evol.</em> 31: 2805–23.</p>
</div>
<div>
<p>Leblois, Raphael, Arnaud Estoup, and François Rousset. 2003. “Influence of Mutational and Sampling Factors on the Estimation of Demographic Parameters in a ‘Continuous’ Population Under Isolation by Distance.” <em>Mol. Biol. Evol.</em> 20: 491–502.</p>
</div>
<div>
<p>Legendre, Pierre, and Marie-Josée Fortin. 2010. “Comparison of the Mantel Test and Alternative Approaches for Detecting Complex Multivariate Relationships in the Spatial Analysis of Genetic Data.” <em>Mol. Ecol. Resources</em> 10 (5): 831–44.</p>
</div>
<div>
<p>Legendre, Pierre, Marie-Josée Fortin, and Daniel Borcard. 2015. “Should the Mantel Test Be Used in Spatial Analysis?” <em>Methods Ecol. Evol.</em> 6: 1239–47.</p>
</div>
<div>
<p>Lehmann, E. L. 1994. <em>Testing Statistical Hypotheses</em>. Second. New York: Chapman &amp; Hall.</p>
</div>
<div>
<p>Levene, Howard. 1949. “On a Matching Problem Arising in Genetics.” <em>Annals of Mathematical Statistics</em> 20: 91–94.</p>
</div>
<div>
<p>Loiselle, Bette A., Victoria L. Sork, John Nason, and Catherine Graham. 1995. “Spatial Genetic Structure of a Tropical Understory Shrub <em>Psychotria Officinalis</em> (Rubiaceae).” <em>Am. J. Bot.</em> 82: 1420–5.</p>
</div>
<div>
<p>Louis, Edward J., and Everett R. Dempster. 1987. “An Exact Test for Hardy-Weinberg and Multiple Alleles.” <em>Biometrics</em> 43: 805–11.</p>
</div>
<div>
<p>Mantel, Nathan. 1967. “The Detection of Disease Clustering and a Generalized Regression Approach.” <em>Cancer Research</em> 27: 209–20.</p>
</div>
<div>
<p>Mehta, Cyrus R., and Nitin R. Patel. 1983. “A Network Algorithm for Performing Fisher’s Exact Test in <span class="math inline">\(r \times
	c\)</span> Contingency Tables.” <em>JASA</em> 78: 427–34.</p>
</div>
<div>
<p>Michalakis, Yannis, and Laurent Excoffier. 1996. “A Generic Estimation of Population Subdivision Using Distances Between Alleles with Special Interest to Microsatellite Loci.” <em>Genetics</em> 142: 1061–4.</p>
</div>
<div>
<p>Oden, Neal L., and Robert R. Sokal. 1992. “An Investigation of Three-Matrix Permutation Tests.” <em>J. Classif.</em> 9: 275–90.</p>
</div>
<div>
<p>Ohta, Tomoko. 1982. “Linkage Disequilibrium Due to Random Genetic Drift in Finite Subdivided Populations.” <em>PNAS</em> 79: 1940–4.</p>
</div>
<div>
<p>Raufaste, Nathalie, and François Rousset. 2001. “Are Partial Mantel Tests Adequate?” <em>Evolution</em> 55: 1703–5.</p>
</div>
<div>
<p>Raymond, Michel, and François Rousset. 1995a. “An Exact Test for Population Differentiation.” <em>Evolution</em> 49: 1283–6.</p>
</div>
<div>
<p>———. 1995b. “GENEPOP Version 1.2: Population Genetics Software for Exact Tests and Ecumenicism.” <em>J. Hered.</em> 86: 248–49.</p>
</div>
<div>
<p>Robertson, Alan, and William G. Hill. 1984. “Deviations from Hardy-Weinberg Proportions: Sampling Variances and Use in Estimation of Inbreeding Coefficients.” <em>Genetics</em> 107: 703–18.</p>
</div>
<div>
<p>Rousset, François. 1996. “Equilibrium Values of Measures of Population Subdivision for Stepwise Mutation Processes.” <em>Genetics</em> 142: 1357–62.</p>
</div>
<div>
<p>———. 1997. “Genetic Differentiation and Estimation of Gene Flow from <span class="math inline">\(F\)</span>-Statistics Under Isolation by Distance.” <em>Genetics</em> 145: 1219–28.</p>
</div>
<div>
<p>———. 1999. “Genetic Differentiation Within and Between Two Habitats.” <em>Genetics</em> 151: 397–407.</p>
</div>
<div>
<p>———. 2000. “Genetic Differentiation Between Individuals.” <em>J. Evol. Biol.</em> 13: 58–62.</p>
</div>
<div>
<p>———. 2002a. “Inbreeding and Relatedness Coefficients: What Do They Measure?” <em>Heredity</em> 88: 371–80.</p>
</div>
<div>
<p>———. 2002b. “Partial Mantel Tests: Reply to Castellano and Balletto.” <em>Evolution</em> 56: 1874–5.</p>
</div>
<div>
<p>———. 2007. “Inferences from Spatial Population Genetics.” In <em>Handbook of Statistical Genetics</em>, edited by D. J. Balding, M. Bishop, and C. Cannings, third, 945–79. Chichester, U.K.: Wiley.</p>
</div>
<div>
<p>Rousset, François, and Raphaël Leblois. 2007. “Likelihood and Approximate Likelihood Analyses of Genetic Structure in a Linear Habitat: Performance and Robustness to Model Mis-Specification.” <em>Mol. Biol. Evol.</em> 24: 2730–45.</p>
</div>
<div>
<p>Rousset, François, and Raphaël Leblois. 2012. “Likelihood-Based Inferences Under Isolation by Distance: Two-Dimensional Habitats and Confidence Intervals.” <em>Mol. Biol. Evol.</em> 29: 957–73.</p>
</div>
<div>
<p>Rousset, François, and Michel Raymond. 1995. “Testing Heterozygote Excess and Deficiency.” <em>Genetics</em> 140: 1413–9.</p>
</div>
<div>
<p>———. 1997. “Statistical Analyses of Population Genetic Data: Old Tools, New Concepts.” <em>Tr. Ecol. Evol.</em> 12: 313–17.</p>
</div>
<div>
<p>Slatkin, Montgomery. 1995. “A Measure of Population Subdivision Based on Microsatellite Allele Frequencies.” <em>Genetics</em> 139: 457–62.</p>
</div>
<div>
<p>Swofford, D. L., and R. B. Selander. 1989. <em>BIOSYS-1. a Computer Program for the Analysis of Allelic Variation in Population Genetics and Biochemical Systematics. Release 1.7.</em> Champaign: Illinois Natural History Survey.</p>
</div>
<div>
<p>Vekemans, X., and O. J. Hardy. 2004. “New Insights from Fine-Scale Spatial Genetic Structure Analyses in Plant Populations.” <em>Mol. Ecol.</em> 13: 921–34.</p>
</div>
<div>
<p>Watts, Phillip C., François Rousset, Ilik J. Saccheri, Raphaël Leblois, Stephen J. Kemp, and David J. Thompson. 2007. “Compatible Genetic and Ecological Estimates of Dispersal Rates in Insect (<em>Coenagrion Mercuriale</em>: Odonata: Zygoptera) Populations: Analysis of ‘Neighbourhood Size’ Using a More Precise Estimator.” <em>Mol. Ecol.</em> 16: 737–51.</p>
</div>
<div>
<p>Weir, B. S., and C. Clark Cockerham. 1984. “Estimating <span class="math inline">\(F\)</span>-Statistics for the Analysis of Population Structure.” <em>Evolution</em> 38: 1358–70.</p>
</div>
<div>
<p>Weir, Bruce S. 1996. <em>Genetic Data Analysis Ii</em>. Sunderland, Mass.: Sinauer.</p>
</div>
<div>
<p>Whitlock, M. C. 2005. “Combining Probability from Independent Tests: The Weighted Z-Method Is Superior to Fisher’s Approach.” <em>J. Evol. Biol.</em> 18 (5): 1368–73.</p>
</div>
</div>
</div>































            </section>

          </div>
        </div>
      </div>
<a href="copyright.html" class="navigation navigation-prev navigation-unique" aria-label="Previous page"><i class="fa fa-angle-left"></i></a>

    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
},
"search": false
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:" && /^https?:/.test(src))
      src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
