<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>2 Installing Genepop and session examples | Genepop version 4.7.5</title>
  <meta name="description" content="2 Installing Genepop and session examples | Genepop version 4.7.5" />
  <meta name="generator" content="bookdown 0.10 and GitBook 2.6.7" />

  <meta property="og:title" content="2 Installing Genepop and session examples | Genepop version 4.7.5" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="2 Installing Genepop and session examples | Genepop version 4.7.5" />
  
  
  

<meta name="author" content="F. Rousset" />



  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="introduction.html">
<link rel="next" href="the-input-file.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />










</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="1" data-path="introduction.html"><a href="introduction.html"><i class="fa fa-check"></i><b>1</b> Introduction</a><ul>
<li class="chapter" data-level="1.1" data-path="introduction.html"><a href="introduction.html#purpose"><i class="fa fa-check"></i><b>1.1</b> Purpose</a></li>
<li class="chapter" data-level="1.2" data-path="introduction.html"><a href="introduction.html#the-two-genepop-distributions"><i class="fa fa-check"></i><b>1.2</b> The two Genepop distributions</a></li>
<li class="chapter" data-level="1.3" data-path="introduction.html"><a href="introduction.html#changes-since-version-4.0"><i class="fa fa-check"></i><b>1.3</b> Changes since version 4.0</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="installing-genepop-and-session-examples.html"><a href="installing-genepop-and-session-examples.html"><i class="fa fa-check"></i><b>2</b> Installing Genepop and session examples</a><ul>
<li class="chapter" data-level="2.1" data-path="installing-genepop-and-session-examples.html"><a href="installing-genepop-and-session-examples.html#installation"><i class="fa fa-check"></i><b>2.1</b> Installation</a><ul>
<li class="chapter" data-level="2.1.1" data-path="installing-genepop-and-session-examples.html"><a href="installing-genepop-and-session-examples.html#r-package"><i class="fa fa-check"></i><b>2.1.1</b> R package</a></li>
<li class="chapter" data-level="2.1.2" data-path="installing-genepop-and-session-examples.html"><a href="installing-genepop-and-session-examples.html#stand-alone-executable"><i class="fa fa-check"></i><b>2.1.2</b> Stand-alone executable</a></li>
</ul></li>
<li class="chapter" data-level="2.2" data-path="installing-genepop-and-session-examples.html"><a href="installing-genepop-and-session-examples.html#example-sessions"><i class="fa fa-check"></i><b>2.2</b> Example sessions</a><ul>
<li class="chapter" data-level="2.2.1" data-path="installing-genepop-and-session-examples.html"><a href="installing-genepop-and-session-examples.html#example-1-basic-session"><i class="fa fa-check"></i><b>2.2.1</b> Example 1: basic session</a></li>
<li class="chapter" data-level="2.2.2" data-path="installing-genepop-and-session-examples.html"><a href="installing-genepop-and-session-examples.html#example-2-using-the-settings-file"><i class="fa fa-check"></i><b>2.2.2</b> Example 2: using the settings file</a></li>
<li class="chapter" data-level="2.2.3" data-path="installing-genepop-and-session-examples.html"><a href="installing-genepop-and-session-examples.html#example-3-batch-processing"><i class="fa fa-check"></i><b>2.2.3</b> Example 3: Batch processing</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="the-input-file.html"><a href="the-input-file.html"><i class="fa fa-check"></i><b>3</b> The input file</a></li>
<li class="chapter" data-level="4" data-path="sec-settings.html"><a href="sec-settings.html"><i class="fa fa-check"></i><b>4</b> The settings file and command line arguments</a></li>
<li class="chapter" data-level="5" data-path="all-menu-options.html"><a href="all-menu-options.html"><i class="fa fa-check"></i><b>5</b> All menu options</a><ul>
<li class="chapter" data-level="5.1" data-path="all-menu-options.html"><a href="all-menu-options.html#option-1-hardy-weinberg-hw-exact-tests"><i class="fa fa-check"></i><b>5.1</b> Option 1: Hardy-Weinberg (HW) exact tests</a><ul>
<li class="chapter" data-level="5.1.1" data-path="all-menu-options.html"><a href="all-menu-options.html#sub-options-13-tests-for-each-locus-in-each-population"><i class="fa fa-check"></i><b>5.1.1</b> Sub-options 1–3: Tests for each locus in each population</a></li>
<li class="chapter" data-level="5.1.2" data-path="all-menu-options.html"><a href="all-menu-options.html#output"><i class="fa fa-check"></i><b>5.1.2</b> Output</a></li>
<li class="chapter" data-level="5.1.3" data-path="all-menu-options.html"><a href="all-menu-options.html#sub-options-45-global-tests-across-loci-or-across-samples"><i class="fa fa-check"></i><b>5.1.3</b> Sub-options 4,5: Global tests across loci or across samples</a></li>
<li class="chapter" data-level="5.1.4" data-path="all-menu-options.html"><a href="all-menu-options.html#analyzing-a-single-genotypic-matrix"><i class="fa fa-check"></i><b>5.1.4</b> Analyzing a single genotypic matrix</a></li>
<li class="chapter" data-level="5.1.5" data-path="all-menu-options.html"><a href="all-menu-options.html#code-checks"><i class="fa fa-check"></i><b>5.1.5</b> Code checks</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="all-menu-options.html"><a href="all-menu-options.html#option-2-tests-and-tables-for-linkage-disequilibrium"><i class="fa fa-check"></i><b>5.2</b> Option 2: Tests and tables for linkage disequilibrium</a><ul>
<li class="chapter" data-level="5.2.1" data-path="all-menu-options.html"><a href="all-menu-options.html#sub-option-1-tests"><i class="fa fa-check"></i><b>5.2.1</b> Sub-option 1: Tests</a></li>
<li class="chapter" data-level="5.2.2" data-path="all-menu-options.html"><a href="all-menu-options.html#output-1"><i class="fa fa-check"></i><b>5.2.2</b> Output</a></li>
<li class="chapter" data-level="5.2.3" data-path="all-menu-options.html"><a href="all-menu-options.html#sub-option-2-create-tables"><i class="fa fa-check"></i><b>5.2.3</b> Sub-option 2: create tables</a></li>
<li class="chapter" data-level="5.2.4" data-path="all-menu-options.html"><a href="all-menu-options.html#code-checks-1"><i class="fa fa-check"></i><b>5.2.4</b> Code checks</a></li>
</ul></li>
<li class="chapter" data-level="5.3" data-path="all-menu-options.html"><a href="all-menu-options.html#option-3-population-differentiation"><i class="fa fa-check"></i><b>5.3</b> Option 3: population differentiation</a><ul>
<li class="chapter" data-level="5.3.1" data-path="all-menu-options.html"><a href="all-menu-options.html#sub-options-1-or-2-genic-differentiation"><i class="fa fa-check"></i><b>5.3.1</b> Sub-options 1 or 2 (genic differentiation)</a></li>
<li class="chapter" data-level="5.3.2" data-path="all-menu-options.html"><a href="all-menu-options.html#sub-options-3-or-4-genotypic-differentiation"><i class="fa fa-check"></i><b>5.3.2</b> Sub-options 3 or 4 (genotypic differentiation)</a></li>
<li class="chapter" data-level="5.3.3" data-path="all-menu-options.html"><a href="all-menu-options.html#output-2"><i class="fa fa-check"></i><b>5.3.3</b> Output</a></li>
<li class="chapter" data-level="5.3.4" data-path="all-menu-options.html"><a href="all-menu-options.html#gene-diversity-as-a-test-statistic"><i class="fa fa-check"></i><b>5.3.4</b> Gene diversity as a test statistic</a></li>
<li class="chapter" data-level="5.3.5" data-path="all-menu-options.html"><a href="all-menu-options.html#analyzing-a-single-contingency-table"><i class="fa fa-check"></i><b>5.3.5</b> Analyzing a single contingency table</a></li>
<li class="chapter" data-level="5.3.6" data-path="all-menu-options.html"><a href="all-menu-options.html#code-checks-2"><i class="fa fa-check"></i><b>5.3.6</b> Code checks</a></li>
</ul></li>
<li class="chapter" data-level="5.4" data-path="all-menu-options.html"><a href="all-menu-options.html#option-4-private-alleles"><i class="fa fa-check"></i><b>5.4</b> Option 4: private alleles</a></li>
<li class="chapter" data-level="5.5" data-path="all-menu-options.html"><a href="all-menu-options.html#option-5-basic-information-f_mathrmis-and-gene-diversities"><i class="fa fa-check"></i><b>5.5</b> Option 5: Basic information, <span class="math inline">\(F_\mathrm{IS}\)</span>, and gene diversities</a><ul>
<li class="chapter" data-level="5.5.1" data-path="all-menu-options.html"><a href="all-menu-options.html#sub-option-1-allele-and-genotype-frequencies"><i class="fa fa-check"></i><b>5.5.1</b> Sub-option 1: Allele and genotype frequencies</a></li>
<li class="chapter" data-level="5.5.2" data-path="all-menu-options.html"><a href="all-menu-options.html#sub-option-2-identity-based-gene-diversities-and-f_mathrmis"><i class="fa fa-check"></i><b>5.5.2</b> Sub-option 2: Identity-based gene diversities and <span class="math inline">\(F_\mathrm{IS}\)</span></a></li>
<li class="chapter" data-level="5.5.3" data-path="all-menu-options.html"><a href="all-menu-options.html#sub-option-3-allele-size-based-gene-diversities-and-rho_mathrmis"><i class="fa fa-check"></i><b>5.5.3</b> Sub-option 3: Allele size-based gene diversities and <span class="math inline">\(\rho_{\mathrm{IS}}\)</span></a></li>
</ul></li>
<li class="chapter" data-level="5.6" data-path="all-menu-options.html"><a href="all-menu-options.html#option-6-fst-and-other-correlations-isolation-by-distance"><i class="fa fa-check"></i><b>5.6</b> Option 6: Fst and other correlations, isolation by distance</a><ul>
<li class="chapter" data-level="5.6.1" data-path="all-menu-options.html"><a href="all-menu-options.html#sub-options-14-f-statistics-and-rho-statistics"><i class="fa fa-check"></i><b>5.6.1</b> Sub-options 1–4: <span class="math inline">\(F\)</span>-statistics and <span class="math inline">\(\rho\)</span>-statistics</a></li>
<li class="chapter" data-level="5.6.2" data-path="all-menu-options.html"><a href="all-menu-options.html#sub-option-5-isolation-by-distance-between-individuals"><i class="fa fa-check"></i><b>5.6.2</b> Sub-option 5: isolation by distance between individuals</a></li>
<li class="chapter" data-level="5.6.3" data-path="all-menu-options.html"><a href="all-menu-options.html#sub-option-6-isolation-by-distance-between-groups"><i class="fa fa-check"></i><b>5.6.3</b> Sub-option 6: isolation by distance between groups</a></li>
<li class="chapter" data-level="5.6.4" data-path="all-menu-options.html"><a href="all-menu-options.html#former-sub-option-5-of-genepop-analysis-of-isolation-by-distance-from-a-genetic-distance-matrix"><i class="fa fa-check"></i><b>5.6.4</b> Former sub-option 5 of Genepop: analysis of isolation by distance from a genetic distance matrix</a></li>
<li class="chapter" data-level="5.6.5" data-path="all-menu-options.html"><a href="all-menu-options.html#user-provided-geographic-distance-matrices"><i class="fa fa-check"></i><b>5.6.5</b> User-provided geographic distance matrices</a></li>
<li class="chapter" data-level="5.6.6" data-path="all-menu-options.html"><a href="all-menu-options.html#analysis-of-isolation-by-distance-from-multiple-genetic-distance-matrices"><i class="fa fa-check"></i><b>5.6.6</b> Analysis of isolation by distance from multiple genetic distance matrices</a></li>
<li class="chapter" data-level="5.6.7" data-path="all-menu-options.html"><a href="all-menu-options.html#analysis-of-mean-differentiation"><i class="fa fa-check"></i><b>5.6.7</b> Analysis of mean differentiation</a></li>
</ul></li>
<li class="chapter" data-level="5.7" data-path="all-menu-options.html"><a href="all-menu-options.html#data-selection-for-analyses-of-isolation-by-distance"><i class="fa fa-check"></i><b>5.7</b> Data selection for analyses of isolation by distance</a><ul>
<li class="chapter" data-level="5.7.1" data-path="all-menu-options.html"><a href="all-menu-options.html#selecting-a-subset-of-samples"><i class="fa fa-check"></i><b>5.7.1</b> Selecting a subset of samples</a></li>
</ul></li>
<li class="chapter" data-level="5.8" data-path="all-menu-options.html"><a href="all-menu-options.html#option-7-file-conversions"><i class="fa fa-check"></i><b>5.8</b> Option 7: File conversions</a></li>
<li class="chapter" data-level="5.9" data-path="all-menu-options.html"><a href="all-menu-options.html#option-8-null-alleles-and-some-input-file-utilities"><i class="fa fa-check"></i><b>5.9</b> Option 8: Null alleles and some input file utilities</a><ul>
<li class="chapter" data-level="5.9.1" data-path="all-menu-options.html"><a href="all-menu-options.html#sub-option-1-null-alleles"><i class="fa fa-check"></i><b>5.9.1</b> Sub-option 1: null alleles</a></li>
<li class="chapter" data-level="5.9.2" data-path="all-menu-options.html"><a href="all-menu-options.html#sub-option-2-diploidisation-of-haploid-data"><i class="fa fa-check"></i><b>5.9.2</b> Sub-option 2: Diploidisation of haploid data</a></li>
<li class="chapter" data-level="5.9.3" data-path="all-menu-options.html"><a href="all-menu-options.html#sub-option-3-relabeling-alleles-names"><i class="fa fa-check"></i><b>5.9.3</b> Sub-option 3: Relabeling alleles names</a></li>
<li class="chapter" data-level="5.9.4" data-path="all-menu-options.html"><a href="all-menu-options.html#sub-options-4-and-5-conversion-of-population-data-to-individual-data"><i class="fa fa-check"></i><b>5.9.4</b> Sub-options 4 and 5: Conversion of population data to individual data</a></li>
<li class="chapter" data-level="5.9.5" data-path="all-menu-options.html"><a href="all-menu-options.html#sub-option-6-random-sampling-of-haploid-genotypes-from-diploid-ones"><i class="fa fa-check"></i><b>5.9.5</b> Sub-option 6: Random sampling of haploid genotypes from diploid ones</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="evaluating-the-performance-of-inferences-for-isolation-by-distance.html"><a href="evaluating-the-performance-of-inferences-for-isolation-by-distance.html"><i class="fa fa-check"></i><b>6</b> Evaluating the performance of inferences for Isolation by distance</a></li>
<li class="chapter" data-level="7" data-path="methods.html"><a href="methods.html"><i class="fa fa-check"></i><b>7</b> Methods</a><ul>
<li class="chapter" data-level="7.1" data-path="methods.html"><a href="methods.html#null-alleles"><i class="fa fa-check"></i><b>7.1</b> Null alleles</a></li>
<li class="chapter" data-level="7.2" data-path="methods.html"><a href="methods.html#exact-tests"><i class="fa fa-check"></i><b>7.2</b> Exact tests</a></li>
<li class="chapter" data-level="7.3" data-path="methods.html"><a href="methods.html#algorithms-for-exact-tests"><i class="fa fa-check"></i><b>7.3</b> Algorithms for exact tests</a></li>
<li class="chapter" data-level="7.4" data-path="methods.html"><a href="methods.html#accuracy-of-p-values-estimated-by-the-markov-chain-algorithms"><i class="fa fa-check"></i><b>7.4</b> Accuracy of P values estimated by the Markov chain algorithms</a></li>
<li class="chapter" data-level="7.5" data-path="methods.html"><a href="methods.html#test-statistics"><i class="fa fa-check"></i><b>7.5</b> Test statistics</a></li>
<li class="chapter" data-level="7.6" data-path="methods.html"><a href="methods.html#estimating-f-statistics-and-related-quantities"><i class="fa fa-check"></i><b>7.6</b> Estimating <span class="math inline">\(F\)</span>-statistics and related quantities</a><ul>
<li class="chapter" data-level="7.6.1" data-path="methods.html"><a href="methods.html#Fmulti"><i class="fa fa-check"></i><b>7.6.1</b> ANOVA estimators: single- and multilocus definitions</a></li>
<li class="chapter" data-level="7.6.2" data-path="methods.html"><a href="methods.html#rho-stats"><i class="fa fa-check"></i><b>7.6.2</b> Microsatellite allele sizes, <span class="math inline">\(R_\mathrm{ST}\)</span>, and <span class="math inline">\(\rho_\mathrm{ST}\)</span></a></li>
<li class="chapter" data-level="7.6.3" data-path="methods.html"><a href="methods.html#robertson-and-hills-estimator-of-f_mathrmis"><i class="fa fa-check"></i><b>7.6.3</b> Robertson and Hill’s estimator of <span class="math inline">\(F_\mathrm{IS}\)</span></a></li>
</ul></li>
<li class="chapter" data-level="7.7" data-path="methods.html"><a href="methods.html#bootstraps"><i class="fa fa-check"></i><b>7.7</b> Bootstraps</a></li>
<li class="chapter" data-level="7.8" data-path="methods.html"><a href="methods.html#mantel-test"><i class="fa fa-check"></i><b>7.8</b> Mantel test</a><ul>
<li class="chapter" data-level="7.8.1" data-path="methods.html"><a href="methods.html#misuse-1-tests-of-correlation-at-different-distance"><i class="fa fa-check"></i><b>7.8.1</b> Misuse 1: tests of correlation at different distance</a></li>
<li class="chapter" data-level="7.8.2" data-path="methods.html"><a href="methods.html#misuse-2-partial-mantel-tests"><i class="fa fa-check"></i><b>7.8.2</b> Misuse 2: partial Mantel tests</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="8" data-path="code-maintenance-credits-contact-etc-.html"><a href="code-maintenance-credits-contact-etc-.html"><i class="fa fa-check"></i><b>8</b> Code maintenance, credits, contact, etc.</a><ul>
<li class="chapter" data-level="8.1" data-path="code-maintenance-credits-contact-etc-.html"><a href="code-maintenance-credits-contact-etc-.html#code-maintenance"><i class="fa fa-check"></i><b>8.1</b> Code maintenance</a></li>
<li class="chapter" data-level="8.2" data-path="code-maintenance-credits-contact-etc-.html"><a href="code-maintenance-credits-contact-etc-.html#credits-for-the-current-version"><i class="fa fa-check"></i><b>8.2</b> Credits for the current version</a></li>
<li class="chapter" data-level="8.3" data-path="code-maintenance-credits-contact-etc-.html"><a href="code-maintenance-credits-contact-etc-.html#previous-history"><i class="fa fa-check"></i><b>8.3</b> Previous history</a></li>
<li class="chapter" data-level="8.4" data-path="code-maintenance-credits-contact-etc-.html"><a href="code-maintenance-credits-contact-etc-.html#contact"><i class="fa fa-check"></i><b>8.4</b> Contact</a><ul>
<li class="chapter" data-level="8.4.1" data-path="code-maintenance-credits-contact-etc-.html"><a href="code-maintenance-credits-contact-etc-.html#bug-fixes-since-release-of-genepop-version-3.4-in-may-2003-until-first-release-of-genepop-4.0"><i class="fa fa-check"></i><b>8.4.1</b> Bug fixes since release of Genepop version 3.4 in May 2003 until first release of Genepop 4.0:</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="9" data-path="copyright.html"><a href="copyright.html"><i class="fa fa-check"></i><b>9</b> Copyright</a></li>
<li class="chapter" data-level="" data-path="bibliography.html"><a href="bibliography.html"><i class="fa fa-check"></i>Bibliography</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Genepop version 4.7.5</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="installing-genepop-and-session-examples" class="section level1">
<h1><span class="header-section-number">2</span> Installing Genepop and session examples</h1>
<div id="installation" class="section level2">
<h2><span class="header-section-number">2.1</span> Installation</h2>
<div id="r-package" class="section level3">
<h3><span class="header-section-number">2.1.1</span> R package</h3>
<p>As any R package, it can be installed by <code>install.packages(&quot;genepop&quot;)</code> if on CRAN, and more generally by <code>install.packages(</code><archive file name><code>,type=&quot;source&quot;,repos=NULL)</code>. See the R documentation for more information.</p>
</div>
<div id="stand-alone-executable" class="section level3">
<h3><span class="header-section-number">2.1.2</span> Stand-alone executable</h3>
<p>Under <strong>Microsoft Windows</strong>, one only needs to unzip/copy the executable on hard disk. Both 32- and 64-bit versions of the executables are distributed. Under <strong>Linux/Mac OSX</strong>,  extract all c++ sources from the distributed <code>sources.tar.gz</code> (or from the <code>src/</code> subdirectory of the R package sources, <strong>except</strong> <code>RcppExports.cpp</code>), and compile with a compiler that supports the C++11 standards. For <strong>Windows</strong>, one can use <code>g++</code> version 4.9.3 (distributed with recent versions of the R tools) with an ad hoc flag:</p>
<p><code>g++ -std=c++11 -o Genepop *.cpp -O3</code></p>
<p>(<code>O</code> in <code>-O3</code> is the letter O, not zero). With more recent versions of <code>g++</code> (&gt;=6.0) or <code>clang++</code>, no such flag is required:</p>
<p><code>g++ -o Genepop *.cpp -O3</code>.</p>
<p>The data files do not need to be in the same directory as the executable<a href="#fn1" class="footnoteRef" id="fnref1"><sup>1</sup></a>; however, users might find that specifying path names under Windows is not as easy at it should.</p>
<p>Examples and documentation files are included in the R package and are available on the <a href="http://kimura.univ-montp2.fr/~rousset/Genepop.htm">Genepop distribution page</a>.</p>
<p>Linkdos, a program described by <span class="citation">Garnier-Géré and Dillmann (1992)</span>, is distributed with (but is not part of) Genepop. It is originally a DOS program, but the source file distributed can be recompiled under Linux using the Free Pascal compiler (or at least ``could’’, since this is no longer maintained/checked).</p>
</div>
</div>
<div id="example-sessions" class="section level2">
<h2><span class="header-section-number">2.2</span> Example sessions</h2>
<p>To reproduce the examples of this session one should copy in a personal directory the examples files found in the <code>extdata/</code> subdirectory of the packageor on the <a href="http://kimura.univ-montp2.fr/~rousset/Genepop.htm">Genepop distribution page</a>.</p>
<div id="example-1-basic-session" class="section level3">
<h3><span class="header-section-number">2.2.1</span> Example 1: basic session</h3>
<p>Open a console window in the directory where Genepop has been installed and just execute</p>
<pre><code> Genepop</code></pre>
<p>If Genepop has never been run before, it will ask for an input file. Otherwise, the main menu should appear, in which case you should use the <code>C</code> option to load this input file. For this sample session, the file name to be given is <code>sample.txt</code>. Genepop will display some information about the file read, then display the main menu:</p>
<pre><code>-------&gt; Change Data ................... C


Testing :
    Hardy-Weinberg exact tests (several options) ...................... 1
    Exact tests for genotypic disequilibrium (several options) ........ 2
    Exact tests for population differentiation (several options) ...... 3

Estimating:
    Nm estimates (private allele method) .............................. 4
    Allele frequencies, various Fis and gene diversities .............. 5
    Fst &amp; other correlations, isolation by distance (several options).. 6

Ecumenicism and various utilities:
    Ecumenicism: file conversion (several options) .................... 7
    Null alleles and miscellaneous input file utilities ............... 8

QUIT Genepop .......................................................... 9

Your choice? :</code></pre>
<p>Each option will be described later. Let us see some tests for heterozygote deficiency. Reply <code>1</code>, next <code>1</code>, next <code>y</code>(es). As indicated, the results of the analysis are stored in the file <code>sample.txt.D</code>.</p>
<p>The next example illustrates a slightly more elaborate use of Genepop.</p>
</div>
<div id="example-2-using-the-settings-file" class="section level3">
<h3><span class="header-section-number">2.2.2</span> Example 2: using the settings file</h3>
<p>Execute</p>
<pre><code> Genepop settingsFile=SampleSettings.txt</code></pre>
<p>Do not add spaces in the arguments. Capitalisation matters for file names (here <code>SampleSettings.txt</code>) if it matters for the operating system (i.e. for Linux).</p>
<p>You can see that the previous and additional analyses are performed, and that you just need to hit Return each time Genepop stops and waits for feedback. Finally, you are brought back to the main menu. Simple instructions for performing the analyses are contained in the <code>SampleSettings.txt</code> file, which you may edit. Section <a href="sec-settings.html#sec-settings">4</a> will explain how to use this file. By default, Genepop seeks and eventually reads instructions in a <code>Genepop.txt</code> file. You can see that one such file is present and was thus read when performing Example 1.</p>
</div>
<div id="example-3-batch-processing" class="section level3">
<h3><span class="header-section-number">2.2.3</span> Example 3: Batch processing</h3>
<p>Execute the same command as in the previous example but with one more statement:</p>
<pre><code> Genepop settingsFile=SampleSettings.txt Mode=Batch</code></pre>
<p> Genepop should perform the same computations as in the previous example but it will not stop and wait for feedback, and will exit after completion of the computations. Note again that spaces are not allowed within each of the arguments <code>settingsFile=SampleSettings.txt</code> and <code>Mode=Batch</code>, nor more generally in arguments specified on the command line.</p>
<p>The batch mode makes it easy to analyze multiple files. However, note that concurrent Genepop processes should be run in distinct directories. Otherwise, the temporary files of each process might conflict with each other.</p>
</div>
</div>
</div>
<div class="footnotes">
<hr />
<ol start="1">
<li id="fn1"><p>…in contrast to earlier versions of Genepop.<a href="installing-genepop-and-session-examples.html#fnref1">↩</a></p></li>
</ol>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="introduction.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="the-input-file.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
},
"search": false
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:" && /^https?:/.test(src))
      src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
