<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>7 Methods | Genepop version 4.7.5</title>
  <meta name="description" content="7 Methods | Genepop version 4.7.5" />
  <meta name="generator" content="bookdown 0.10 and GitBook 2.6.7" />

  <meta property="og:title" content="7 Methods | Genepop version 4.7.5" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="7 Methods | Genepop version 4.7.5" />
  
  
  

<meta name="author" content="F. Rousset" />



  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="evaluating-the-performance-of-inferences-for-isolation-by-distance.html">
<link rel="next" href="code-maintenance-credits-contact-etc-.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />










</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="1" data-path="introduction.html"><a href="introduction.html"><i class="fa fa-check"></i><b>1</b> Introduction</a><ul>
<li class="chapter" data-level="1.1" data-path="introduction.html"><a href="introduction.html#purpose"><i class="fa fa-check"></i><b>1.1</b> Purpose</a></li>
<li class="chapter" data-level="1.2" data-path="introduction.html"><a href="introduction.html#the-two-genepop-distributions"><i class="fa fa-check"></i><b>1.2</b> The two Genepop distributions</a></li>
<li class="chapter" data-level="1.3" data-path="introduction.html"><a href="introduction.html#changes-since-version-4.0"><i class="fa fa-check"></i><b>1.3</b> Changes since version 4.0</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="installing-genepop-and-session-examples.html"><a href="installing-genepop-and-session-examples.html"><i class="fa fa-check"></i><b>2</b> Installing Genepop and session examples</a><ul>
<li class="chapter" data-level="2.1" data-path="installing-genepop-and-session-examples.html"><a href="installing-genepop-and-session-examples.html#installation"><i class="fa fa-check"></i><b>2.1</b> Installation</a><ul>
<li class="chapter" data-level="2.1.1" data-path="installing-genepop-and-session-examples.html"><a href="installing-genepop-and-session-examples.html#r-package"><i class="fa fa-check"></i><b>2.1.1</b> R package</a></li>
<li class="chapter" data-level="2.1.2" data-path="installing-genepop-and-session-examples.html"><a href="installing-genepop-and-session-examples.html#stand-alone-executable"><i class="fa fa-check"></i><b>2.1.2</b> Stand-alone executable</a></li>
</ul></li>
<li class="chapter" data-level="2.2" data-path="installing-genepop-and-session-examples.html"><a href="installing-genepop-and-session-examples.html#example-sessions"><i class="fa fa-check"></i><b>2.2</b> Example sessions</a><ul>
<li class="chapter" data-level="2.2.1" data-path="installing-genepop-and-session-examples.html"><a href="installing-genepop-and-session-examples.html#example-1-basic-session"><i class="fa fa-check"></i><b>2.2.1</b> Example 1: basic session</a></li>
<li class="chapter" data-level="2.2.2" data-path="installing-genepop-and-session-examples.html"><a href="installing-genepop-and-session-examples.html#example-2-using-the-settings-file"><i class="fa fa-check"></i><b>2.2.2</b> Example 2: using the settings file</a></li>
<li class="chapter" data-level="2.2.3" data-path="installing-genepop-and-session-examples.html"><a href="installing-genepop-and-session-examples.html#example-3-batch-processing"><i class="fa fa-check"></i><b>2.2.3</b> Example 3: Batch processing</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="the-input-file.html"><a href="the-input-file.html"><i class="fa fa-check"></i><b>3</b> The input file</a></li>
<li class="chapter" data-level="4" data-path="sec-settings.html"><a href="sec-settings.html"><i class="fa fa-check"></i><b>4</b> The settings file and command line arguments</a></li>
<li class="chapter" data-level="5" data-path="all-menu-options.html"><a href="all-menu-options.html"><i class="fa fa-check"></i><b>5</b> All menu options</a><ul>
<li class="chapter" data-level="5.1" data-path="all-menu-options.html"><a href="all-menu-options.html#option-1-hardy-weinberg-hw-exact-tests"><i class="fa fa-check"></i><b>5.1</b> Option 1: Hardy-Weinberg (HW) exact tests</a><ul>
<li class="chapter" data-level="5.1.1" data-path="all-menu-options.html"><a href="all-menu-options.html#sub-options-13-tests-for-each-locus-in-each-population"><i class="fa fa-check"></i><b>5.1.1</b> Sub-options 1–3: Tests for each locus in each population</a></li>
<li class="chapter" data-level="5.1.2" data-path="all-menu-options.html"><a href="all-menu-options.html#output"><i class="fa fa-check"></i><b>5.1.2</b> Output</a></li>
<li class="chapter" data-level="5.1.3" data-path="all-menu-options.html"><a href="all-menu-options.html#sub-options-45-global-tests-across-loci-or-across-samples"><i class="fa fa-check"></i><b>5.1.3</b> Sub-options 4,5: Global tests across loci or across samples</a></li>
<li class="chapter" data-level="5.1.4" data-path="all-menu-options.html"><a href="all-menu-options.html#analyzing-a-single-genotypic-matrix"><i class="fa fa-check"></i><b>5.1.4</b> Analyzing a single genotypic matrix</a></li>
<li class="chapter" data-level="5.1.5" data-path="all-menu-options.html"><a href="all-menu-options.html#code-checks"><i class="fa fa-check"></i><b>5.1.5</b> Code checks</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="all-menu-options.html"><a href="all-menu-options.html#option-2-tests-and-tables-for-linkage-disequilibrium"><i class="fa fa-check"></i><b>5.2</b> Option 2: Tests and tables for linkage disequilibrium</a><ul>
<li class="chapter" data-level="5.2.1" data-path="all-menu-options.html"><a href="all-menu-options.html#sub-option-1-tests"><i class="fa fa-check"></i><b>5.2.1</b> Sub-option 1: Tests</a></li>
<li class="chapter" data-level="5.2.2" data-path="all-menu-options.html"><a href="all-menu-options.html#output-1"><i class="fa fa-check"></i><b>5.2.2</b> Output</a></li>
<li class="chapter" data-level="5.2.3" data-path="all-menu-options.html"><a href="all-menu-options.html#sub-option-2-create-tables"><i class="fa fa-check"></i><b>5.2.3</b> Sub-option 2: create tables</a></li>
<li class="chapter" data-level="5.2.4" data-path="all-menu-options.html"><a href="all-menu-options.html#code-checks-1"><i class="fa fa-check"></i><b>5.2.4</b> Code checks</a></li>
</ul></li>
<li class="chapter" data-level="5.3" data-path="all-menu-options.html"><a href="all-menu-options.html#option-3-population-differentiation"><i class="fa fa-check"></i><b>5.3</b> Option 3: population differentiation</a><ul>
<li class="chapter" data-level="5.3.1" data-path="all-menu-options.html"><a href="all-menu-options.html#sub-options-1-or-2-genic-differentiation"><i class="fa fa-check"></i><b>5.3.1</b> Sub-options 1 or 2 (genic differentiation)</a></li>
<li class="chapter" data-level="5.3.2" data-path="all-menu-options.html"><a href="all-menu-options.html#sub-options-3-or-4-genotypic-differentiation"><i class="fa fa-check"></i><b>5.3.2</b> Sub-options 3 or 4 (genotypic differentiation)</a></li>
<li class="chapter" data-level="5.3.3" data-path="all-menu-options.html"><a href="all-menu-options.html#output-2"><i class="fa fa-check"></i><b>5.3.3</b> Output</a></li>
<li class="chapter" data-level="5.3.4" data-path="all-menu-options.html"><a href="all-menu-options.html#gene-diversity-as-a-test-statistic"><i class="fa fa-check"></i><b>5.3.4</b> Gene diversity as a test statistic</a></li>
<li class="chapter" data-level="5.3.5" data-path="all-menu-options.html"><a href="all-menu-options.html#analyzing-a-single-contingency-table"><i class="fa fa-check"></i><b>5.3.5</b> Analyzing a single contingency table</a></li>
<li class="chapter" data-level="5.3.6" data-path="all-menu-options.html"><a href="all-menu-options.html#code-checks-2"><i class="fa fa-check"></i><b>5.3.6</b> Code checks</a></li>
</ul></li>
<li class="chapter" data-level="5.4" data-path="all-menu-options.html"><a href="all-menu-options.html#option-4-private-alleles"><i class="fa fa-check"></i><b>5.4</b> Option 4: private alleles</a></li>
<li class="chapter" data-level="5.5" data-path="all-menu-options.html"><a href="all-menu-options.html#option-5-basic-information-f_mathrmis-and-gene-diversities"><i class="fa fa-check"></i><b>5.5</b> Option 5: Basic information, <span class="math inline">\(F_\mathrm{IS}\)</span>, and gene diversities</a><ul>
<li class="chapter" data-level="5.5.1" data-path="all-menu-options.html"><a href="all-menu-options.html#sub-option-1-allele-and-genotype-frequencies"><i class="fa fa-check"></i><b>5.5.1</b> Sub-option 1: Allele and genotype frequencies</a></li>
<li class="chapter" data-level="5.5.2" data-path="all-menu-options.html"><a href="all-menu-options.html#sub-option-2-identity-based-gene-diversities-and-f_mathrmis"><i class="fa fa-check"></i><b>5.5.2</b> Sub-option 2: Identity-based gene diversities and <span class="math inline">\(F_\mathrm{IS}\)</span></a></li>
<li class="chapter" data-level="5.5.3" data-path="all-menu-options.html"><a href="all-menu-options.html#sub-option-3-allele-size-based-gene-diversities-and-rho_mathrmis"><i class="fa fa-check"></i><b>5.5.3</b> Sub-option 3: Allele size-based gene diversities and <span class="math inline">\(\rho_{\mathrm{IS}}\)</span></a></li>
</ul></li>
<li class="chapter" data-level="5.6" data-path="all-menu-options.html"><a href="all-menu-options.html#option-6-fst-and-other-correlations-isolation-by-distance"><i class="fa fa-check"></i><b>5.6</b> Option 6: Fst and other correlations, isolation by distance</a><ul>
<li class="chapter" data-level="5.6.1" data-path="all-menu-options.html"><a href="all-menu-options.html#sub-options-14-f-statistics-and-rho-statistics"><i class="fa fa-check"></i><b>5.6.1</b> Sub-options 1–4: <span class="math inline">\(F\)</span>-statistics and <span class="math inline">\(\rho\)</span>-statistics</a></li>
<li class="chapter" data-level="5.6.2" data-path="all-menu-options.html"><a href="all-menu-options.html#sub-option-5-isolation-by-distance-between-individuals"><i class="fa fa-check"></i><b>5.6.2</b> Sub-option 5: isolation by distance between individuals</a></li>
<li class="chapter" data-level="5.6.3" data-path="all-menu-options.html"><a href="all-menu-options.html#sub-option-6-isolation-by-distance-between-groups"><i class="fa fa-check"></i><b>5.6.3</b> Sub-option 6: isolation by distance between groups</a></li>
<li class="chapter" data-level="5.6.4" data-path="all-menu-options.html"><a href="all-menu-options.html#former-sub-option-5-of-genepop-analysis-of-isolation-by-distance-from-a-genetic-distance-matrix"><i class="fa fa-check"></i><b>5.6.4</b> Former sub-option 5 of Genepop: analysis of isolation by distance from a genetic distance matrix</a></li>
<li class="chapter" data-level="5.6.5" data-path="all-menu-options.html"><a href="all-menu-options.html#user-provided-geographic-distance-matrices"><i class="fa fa-check"></i><b>5.6.5</b> User-provided geographic distance matrices</a></li>
<li class="chapter" data-level="5.6.6" data-path="all-menu-options.html"><a href="all-menu-options.html#analysis-of-isolation-by-distance-from-multiple-genetic-distance-matrices"><i class="fa fa-check"></i><b>5.6.6</b> Analysis of isolation by distance from multiple genetic distance matrices</a></li>
<li class="chapter" data-level="5.6.7" data-path="all-menu-options.html"><a href="all-menu-options.html#analysis-of-mean-differentiation"><i class="fa fa-check"></i><b>5.6.7</b> Analysis of mean differentiation</a></li>
</ul></li>
<li class="chapter" data-level="5.7" data-path="all-menu-options.html"><a href="all-menu-options.html#data-selection-for-analyses-of-isolation-by-distance"><i class="fa fa-check"></i><b>5.7</b> Data selection for analyses of isolation by distance</a><ul>
<li class="chapter" data-level="5.7.1" data-path="all-menu-options.html"><a href="all-menu-options.html#selecting-a-subset-of-samples"><i class="fa fa-check"></i><b>5.7.1</b> Selecting a subset of samples</a></li>
</ul></li>
<li class="chapter" data-level="5.8" data-path="all-menu-options.html"><a href="all-menu-options.html#option-7-file-conversions"><i class="fa fa-check"></i><b>5.8</b> Option 7: File conversions</a></li>
<li class="chapter" data-level="5.9" data-path="all-menu-options.html"><a href="all-menu-options.html#option-8-null-alleles-and-some-input-file-utilities"><i class="fa fa-check"></i><b>5.9</b> Option 8: Null alleles and some input file utilities</a><ul>
<li class="chapter" data-level="5.9.1" data-path="all-menu-options.html"><a href="all-menu-options.html#sub-option-1-null-alleles"><i class="fa fa-check"></i><b>5.9.1</b> Sub-option 1: null alleles</a></li>
<li class="chapter" data-level="5.9.2" data-path="all-menu-options.html"><a href="all-menu-options.html#sub-option-2-diploidisation-of-haploid-data"><i class="fa fa-check"></i><b>5.9.2</b> Sub-option 2: Diploidisation of haploid data</a></li>
<li class="chapter" data-level="5.9.3" data-path="all-menu-options.html"><a href="all-menu-options.html#sub-option-3-relabeling-alleles-names"><i class="fa fa-check"></i><b>5.9.3</b> Sub-option 3: Relabeling alleles names</a></li>
<li class="chapter" data-level="5.9.4" data-path="all-menu-options.html"><a href="all-menu-options.html#sub-options-4-and-5-conversion-of-population-data-to-individual-data"><i class="fa fa-check"></i><b>5.9.4</b> Sub-options 4 and 5: Conversion of population data to individual data</a></li>
<li class="chapter" data-level="5.9.5" data-path="all-menu-options.html"><a href="all-menu-options.html#sub-option-6-random-sampling-of-haploid-genotypes-from-diploid-ones"><i class="fa fa-check"></i><b>5.9.5</b> Sub-option 6: Random sampling of haploid genotypes from diploid ones</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="evaluating-the-performance-of-inferences-for-isolation-by-distance.html"><a href="evaluating-the-performance-of-inferences-for-isolation-by-distance.html"><i class="fa fa-check"></i><b>6</b> Evaluating the performance of inferences for Isolation by distance</a></li>
<li class="chapter" data-level="7" data-path="methods.html"><a href="methods.html"><i class="fa fa-check"></i><b>7</b> Methods</a><ul>
<li class="chapter" data-level="7.1" data-path="methods.html"><a href="methods.html#null-alleles"><i class="fa fa-check"></i><b>7.1</b> Null alleles</a></li>
<li class="chapter" data-level="7.2" data-path="methods.html"><a href="methods.html#exact-tests"><i class="fa fa-check"></i><b>7.2</b> Exact tests</a></li>
<li class="chapter" data-level="7.3" data-path="methods.html"><a href="methods.html#algorithms-for-exact-tests"><i class="fa fa-check"></i><b>7.3</b> Algorithms for exact tests</a></li>
<li class="chapter" data-level="7.4" data-path="methods.html"><a href="methods.html#accuracy-of-p-values-estimated-by-the-markov-chain-algorithms"><i class="fa fa-check"></i><b>7.4</b> Accuracy of P values estimated by the Markov chain algorithms</a></li>
<li class="chapter" data-level="7.5" data-path="methods.html"><a href="methods.html#test-statistics"><i class="fa fa-check"></i><b>7.5</b> Test statistics</a></li>
<li class="chapter" data-level="7.6" data-path="methods.html"><a href="methods.html#estimating-f-statistics-and-related-quantities"><i class="fa fa-check"></i><b>7.6</b> Estimating <span class="math inline">\(F\)</span>-statistics and related quantities</a><ul>
<li class="chapter" data-level="7.6.1" data-path="methods.html"><a href="methods.html#Fmulti"><i class="fa fa-check"></i><b>7.6.1</b> ANOVA estimators: single- and multilocus definitions</a></li>
<li class="chapter" data-level="7.6.2" data-path="methods.html"><a href="methods.html#rho-stats"><i class="fa fa-check"></i><b>7.6.2</b> Microsatellite allele sizes, <span class="math inline">\(R_\mathrm{ST}\)</span>, and <span class="math inline">\(\rho_\mathrm{ST}\)</span></a></li>
<li class="chapter" data-level="7.6.3" data-path="methods.html"><a href="methods.html#robertson-and-hills-estimator-of-f_mathrmis"><i class="fa fa-check"></i><b>7.6.3</b> Robertson and Hill’s estimator of <span class="math inline">\(F_\mathrm{IS}\)</span></a></li>
</ul></li>
<li class="chapter" data-level="7.7" data-path="methods.html"><a href="methods.html#bootstraps"><i class="fa fa-check"></i><b>7.7</b> Bootstraps</a></li>
<li class="chapter" data-level="7.8" data-path="methods.html"><a href="methods.html#mantel-test"><i class="fa fa-check"></i><b>7.8</b> Mantel test</a><ul>
<li class="chapter" data-level="7.8.1" data-path="methods.html"><a href="methods.html#misuse-1-tests-of-correlation-at-different-distance"><i class="fa fa-check"></i><b>7.8.1</b> Misuse 1: tests of correlation at different distance</a></li>
<li class="chapter" data-level="7.8.2" data-path="methods.html"><a href="methods.html#misuse-2-partial-mantel-tests"><i class="fa fa-check"></i><b>7.8.2</b> Misuse 2: partial Mantel tests</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="8" data-path="code-maintenance-credits-contact-etc-.html"><a href="code-maintenance-credits-contact-etc-.html"><i class="fa fa-check"></i><b>8</b> Code maintenance, credits, contact, etc.</a><ul>
<li class="chapter" data-level="8.1" data-path="code-maintenance-credits-contact-etc-.html"><a href="code-maintenance-credits-contact-etc-.html#code-maintenance"><i class="fa fa-check"></i><b>8.1</b> Code maintenance</a></li>
<li class="chapter" data-level="8.2" data-path="code-maintenance-credits-contact-etc-.html"><a href="code-maintenance-credits-contact-etc-.html#credits-for-the-current-version"><i class="fa fa-check"></i><b>8.2</b> Credits for the current version</a></li>
<li class="chapter" data-level="8.3" data-path="code-maintenance-credits-contact-etc-.html"><a href="code-maintenance-credits-contact-etc-.html#previous-history"><i class="fa fa-check"></i><b>8.3</b> Previous history</a></li>
<li class="chapter" data-level="8.4" data-path="code-maintenance-credits-contact-etc-.html"><a href="code-maintenance-credits-contact-etc-.html#contact"><i class="fa fa-check"></i><b>8.4</b> Contact</a><ul>
<li class="chapter" data-level="8.4.1" data-path="code-maintenance-credits-contact-etc-.html"><a href="code-maintenance-credits-contact-etc-.html#bug-fixes-since-release-of-genepop-version-3.4-in-may-2003-until-first-release-of-genepop-4.0"><i class="fa fa-check"></i><b>8.4.1</b> Bug fixes since release of Genepop version 3.4 in May 2003 until first release of Genepop 4.0:</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="9" data-path="copyright.html"><a href="copyright.html"><i class="fa fa-check"></i><b>9</b> Copyright</a></li>
<li class="chapter" data-level="" data-path="bibliography.html"><a href="bibliography.html"><i class="fa fa-check"></i>Bibliography</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Genepop version 4.7.5</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="methods" class="section level1">
<h1><span class="header-section-number">7</span> Methods</h1>
<p>This section is only intended as a quick reference guide. The primary literature should be consulted for further information about the methods implemented in Genepop.</p>
<div id="null-alleles" class="section level2">
<h2><span class="header-section-number">7.1</span> Null alleles</h2>
<p> When apparent null homozygotes are observed, one may wonder whether these are truly null homozygotes, or whether some technical failure independent of genotype has occurred. Maximum likelihood estimates of null allele frequency, or of this frequency jointly with the failure rate, can be obtained by the EM algorithm <span class="citation">(Dempster, Laird, and Rubin 1977; Hartl and Clark 1989; Kalinowski and Taper 2006)</span>, which is one of the methods implemented in Genepop (menu option 8.1).</p>
<p>Also implemented is a simpler estimator defined by <span class="citation">Brookfield (1996)</span> for the case where apparent null homozygotes are true null homozygotes. He also described this as a maximum likelihood estimator, but there are some (often small) differences with the ML estimates derived by the EM algorithm as implemented in this and previous versions of Genepop, which may to be due to the fact that Brookfield wrote a likelihood formula for the number of apparent homozygotes and heterozygotes, while the EM implementation is based on a likelihood formula where apparent homozygotes and heterozygotes for different alleles are distinguished.</p>
<p>For the case where one is unsure whether apparent null homozygotes are true null homozygotes, <span class="citation">Chakraborty et al. (1992)</span> described a method to estimate the null allele frequency from the other data, excluding any apparent null homozygote. The estimator is not implemented in Genepop because, beyond its relatively low efficiency, its behavior is sometimes puzzling (for example, where there is no obvious heterozygote in a sample, the estimated null allele frequency is always 1, whatever the number of alleles obviously present and even if only non-null genotypes are present). Actually, even if apparent null homozygotes are not true null homozygotes, their number bring some information, and it is more logical to estimate the null allele frequency jointly with the nonspecific genotyping failure rate by maximum likelihood <span class="citation">(Kalinowski and Taper 2006)</span>. This analysis is possible when at least three alleles are obviously present.</p>
</div>
<div id="exact-tests" class="section level2">
<h2><span class="header-section-number">7.2</span> Exact tests</h2>
<p>The probability of a sample of genotypes depends on allele frequencies at one or more loci. In the tests of Hardy Weinberg equilibrium, population differentiation and pairwise independence between loci (“linkage equilibrium”) implemented in Genepop, one is not interested in the allele frequencies themselves and, given they are unknown, the aim is to derive valid conclusions whatever their values. In these different cases, this can be achieved by considering only the probability of samples conditional on observed allelic (e.g. for HW tests) or genotypic counts (e.g. for tests of population differentiation not assuming HW equilibrium). Because exact probabilities are computed, these conditional tests are also known as exact tests. See <span class="citation">Cox and Hinkley (1974)</span> and <span class="citation">Lehmann (1994)</span> for the underlying theory; a much more elementary introduction to the tests implemented in Genepop is <span class="citation">Rousset and Raymond (1997)</span>.</p>
</div>
<div id="algorithms-for-exact-tests" class="section level2">
<h2><span class="header-section-number">7.3</span> Algorithms for exact tests</h2>
<p>Conditional tests require in principle the complete enumeration of all possible samples satisfying the given condition. In many cases this is not practical, and the <span class="math inline">\(P\)</span>-value may be computed by simple permutation algorithms or by more elaborate Markov chain algorithms, in particular the Metropolis-Hastings algorithm <span class="citation">(Hastings 1970)</span>. The latter algorithm explores the universe of samples satisfying the given condition in a “random walk” fashion. For HW testing <span class="citation">Guo and Thompson (1992)</span> found a Metropolis-Hastings algorithm to be efficient compared to permutations. A slight modification of their algorithm is implemented in Genepop. Guo and Thompson also considered tests for contingency tables (Technical report No. 187, Department of Statistics, University of Washington, Seattle, USA, 1989) and again a slightly modified algorithm is implemented in Genepop <span class="citation">(Raymond and Rousset 1995a)</span>. A run of the Markov chain (MC) algorithms starts with a dememorization step; if this step is long enough, the state of the chain at the end of the dememorization is independent of the initial state. Then, further simulation of the MC is divided in batches. In each batch a P-value estimate is derived by counting the proportion of time the MC spends visiting sample configurations more extreme (according to the given test statistic) than the observed sample. If the batches are long enough, the P-value estimates from successive batches are essentially independent from each other and a standard error for the P-value can be derived from the variance of per-batch P-values <span class="citation">(Hastings 1970)</span>. As could be expected, the longer the runs, the lower the standard error.</p>
</div>
<div id="accuracy-of-p-values-estimated-by-the-markov-chain-algorithms" class="section level2">
<h2><span class="header-section-number">7.4</span> Accuracy of P values estimated by the Markov chain algorithms</h2>
<p> For most data sets the MC “mixes well” so that the default values of the dememorization length and batch length implemented in Genepop appear quite sufficient <span class="citation">(in many other applications of MC algorithms, things are not so simple; e.g. Brooks and Gelman 1998)</span>. Nevertheless, inaccurate P-values can be detected when the standard error is large, or else if the number of switches (the number of times the sample configuration changes in the MC run) is low (this may occur when the P-value estimate is close to 0 or 1). Therefore, it is wise to increase the number of batches if the standard error is too large, in particular if it is of the order of <span class="math inline">\(P\)</span> (the P-value) for small <span class="math inline">\(P\)</span> or of the order of <span class="math inline">\(1-P\)</span> for large <span class="math inline">\(P\)</span>, or else if the number of switches is low (<span class="math inline">\(&lt;1000\)</span>).</p>
</div>
<div id="test-statistics" class="section level2">
<h2><span class="header-section-number">7.5</span> Test statistics</h2>
<p>The Markov chain algorithms were first implemented for probability tests, i.e. tests where the rejection zone is defined out of the least likely samples under the null hypothesis. Such tests also had Fisher’s preference <span class="citation">(e.g. Fisher 1935)</span>; in particular the probability test for independence in contingency tables is known as Fisher’s exact test. However, probability tests are not necessarily the most powerful. Depending on the alternative hypothesis of importance, other test statistics are often preferable <span class="citation">(see again Cox and Hinkley 1974 or <span class="citation">Lehmann (1994)</span> for textbook accounts)</span>. Efficient tests for detecting heterozygote excesses and deficits <span class="citation">(Rousset and Raymond 1995)</span> were introduced in Genepop from the start (see option 1), and log likelihood ratio (<span class="math inline">\(G\)</span>) tests were introduced with the implementation of the genotypic tests for population differentiation <span class="citation">(Goudet et al. 1996)</span>. The allelic weighting implicit in the <span class="math inline">\(G\)</span> statistic is indeed optimal for detecting differentiation under an island model <span class="citation">(Rousset 2007)</span> and use of the <span class="math inline">\(G\)</span> statistic has been generalized to all contingency table tests in Genepop 4.0, though probability tests performed in earlier versions of Genepop are still available.</p>
<p>Global tests are performed either using methods tuned to specific alternative hypotheses (for heterozygote excess or deficiency) or using Fisher’s combination of probabilities technique. While the latter has been criticized <span class="citation">(Whitlock 2005)</span>, the recommended alternative can fail spectacularly on discrete data.</p>
</div>
<div id="estimating-f-statistics-and-related-quantities" class="section level2">
<h2><span class="header-section-number">7.6</span> Estimating <span class="math inline">\(F\)</span>-statistics and related quantities</h2>
<p>The definition of <span class="math inline">\(F\)</span>-statistics used here is</p>
<p><span class="math display">\[\begin{aligned}
 {F_\mathrm{IS}}\equiv &amp;\frac{Q_1-Q_2}{1-Q_2}\\
 {F_\mathrm{ST}}\equiv &amp;\frac{Q_2-Q_3}{1-Q_3}\\
 {F_\mathrm{IT}}\equiv &amp;\frac{Q_1-Q_3}{1-Q_3}
 \end{aligned}\]</span></p>
<p>where the <span class="math inline">\(Q\)</span> are probabilities of identity in state, <span class="math inline">\(Q_1\)</span> among genes (gametes) within individuals, <span class="math inline">\(Q_2\)</span> among genes in different individuals within groups (populations), and <span class="math inline">\(Q_3\)</span> among groups (populations). Such formulas appear in <span class="citation">Cockerham and Weir (1987)</span>; see <span class="citation">François Rousset (2002a)</span> for an account of most implications of such definitions, except estimation.</p>
<p>The commonly held idea that it is more difficult to estimate <span class="math inline">\(F\)</span>-statistics when there are more alleles is generally incorrect; actually many inferences may be more accurate when more alleles are present <span class="citation">(e.g. Leblois, Estoup, and Rousset 2003, at least as long as gene diversity is less than 0.8)</span>. The issue is not to estimate the frequencies of all alleles, but only to estimate the above ratios. Any expression of the form <span class="math inline">\((Q_i-Q_j)/(1-Q_j)\)</span> can be estimated as <span class="math inline">\((\hat{Q}_i-\hat{Q}_j)/(1-\hat{Q}_j)\)</span> where any <span class="math inline">\(\hat{Q}_k\)</span> is the observed frequency of identical pairs of genes in the sample, among pairs satisfying the condition designated by the <span class="math inline">\(k\)</span> index. This is only slightly different <span class="citation">(see Rousset 2007)</span> from what the following estimators achieve.</p>
<div id="Fmulti" class="section level3">
<h3><span class="header-section-number">7.6.1</span> ANOVA estimators: single- and multilocus definitions</h3>
<p>Well-known work by Cockerham <span class="citation">(e.g. Cockerham 1973; Weir and Cockerham 1984)</span> has used the formalism of analysis of variance (ANOVA) to define estimators of <span class="math inline">\(F\)</span>-statistics. These estimators may be expressed in terms of the mean sums of squares <span class="math inline">\(MSG\)</span>, <span class="math inline">\(MSI\)</span>, <span class="math inline">\(MSP\)</span> (for Gametes, Individuals, and Populations) computed by an analysis of variance <span class="citation">(see e.g. Weir 1996)</span>. Equivalently, they can be expressed in terms of “components of variances” <span class="math inline">\(\hat{\sigma}^2_G\)</span>, <span class="math inline">\(\hat{\sigma}^2_I\)</span>, <span class="math inline">\(\hat{\sigma}^2_P\)</span> which are unbiased estimates of the corresponding parametric “components of variances” <span class="math inline">\(\sigma^2_G\)</span>, <span class="math inline">\(\sigma^2_I\)</span>, <span class="math inline">\(\sigma^2_P\)</span> in an ANOVA model. The snag is, in general (and in some notable applications), these parametric “components of variance” are not variances but rather differences between variances and can be negative. The <span class="math inline">\(\sigma^2\)</span> notation is misleading in this respect; this is a lasting source of confusion, explained in <span class="citation">Rousset (2007)</span>. Of course, the <span class="math inline">\(\hat{\sigma}^2\)</span> estimators can be negative even if the <span class="math inline">\(\sigma^2\)</span> parameters are positive, but this is a distinct issue.</p>
<p>The mean squares can themselves be interpreted in terms of observed frequencies <span class="math inline">\(\hat{Q}\)</span> of identical pairs of genes in the sample. For balanced samples, the relationships are simple:</p>
<p><span class="math inline">\(1-\hat{Q}_1=MSG\equiv \hat{\sigma}^2_G\)</span>, <span class="math inline">\(\hat{Q}_1-\hat{Q}_2=(MSI-MSG)/2\equiv \hat{\sigma}^2_I\)</span> and <span class="math inline">\(\hat{Q}_2-\hat{Q}_3=(MSP-MSI)/(2n)\equiv \hat{\sigma}^2_P\)</span> where <span class="math inline">\(n\)</span> is group size. Hence the single-group (single-population) <span class="math inline">\(F_\mathrm{IS}\)</span> estimator is</p>
<p><span class="math display">\[\label{}
    \frac{\hat{Q}_1-\hat{Q}_2}{1-\hat{Q}_2}=
    \frac{MSI-MSG}{MSI+MSG}=
    \frac{\hat{\sigma}^2_I}{\hat{\sigma}^2_I+\hat{\sigma}^2_G}.\]</span></p>
<p>For unbalanced groups (“populations” of unequal size), estimates over several groups are complex weighted averages of observed frequencies of identical pairs of genes within groups, not detailed here <span class="citation">(see Rousset 2007)</span>. However, ANOVA expressions still satisfy <span class="math inline">\(MSG\equiv \hat{\sigma}^2_G\)</span> and <span class="math inline">\((MSI-MSG)/2\equiv \hat{\sigma}^2_I\)</span>, and <span class="math inline">\((MSP-MSI)/(2n_c)\equiv \hat{\sigma}^2_P\)</span> where <span class="math inline">\(n_c\)</span> is a function of the size of each group (<span class="math inline">\(n_c\equiv [S_1-S_2/S_1]/(n-1)\)</span>, where <span class="math inline">\(S_1\)</span> is the total sample size, <span class="math inline">\(S_2\)</span> is the sum of squared group sizes, and <span class="math inline">\(n\)</span> is the number of non-empty groups). Then</p>
<p><span class="math display">\[\begin{gathered}
    \hat{F}_{\mathrm{IS}}=    \frac{MSI-MSG}{MSI+MSG}=
    \frac{\hat{\sigma}^2_I}{\hat{\sigma}^2_I+\hat{\sigma}^2_G}, \\
        \hat{F}_{\mathrm{ST}}=    \frac{MSP-MSI}{MSP+(n_c-1)MSI+n_cMSG}=
    \frac{\hat{\sigma}^2_P}{\hat{\sigma}^2_P+\hat{\sigma}^2_I+\hat{\sigma}^2_G}, \\
        \hat{F}_{\mathrm{IT}}=    \frac{MSP+(n_c-1)MSI-n_cMSG}{MSP+(n_c-1)MSI+n_cMSG}=
    \frac{\hat{\sigma}^2_P+\hat{\sigma}^2_I}{\hat{\sigma}^2_P+\hat{\sigma}^2_I+\hat{\sigma}^2_G}.\end{gathered}\]</span></p>
<p>With several loci, such an analysis is performed for each locus <span class="math inline">\(i\)</span> and the multilocus estimate is the ratio of a weighted sum of the above locus-specific numerators over locus-specific denominators. However, there is no single consistent way to compute the weighted sums. <span class="citation">Weir and Cockerham (1984)</span>’s multilocus estimators are defined in terms of intermediate statistics <span class="math inline">\(a\)</span>, <span class="math inline">\(b\)</span>, and <span class="math inline">\(c\)</span> for each locus, which appear to be the <span class="math inline">\(\hat{\sigma}^2\)</span>’s. The numerator of the multilocus estimator of <span class="math inline">\(F_\mathrm{ST}\)</span> is thus <span class="math inline">\(\sum_{\textrm{loci }i}a_i=\sum_{i}[(MSP-MSI)/(2n_c)]_i\)</span>. On the other hand <span class="citation">(Weir 1996’s)</span> multilocus estimators are defined from distinct intermediate statistics <span class="math inline">\(S_1\)</span>, <span class="math inline">\(S_2\)</span>, and <span class="math inline">\(S_3\)</span> for each locus, where for locus <span class="math inline">\(i\)</span>, <span class="math inline">\(S_{1i}=[(MSP-MSI)]_i/(2\bar{n})\)</span> for an average sample size across loci <span class="math inline">\(\bar{n}\)</span>, and the numerator of the multilocus estimate is <span class="math inline">\(\sum_{\textrm{loci }i}S_i=\sum_{i}[a n_c]_i/\bar{n}\)</span>. Hence the 1984 and 1996 estimators slightly differ.</p>
<p>However, both give the same weight to the estimates of the <span class="math inline">\(Q\)</span>’s for a locus typed at 5 individuals in each subpopulation as for a locus typed at 50 individuals in each subpopulation. Genepop follows another logic. The multilocus estimator of <span class="math inline">\(F_\mathrm{ST}\)</span> has numerator <span class="math inline">\(\sum_i [n_c(MSP-MSI)]_i\)</span>, which will give 10 time more weight to the <span class="math inline">\(Q\)</span> estimates for the more intensively typed locus. ‘Explicit’ formulas for the estimators are:</p>
<p><span class="math display">\[\begin{gathered}
    \hat{F}_{\mathrm{IS}}=    \frac{\sum_i [n_c(MSI-MSG)]_i}{\sum_i [n_c(MSI+MSG)]_i}=
    \frac{\sum_i [n_c\hat{\sigma}^2_I]_i}{\sum_i [n_c\hat{\sigma}^2_I+n_c\hat{\sigma}^2_G]_i}, \\
        \hat{F}_{\mathrm{ST}}=    \frac{\sum_i [MSP-MSI]_i}{\sum_i [MSP+(n_c-1)MSI+n_cMSG]_i}=
    \frac{\sum_i [n_c\hat{\sigma}^2_P]_i}{\sum_i [n_c\hat{\sigma}^2_P+n_c\hat{\sigma}^2_I+n_c\hat{\sigma}^2_G]_i}, \\
        \hat{F}_{\mathrm{IT}}=    \frac{\sum_i [MSP+(n_c-1)MSI-n_cMSG]_i}{\sum_i [MSP+(n_c-1)MSI+n_cMSG]_i}=
    \frac{\sum_i [n_c\hat{\sigma}^2_P+n_c\hat{\sigma}^2_I]_i}{\sum_i [n_c\hat{\sigma}^2_P+n_c\hat{\sigma}^2_I+n_c\hat{\sigma}^2_G]_i}.\end{gathered}\]</span></p>
<p>Data from the example file <code>Fmulti.txt</code> (3 samples, 3 loci) illustrate the difference between results obtained by the different methods:</p>
<table>
<thead>
<tr class="header">
<th>Estimate</th>
<th><span class="math inline">\(F_\mathrm{IS}\)</span></th>
<th><span class="math inline">\(F_\mathrm{ST}\)</span></th>
<th><span class="math inline">\(F_\mathrm{IT}\)</span></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>locus 1</td>
<td>-0.0483</td>
<td>0.5712</td>
<td>0.5505</td>
</tr>
<tr class="even">
<td>locus 2</td>
<td>-0.1161</td>
<td>0.8560</td>
<td>0.8393</td>
</tr>
<tr class="odd">
<td>locus 3</td>
<td>0.0051</td>
<td>-0.0023</td>
<td>0.0028</td>
</tr>
<tr class="even">
<td>Multilocus (1984 a,b,c method)</td>
<td>-0.0286</td>
<td>0.5606</td>
<td>0.5480</td>
</tr>
<tr class="odd">
<td>Multilocus (1996 S1,S2,S3 method)</td>
<td>-0.0286</td>
<td>0.5633</td>
<td>0.5508</td>
</tr>
<tr class="even">
<td>Multilocus (Genepop v3.3 and later)</td>
<td>-0.0275</td>
<td>0.5436</td>
<td>0.5310</td>
</tr>
</tbody>
</table>
<p>Most of the time the different estimators yield close values; I expect the Genepop method to provide better <span class="math inline">\(F_\mathrm{ST}\)</span> estimates under weak differentiation.</p>
</div>
<div id="rho-stats" class="section level3">
<h3><span class="header-section-number">7.6.2</span> Microsatellite allele sizes, <span class="math inline">\(R_\mathrm{ST}\)</span>, and <span class="math inline">\(\rho_\mathrm{ST}\)</span></h3>
<p> Following <span class="citation">Slatkin (1995)</span>, statistics based on allele size have been widely used. The parameters <span class="math inline">\(\rho_\mathrm{IS}\)</span>, <span class="math inline">\(\rho_\mathrm{ST}\)</span> and <span class="math inline">\(\rho_\mathrm{IT}\)</span> and their estimators are defined by replacing any <span class="math inline">\(1-Q_k\)</span> by the expected square difference in allele size between the genes compared <span class="citation">(Rousset 1996)</span> in all formulas above, and any <span class="math inline">\(1-\hat{Q}_k\)</span> by the observed mean square difference <span class="citation">(more formulas are given in Michalakis and Excoffier 1996)</span>. Then the estimators become plain ANOVA estimators of intraclass correlation for allele size; if there are only two alleles, <span class="math inline">\(\hat{\rho}_{\mathrm{ST}}=\hat{F}_{\mathrm{ST}}\)</span>, but Slatkin’s <span class="math inline">\(R_{\mathrm{ST}}\neq\hat{F}_{\mathrm{ST}}\)</span>.</p>
</div>
<div id="robertson-and-hills-estimator-of-f_mathrmis" class="section level3">
<h3><span class="header-section-number">7.6.3</span> Robertson and Hill’s estimator of <span class="math inline">\(F_\mathrm{IS}\)</span></h3>
<p>This estimator, reported in options 1 and 5, was designed to have lower variance than the ANOVA estimator and no small-sample bias when <span class="math inline">\(F_\mathrm{IS}\)</span> is low, assuming that deviations from Hardy-Weinberg proportions are characterized by the same <span class="math inline">\(F_\mathrm{IS}\)</span> for all pairs of alleles <span class="citation">(Robertson and Hill 1984)</span>. The score test computed in heterozygote excess and deficiency sub-options of option 1 is equivalent to this estimator for testing purposes.</p>
</div>
</div>
<div id="bootstraps" class="section level2">
<h2><span class="header-section-number">7.7</span> Bootstraps</h2>
<p> Option 6 constructs approximate bootstrap confidence (ABC) intervals <span class="citation">(DiCiccio and Efron 1996)</span>, assuming that each locus is an independent realization of genealogical and mutation processes. The bootstrap is a general methodology with different incarnations. The ABC methods were chosen because they balance moderate computation needs with good accuracy compared to alternatives. Bootstrap methods are approximate, and simulation tests of their performance (a too rare deed in statistical population genetics) for the present application are reported in <span class="citation">Leblois, Estoup, and Rousset (2003)</span> and <span class="citation">Watts et al. (2007)</span>.</p>
<p>The ABC method is also applied over individuals in option 8 to compute confidence intervals for null allele frequency estimates.</p>
</div>
<div id="mantel-test" class="section level2">
<h2><span class="header-section-number">7.8</span> Mantel test</h2>
<p> The principle of the Mantel permutation procedure is to permute samples between geographical locations, so it generates a distribution conditional on having <span class="math inline">\(n\)</span> given sets of genotypic data in <span class="math inline">\(n\)</span> different samples. The permutations provide the distribution of any statistic under the null hypothesis of independence between the two variables (here, genotype counts and geographic location).</p>
<p><span class="citation">Mantel (1967)</span> considered a particular statistics and approximations for its distribution. Instead, Genepop uses no such approximation. Isolation by distance will generate positive correlations between geographic distance and genetic distance estimates, and this is best tested using one-tailed P-values. The program provides both one-tailed P-values. The probability of observing the sample correlation is the sum of these two P-values minus 1.</p>
<div id="misuse-1-tests-of-correlation-at-different-distance" class="section level3">
<h3><span class="header-section-number">7.8.1</span> Misuse 1: tests of correlation at different distance</h3>
<p>Genetic processes of isolation by distance generate asymptotically decreasing variation in genetic differentiation with increasing geographic distances, and there is some temptation to use the Mantel test to test for the presence of correlation at specific distances. However, Genepop prevents this as this is logically unsound, and the more quantitative methods it provides are better suited to address variation of patterns with distance.</p>
<p>As soon as a process generates data with an expected non-zero correlation at some distance, it contradicts the null hypothesis under which the Mantel test is an exact test. Thus it may not make sense to use a Mantel test for testing correlation at some distance if there is correlation at another distance.</p>
<p>One can still wonder whether a permutation-based test could have some approximate validity for testing absence of correlation at some distance. However, the bootstrap procedure already addresses this case. Alternative procedures would require further definition on an ad-hoc basis to be operational (e.g., the idea of eliminating <em>samples</em> that form <em>pairs</em> below or above a given distance may not unambiguously define a <em>sample</em> selection procedure that will retain power) and would be likely to generate some confusion.</p>
<p>For these reasons, in the present implementation the Mantel tests are always based on all pairs, ignoring all selection of data according to distance.</p>
</div>
<div id="misuse-2-partial-mantel-tests" class="section level3">
<h3><span class="header-section-number">7.8.2</span> Misuse 2: partial Mantel tests</h3>
<p>Partial Mantel tests have been used to test for effects of a variable Y on a response variable Z, while supposedly removing spatial autocorrelation effects on Z. Both standard theory of exact tests <span class="citation">(as used by Raufaste and Rousset 2001)</span> and simulation <span class="citation">(Oden and Sokal 1992; Raufaste and Rousset 2001; François Rousset 2002b; Guillot and Rousset 2013)</span> show that the permutation procedure of the Mantel test is not appropriate for the partial Mantel test when the Y variable itself presents spatial correlations. Asymptotic arguments have also been proposed to support the use of such permutation tests <span class="citation">(e.g. Anderson 2001)</span> but they fail in the same conditions. As shown by <span class="citation">Raufaste and Rousset (2001)</span>, the problem is inherent to the permutation procedure, not to a specific test statistic. Unfortunately, some papers maintain confusion about these different aspects of “partial Mantel tests”. <span class="citation">Legendre and Fortin (2010)</span> argued how miserable the papers by <span class="citation">Raufaste and Rousset (2001)</span> and <span class="citation">François Rousset (2002b)</span> were, and claimed that some versions of the tests should be preferred because they used pivotal statistics (without evidence that the statistics were indeed pivotal, a property that depends on the statistical model). <span class="citation">Guillot and Rousset (2013)</span> reviewed old and more recent literature demonstrating issues with the partial Mantel test, provided new simulations showing that the different tests discussed by <span class="citation">Legendre and Fortin (2010)</span> failed, and criticized their verbal arguments. Despite this, <span class="citation">Legendre, Fortin, and Borcard (2015)</span> criticized this more recent paper again for ignoring the old literature, and repeated the same kind of verbal explanations that have previously failed.</p>
</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="evaluating-the-performance-of-inferences-for-isolation-by-distance.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="code-maintenance-credits-contact-etc-.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
},
"search": false
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:" && /^https?:/.test(src))
      src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
