<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>3 The input file | Genepop version 4.7.5</title>
  <meta name="description" content="3 The input file | Genepop version 4.7.5" />
  <meta name="generator" content="bookdown 0.10 and GitBook 2.6.7" />

  <meta property="og:title" content="3 The input file | Genepop version 4.7.5" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="3 The input file | Genepop version 4.7.5" />
  
  
  

<meta name="author" content="F. Rousset" />



  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="installing-genepop-and-session-examples.html">
<link rel="next" href="sec-settings.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />










</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="1" data-path="introduction.html"><a href="introduction.html"><i class="fa fa-check"></i><b>1</b> Introduction</a><ul>
<li class="chapter" data-level="1.1" data-path="introduction.html"><a href="introduction.html#purpose"><i class="fa fa-check"></i><b>1.1</b> Purpose</a></li>
<li class="chapter" data-level="1.2" data-path="introduction.html"><a href="introduction.html#the-two-genepop-distributions"><i class="fa fa-check"></i><b>1.2</b> The two Genepop distributions</a></li>
<li class="chapter" data-level="1.3" data-path="introduction.html"><a href="introduction.html#changes-since-version-4.0"><i class="fa fa-check"></i><b>1.3</b> Changes since version 4.0</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="installing-genepop-and-session-examples.html"><a href="installing-genepop-and-session-examples.html"><i class="fa fa-check"></i><b>2</b> Installing Genepop and session examples</a><ul>
<li class="chapter" data-level="2.1" data-path="installing-genepop-and-session-examples.html"><a href="installing-genepop-and-session-examples.html#installation"><i class="fa fa-check"></i><b>2.1</b> Installation</a><ul>
<li class="chapter" data-level="2.1.1" data-path="installing-genepop-and-session-examples.html"><a href="installing-genepop-and-session-examples.html#r-package"><i class="fa fa-check"></i><b>2.1.1</b> R package</a></li>
<li class="chapter" data-level="2.1.2" data-path="installing-genepop-and-session-examples.html"><a href="installing-genepop-and-session-examples.html#stand-alone-executable"><i class="fa fa-check"></i><b>2.1.2</b> Stand-alone executable</a></li>
</ul></li>
<li class="chapter" data-level="2.2" data-path="installing-genepop-and-session-examples.html"><a href="installing-genepop-and-session-examples.html#example-sessions"><i class="fa fa-check"></i><b>2.2</b> Example sessions</a><ul>
<li class="chapter" data-level="2.2.1" data-path="installing-genepop-and-session-examples.html"><a href="installing-genepop-and-session-examples.html#example-1-basic-session"><i class="fa fa-check"></i><b>2.2.1</b> Example 1: basic session</a></li>
<li class="chapter" data-level="2.2.2" data-path="installing-genepop-and-session-examples.html"><a href="installing-genepop-and-session-examples.html#example-2-using-the-settings-file"><i class="fa fa-check"></i><b>2.2.2</b> Example 2: using the settings file</a></li>
<li class="chapter" data-level="2.2.3" data-path="installing-genepop-and-session-examples.html"><a href="installing-genepop-and-session-examples.html#example-3-batch-processing"><i class="fa fa-check"></i><b>2.2.3</b> Example 3: Batch processing</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="the-input-file.html"><a href="the-input-file.html"><i class="fa fa-check"></i><b>3</b> The input file</a></li>
<li class="chapter" data-level="4" data-path="sec-settings.html"><a href="sec-settings.html"><i class="fa fa-check"></i><b>4</b> The settings file and command line arguments</a></li>
<li class="chapter" data-level="5" data-path="all-menu-options.html"><a href="all-menu-options.html"><i class="fa fa-check"></i><b>5</b> All menu options</a><ul>
<li class="chapter" data-level="5.1" data-path="all-menu-options.html"><a href="all-menu-options.html#option-1-hardy-weinberg-hw-exact-tests"><i class="fa fa-check"></i><b>5.1</b> Option 1: Hardy-Weinberg (HW) exact tests</a><ul>
<li class="chapter" data-level="5.1.1" data-path="all-menu-options.html"><a href="all-menu-options.html#sub-options-13-tests-for-each-locus-in-each-population"><i class="fa fa-check"></i><b>5.1.1</b> Sub-options 1–3: Tests for each locus in each population</a></li>
<li class="chapter" data-level="5.1.2" data-path="all-menu-options.html"><a href="all-menu-options.html#output"><i class="fa fa-check"></i><b>5.1.2</b> Output</a></li>
<li class="chapter" data-level="5.1.3" data-path="all-menu-options.html"><a href="all-menu-options.html#sub-options-45-global-tests-across-loci-or-across-samples"><i class="fa fa-check"></i><b>5.1.3</b> Sub-options 4,5: Global tests across loci or across samples</a></li>
<li class="chapter" data-level="5.1.4" data-path="all-menu-options.html"><a href="all-menu-options.html#analyzing-a-single-genotypic-matrix"><i class="fa fa-check"></i><b>5.1.4</b> Analyzing a single genotypic matrix</a></li>
<li class="chapter" data-level="5.1.5" data-path="all-menu-options.html"><a href="all-menu-options.html#code-checks"><i class="fa fa-check"></i><b>5.1.5</b> Code checks</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="all-menu-options.html"><a href="all-menu-options.html#option-2-tests-and-tables-for-linkage-disequilibrium"><i class="fa fa-check"></i><b>5.2</b> Option 2: Tests and tables for linkage disequilibrium</a><ul>
<li class="chapter" data-level="5.2.1" data-path="all-menu-options.html"><a href="all-menu-options.html#sub-option-1-tests"><i class="fa fa-check"></i><b>5.2.1</b> Sub-option 1: Tests</a></li>
<li class="chapter" data-level="5.2.2" data-path="all-menu-options.html"><a href="all-menu-options.html#output-1"><i class="fa fa-check"></i><b>5.2.2</b> Output</a></li>
<li class="chapter" data-level="5.2.3" data-path="all-menu-options.html"><a href="all-menu-options.html#sub-option-2-create-tables"><i class="fa fa-check"></i><b>5.2.3</b> Sub-option 2: create tables</a></li>
<li class="chapter" data-level="5.2.4" data-path="all-menu-options.html"><a href="all-menu-options.html#code-checks-1"><i class="fa fa-check"></i><b>5.2.4</b> Code checks</a></li>
</ul></li>
<li class="chapter" data-level="5.3" data-path="all-menu-options.html"><a href="all-menu-options.html#option-3-population-differentiation"><i class="fa fa-check"></i><b>5.3</b> Option 3: population differentiation</a><ul>
<li class="chapter" data-level="5.3.1" data-path="all-menu-options.html"><a href="all-menu-options.html#sub-options-1-or-2-genic-differentiation"><i class="fa fa-check"></i><b>5.3.1</b> Sub-options 1 or 2 (genic differentiation)</a></li>
<li class="chapter" data-level="5.3.2" data-path="all-menu-options.html"><a href="all-menu-options.html#sub-options-3-or-4-genotypic-differentiation"><i class="fa fa-check"></i><b>5.3.2</b> Sub-options 3 or 4 (genotypic differentiation)</a></li>
<li class="chapter" data-level="5.3.3" data-path="all-menu-options.html"><a href="all-menu-options.html#output-2"><i class="fa fa-check"></i><b>5.3.3</b> Output</a></li>
<li class="chapter" data-level="5.3.4" data-path="all-menu-options.html"><a href="all-menu-options.html#gene-diversity-as-a-test-statistic"><i class="fa fa-check"></i><b>5.3.4</b> Gene diversity as a test statistic</a></li>
<li class="chapter" data-level="5.3.5" data-path="all-menu-options.html"><a href="all-menu-options.html#analyzing-a-single-contingency-table"><i class="fa fa-check"></i><b>5.3.5</b> Analyzing a single contingency table</a></li>
<li class="chapter" data-level="5.3.6" data-path="all-menu-options.html"><a href="all-menu-options.html#code-checks-2"><i class="fa fa-check"></i><b>5.3.6</b> Code checks</a></li>
</ul></li>
<li class="chapter" data-level="5.4" data-path="all-menu-options.html"><a href="all-menu-options.html#option-4-private-alleles"><i class="fa fa-check"></i><b>5.4</b> Option 4: private alleles</a></li>
<li class="chapter" data-level="5.5" data-path="all-menu-options.html"><a href="all-menu-options.html#option-5-basic-information-f_mathrmis-and-gene-diversities"><i class="fa fa-check"></i><b>5.5</b> Option 5: Basic information, <span class="math inline">\(F_\mathrm{IS}\)</span>, and gene diversities</a><ul>
<li class="chapter" data-level="5.5.1" data-path="all-menu-options.html"><a href="all-menu-options.html#sub-option-1-allele-and-genotype-frequencies"><i class="fa fa-check"></i><b>5.5.1</b> Sub-option 1: Allele and genotype frequencies</a></li>
<li class="chapter" data-level="5.5.2" data-path="all-menu-options.html"><a href="all-menu-options.html#sub-option-2-identity-based-gene-diversities-and-f_mathrmis"><i class="fa fa-check"></i><b>5.5.2</b> Sub-option 2: Identity-based gene diversities and <span class="math inline">\(F_\mathrm{IS}\)</span></a></li>
<li class="chapter" data-level="5.5.3" data-path="all-menu-options.html"><a href="all-menu-options.html#sub-option-3-allele-size-based-gene-diversities-and-rho_mathrmis"><i class="fa fa-check"></i><b>5.5.3</b> Sub-option 3: Allele size-based gene diversities and <span class="math inline">\(\rho_{\mathrm{IS}}\)</span></a></li>
</ul></li>
<li class="chapter" data-level="5.6" data-path="all-menu-options.html"><a href="all-menu-options.html#option-6-fst-and-other-correlations-isolation-by-distance"><i class="fa fa-check"></i><b>5.6</b> Option 6: Fst and other correlations, isolation by distance</a><ul>
<li class="chapter" data-level="5.6.1" data-path="all-menu-options.html"><a href="all-menu-options.html#sub-options-14-f-statistics-and-rho-statistics"><i class="fa fa-check"></i><b>5.6.1</b> Sub-options 1–4: <span class="math inline">\(F\)</span>-statistics and <span class="math inline">\(\rho\)</span>-statistics</a></li>
<li class="chapter" data-level="5.6.2" data-path="all-menu-options.html"><a href="all-menu-options.html#sub-option-5-isolation-by-distance-between-individuals"><i class="fa fa-check"></i><b>5.6.2</b> Sub-option 5: isolation by distance between individuals</a></li>
<li class="chapter" data-level="5.6.3" data-path="all-menu-options.html"><a href="all-menu-options.html#sub-option-6-isolation-by-distance-between-groups"><i class="fa fa-check"></i><b>5.6.3</b> Sub-option 6: isolation by distance between groups</a></li>
<li class="chapter" data-level="5.6.4" data-path="all-menu-options.html"><a href="all-menu-options.html#former-sub-option-5-of-genepop-analysis-of-isolation-by-distance-from-a-genetic-distance-matrix"><i class="fa fa-check"></i><b>5.6.4</b> Former sub-option 5 of Genepop: analysis of isolation by distance from a genetic distance matrix</a></li>
<li class="chapter" data-level="5.6.5" data-path="all-menu-options.html"><a href="all-menu-options.html#user-provided-geographic-distance-matrices"><i class="fa fa-check"></i><b>5.6.5</b> User-provided geographic distance matrices</a></li>
<li class="chapter" data-level="5.6.6" data-path="all-menu-options.html"><a href="all-menu-options.html#analysis-of-isolation-by-distance-from-multiple-genetic-distance-matrices"><i class="fa fa-check"></i><b>5.6.6</b> Analysis of isolation by distance from multiple genetic distance matrices</a></li>
<li class="chapter" data-level="5.6.7" data-path="all-menu-options.html"><a href="all-menu-options.html#analysis-of-mean-differentiation"><i class="fa fa-check"></i><b>5.6.7</b> Analysis of mean differentiation</a></li>
</ul></li>
<li class="chapter" data-level="5.7" data-path="all-menu-options.html"><a href="all-menu-options.html#data-selection-for-analyses-of-isolation-by-distance"><i class="fa fa-check"></i><b>5.7</b> Data selection for analyses of isolation by distance</a><ul>
<li class="chapter" data-level="5.7.1" data-path="all-menu-options.html"><a href="all-menu-options.html#selecting-a-subset-of-samples"><i class="fa fa-check"></i><b>5.7.1</b> Selecting a subset of samples</a></li>
</ul></li>
<li class="chapter" data-level="5.8" data-path="all-menu-options.html"><a href="all-menu-options.html#option-7-file-conversions"><i class="fa fa-check"></i><b>5.8</b> Option 7: File conversions</a></li>
<li class="chapter" data-level="5.9" data-path="all-menu-options.html"><a href="all-menu-options.html#option-8-null-alleles-and-some-input-file-utilities"><i class="fa fa-check"></i><b>5.9</b> Option 8: Null alleles and some input file utilities</a><ul>
<li class="chapter" data-level="5.9.1" data-path="all-menu-options.html"><a href="all-menu-options.html#sub-option-1-null-alleles"><i class="fa fa-check"></i><b>5.9.1</b> Sub-option 1: null alleles</a></li>
<li class="chapter" data-level="5.9.2" data-path="all-menu-options.html"><a href="all-menu-options.html#sub-option-2-diploidisation-of-haploid-data"><i class="fa fa-check"></i><b>5.9.2</b> Sub-option 2: Diploidisation of haploid data</a></li>
<li class="chapter" data-level="5.9.3" data-path="all-menu-options.html"><a href="all-menu-options.html#sub-option-3-relabeling-alleles-names"><i class="fa fa-check"></i><b>5.9.3</b> Sub-option 3: Relabeling alleles names</a></li>
<li class="chapter" data-level="5.9.4" data-path="all-menu-options.html"><a href="all-menu-options.html#sub-options-4-and-5-conversion-of-population-data-to-individual-data"><i class="fa fa-check"></i><b>5.9.4</b> Sub-options 4 and 5: Conversion of population data to individual data</a></li>
<li class="chapter" data-level="5.9.5" data-path="all-menu-options.html"><a href="all-menu-options.html#sub-option-6-random-sampling-of-haploid-genotypes-from-diploid-ones"><i class="fa fa-check"></i><b>5.9.5</b> Sub-option 6: Random sampling of haploid genotypes from diploid ones</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="evaluating-the-performance-of-inferences-for-isolation-by-distance.html"><a href="evaluating-the-performance-of-inferences-for-isolation-by-distance.html"><i class="fa fa-check"></i><b>6</b> Evaluating the performance of inferences for Isolation by distance</a></li>
<li class="chapter" data-level="7" data-path="methods.html"><a href="methods.html"><i class="fa fa-check"></i><b>7</b> Methods</a><ul>
<li class="chapter" data-level="7.1" data-path="methods.html"><a href="methods.html#null-alleles"><i class="fa fa-check"></i><b>7.1</b> Null alleles</a></li>
<li class="chapter" data-level="7.2" data-path="methods.html"><a href="methods.html#exact-tests"><i class="fa fa-check"></i><b>7.2</b> Exact tests</a></li>
<li class="chapter" data-level="7.3" data-path="methods.html"><a href="methods.html#algorithms-for-exact-tests"><i class="fa fa-check"></i><b>7.3</b> Algorithms for exact tests</a></li>
<li class="chapter" data-level="7.4" data-path="methods.html"><a href="methods.html#accuracy-of-p-values-estimated-by-the-markov-chain-algorithms"><i class="fa fa-check"></i><b>7.4</b> Accuracy of P values estimated by the Markov chain algorithms</a></li>
<li class="chapter" data-level="7.5" data-path="methods.html"><a href="methods.html#test-statistics"><i class="fa fa-check"></i><b>7.5</b> Test statistics</a></li>
<li class="chapter" data-level="7.6" data-path="methods.html"><a href="methods.html#estimating-f-statistics-and-related-quantities"><i class="fa fa-check"></i><b>7.6</b> Estimating <span class="math inline">\(F\)</span>-statistics and related quantities</a><ul>
<li class="chapter" data-level="7.6.1" data-path="methods.html"><a href="methods.html#Fmulti"><i class="fa fa-check"></i><b>7.6.1</b> ANOVA estimators: single- and multilocus definitions</a></li>
<li class="chapter" data-level="7.6.2" data-path="methods.html"><a href="methods.html#rho-stats"><i class="fa fa-check"></i><b>7.6.2</b> Microsatellite allele sizes, <span class="math inline">\(R_\mathrm{ST}\)</span>, and <span class="math inline">\(\rho_\mathrm{ST}\)</span></a></li>
<li class="chapter" data-level="7.6.3" data-path="methods.html"><a href="methods.html#robertson-and-hills-estimator-of-f_mathrmis"><i class="fa fa-check"></i><b>7.6.3</b> Robertson and Hill’s estimator of <span class="math inline">\(F_\mathrm{IS}\)</span></a></li>
</ul></li>
<li class="chapter" data-level="7.7" data-path="methods.html"><a href="methods.html#bootstraps"><i class="fa fa-check"></i><b>7.7</b> Bootstraps</a></li>
<li class="chapter" data-level="7.8" data-path="methods.html"><a href="methods.html#mantel-test"><i class="fa fa-check"></i><b>7.8</b> Mantel test</a><ul>
<li class="chapter" data-level="7.8.1" data-path="methods.html"><a href="methods.html#misuse-1-tests-of-correlation-at-different-distance"><i class="fa fa-check"></i><b>7.8.1</b> Misuse 1: tests of correlation at different distance</a></li>
<li class="chapter" data-level="7.8.2" data-path="methods.html"><a href="methods.html#misuse-2-partial-mantel-tests"><i class="fa fa-check"></i><b>7.8.2</b> Misuse 2: partial Mantel tests</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="8" data-path="code-maintenance-credits-contact-etc-.html"><a href="code-maintenance-credits-contact-etc-.html"><i class="fa fa-check"></i><b>8</b> Code maintenance, credits, contact, etc.</a><ul>
<li class="chapter" data-level="8.1" data-path="code-maintenance-credits-contact-etc-.html"><a href="code-maintenance-credits-contact-etc-.html#code-maintenance"><i class="fa fa-check"></i><b>8.1</b> Code maintenance</a></li>
<li class="chapter" data-level="8.2" data-path="code-maintenance-credits-contact-etc-.html"><a href="code-maintenance-credits-contact-etc-.html#credits-for-the-current-version"><i class="fa fa-check"></i><b>8.2</b> Credits for the current version</a></li>
<li class="chapter" data-level="8.3" data-path="code-maintenance-credits-contact-etc-.html"><a href="code-maintenance-credits-contact-etc-.html#previous-history"><i class="fa fa-check"></i><b>8.3</b> Previous history</a></li>
<li class="chapter" data-level="8.4" data-path="code-maintenance-credits-contact-etc-.html"><a href="code-maintenance-credits-contact-etc-.html#contact"><i class="fa fa-check"></i><b>8.4</b> Contact</a><ul>
<li class="chapter" data-level="8.4.1" data-path="code-maintenance-credits-contact-etc-.html"><a href="code-maintenance-credits-contact-etc-.html#bug-fixes-since-release-of-genepop-version-3.4-in-may-2003-until-first-release-of-genepop-4.0"><i class="fa fa-check"></i><b>8.4.1</b> Bug fixes since release of Genepop version 3.4 in May 2003 until first release of Genepop 4.0:</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="9" data-path="copyright.html"><a href="copyright.html"><i class="fa fa-check"></i><b>9</b> Copyright</a></li>
<li class="chapter" data-level="" data-path="bibliography.html"><a href="bibliography.html"><i class="fa fa-check"></i>Bibliography</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Genepop version 4.7.5</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="the-input-file" class="section level1">
<h1><span class="header-section-number">3</span> The input file</h1>
<p> As illustrated by the following examples, the input format requested by Genepop is:</p>
<ul>
<li><strong>First line: anything</strong> Use this line to store information about your data.</li>
<li><strong>Locus names</strong> They may be given one per line, or on the same line but separated by commas. <code>Pop</code> sample indicator (Capitalization does not matter)<a href="#fn2" class="footnoteRef" id="fnref2"><sup>2</sup></a>. Each sample from a different geographical original is declared by a line with a <code>pop</code> statement.</li>
<li><strong>Information for first individual.</strong> An example is: <code>ind#001 fem  ,0101 0202 0000 0410</code> Here <code>ind#001 fem</code> is an identifier for your personal use. You can use any character (except a comma!). You may leave it blank (at least one space) if you wish. The last identifier of every sub-population is used by Genepop as the sample name in output files. The comma between the identifier and the list of genotypes is required. <code>0101</code> indicates that this individual is homozygous for the <code>01</code> allele at the first locus. The same individual is homozygous for the <code>02</code> allele at the second locus (<code>0202</code>). Data are missing at the third locus (<code>0000</code>). At the fourth locus, the genotype is <code>0410</code>, which indicates the presence of alleles <code>04</code> and <code>10</code>.</li>
<li><strong>More individuals</strong> Each individual information starts on a new line, but may extend over several lines (do not start a new line in the middle of a one-locus genotype!).</li>
<li><strong>More samples</strong> each declared by a <code>pop</code> statement on a new line</li>
<li><strong>Blank lines</strong> at the end of the file are removed by Genepop.</li>
</ul>
<p>An example of a short input file is given below:</p>
<pre><code> Title line: &quot;Grape populations in southern France&quot;
 ADH Locus 1
 ADH #2
 ADH three
 ADH-4
 ADH-5
 mtDNA
 Pop
 Grange des Peres  ,  0201 003003 0102 0302 1011 01
 Grange des Peres  ,  0202 003001 0102 0303 1111 01
 Grange des Peres  ,  0102 004001 0202 0102 1010 01
 Grange des Peres  ,  0103 002002 0101 0202 1011 01
 Grange des Peres  ,  0203 002004 0101 0102 1010 01
 POP
 Tertre Roteboeuf ,      0102 002002 0201 0405 0807 01
 Tertre Roteboeuf ,      0102 002001 0201 0405 0307 01
 Tertre Roteboeuf ,      0201 002003 0101 0505 0402 01
 Tertre Roteboeuf ,      0201 003003 0301 0303 0603 01
 Tertre Roteboeuf ,      0101 002001 0301 0505 0807 01
 pop
 Bonneau 01   , 0101    002002 0304 0805 0304 01
 Bonneau 02   , 0201    002002 0404 0505 0304 01
 Bonneau 03   , 0101    002100 0304 0505 0101 01
 Bonneau 04 , 0101    100100 0204 0805 0304 01
 Bonneau 05   , 0101    100002 0104 0808 0304 01
 Pop
  ,            0000 002001 0202 0402 0007 01
  ,            0200 002001 0202 0205 0707 01
  ,            0010 002001 0101
 0105 0807 01
 last pop,      0101 002001 0101 0401 0807 02</code></pre>
<p>This example shows some useful features of the input file:</p>
<ul>
<li><p>There is no constraint on the number of blanks separating the various fields.</p></li>
<li><p>The individual identifier has a free format.</p></li>
<li><p>Alleles are numbered from 01 to 99 or 001 to 999 if needed.  In 3-digits coding,  (say) homozygotes for the <code>90</code> allele are noted <code>090090</code>, not <code>9090</code> as in the 2-digits format. 2-digits and 3-digits coding of alleles can be intermixed (among loci, not within loci!).<a href="#fn3" class="footnoteRef" id="fnref3"><sup>3</sup></a></p></li>
<li><p>To designate alleles, consecutive numbers are not required.</p></li>
<li><p>haploid and diploid data can be intermixed.<a href="#fn4" class="footnoteRef" id="fnref4"><sup>4</sup></a> 6-digits genotypes are recognized as 3-digits diploid genotypes; 4-digits genotypes are recognized as 2-digits diploid genotypes; 2- and 3-digits genotypes are recognized as haploid genotypes. The same coding should be used consistently within each locus. See the <code>EstimationPloidy</code> setting for more information about analyzing haploid data. For haplo-diploid data at a given locus, the haploid genotypes should be coded as diploid genotypes with one unknown allele; note however that the information from haploid genotypes at haplo-diploid loci will be used only for genic contingency table tests, and will be ignored in estimation of genetic structure.</p></li>
<li><p>Genotypes can extend on more than one line (see penultimate individual)</p></li>
<li><p>To group various samples, just remove each relevant <code>Pop</code> separator.</p></li>
</ul>
<p>It is possible to write all the locus names on one line, provided that a comma is used as separator. This could be useful to clearly label each column. Thus the above input file could have started as</p>
<pre><code> Title line: &quot;Grape populations in southern France&quot;
                      Loc1,Loc2,  ADH3,ADH4,ADH5,mtDNA
 Pop
 Grange des Peres  ,  0201 003003 0102 0302 1011 01
 ...</code></pre>
<p>Note the absence of comma after the last locus name.</p>
<p>There are however constraints to be obeyed</p>
<ul>
<li><p>Missing data should be indicated with <code>00</code> (or <code>000</code> for 3-digits coding) and not with blanks. The first locus in the last sample illustrates the various possibilities of missing data: no information (first individual coded <code>0000</code>) or partial information (only one allele is determined: allele <code>02</code> for the second individual coded <code>0200</code> and allele <code>10</code> for the third individual coded <code>0010</code>).</p></li>
<li><p>The number of locus names should correspond to the number of genotypes in each individual. If you remove one or several loci from your input file, you should remove both their names and the corresponding genotypes.</p></li>
<li><p>No empty line should be present in the data file.</p></li>
<li><p>Genepop accepts input file names either with the extension <code>.txt</code><a href="#fn5" class="footnoteRef" id="fnref5"><sup>5</sup></a> or without any extension.</p></li>
<li><p>Genepop input files are ASCII text files.</p></li>
</ul>
<p>The last point implies that under <strong>Windows</strong>, you should avoid using Microsoft Word to edit input files (and settings files as well). Rather use a text editor such as <a href="http://notepad-plus-plus.org">Notepad++</a>.<a href="#fn6" class="footnoteRef" id="fnref6"><sup>6</sup></a> It has also appeared that certain Microsoft products under <strong>Mac OS X</strong> still produced files formatted according to the older Mac format. Genepop now catches and corrects this miserable feature.</p>
<p>One can also find some conversion tools (e.g. from EXCEL) on the web.</p>
<p>If the input file is correctly read, the name of the larger allele number is indicated for each locus. The number of distinct alleles for each locus is provided upon request. If alleles have been labeled with consecutive numbers from <code>01</code> onwards, then the name of the larger allele will correspond to the number of distinct alleles for each locus.</p>
<p>There are some limits to the number of samples and individuals imposed by the compiler. These values, and a few other ones, are shown by running “<code>Genepop Maxima=</code>” (see the <code>Maxima</code> setting). However, these built-in maxima are so large<a href="#fn7" class="footnoteRef" id="fnref7"><sup>7</sup></a> as to be practically infinite even in the era of whole-genome sequencing. Computer memory, or user patience, are more likely limits.</p>
</div>
<div class="footnotes">
<hr />
<ol start="2">
<li id="fn2"><p>Earlier versions of Genepop only accepted <code>Pop</code>, <code>POP</code> and <code>pop</code>…<a href="the-input-file.html#fnref2">↩</a></p></li>
<li id="fn3"><p>New to Genepop 4.0<a href="the-input-file.html#fnref3">↩</a></p></li>
<li id="fn4"><p>Also new to Genepop 4.0<a href="the-input-file.html#fnref4">↩</a></p></li>
<li id="fn5"><p>New to Genepop 4.0<a href="the-input-file.html#fnref5">↩</a></p></li>
<li id="fn6"><p>Other text editors including the Windows basic text editor may not show all end-of-line characters correctly.<a href="the-input-file.html#fnref6">↩</a></p></li>
<li id="fn7"><p>in constrast to earlier versions of Genepop<a href="the-input-file.html#fnref7">↩</a></p></li>
</ol>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="installing-genepop-and-session-examples.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="sec-settings.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
},
"search": false
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:" && /^https?:/.test(src))
      src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
