<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />



<title>Using GxEScanR</title>

<script>// Hide empty <a> tag within highlighted CodeBlock for screen reader accessibility (see https://github.com/jgm/pandoc/issues/6352#issuecomment-626106786) -->
// v0.0.1
// Written by JooYoung Seo (jooyoung@psu.edu) and Atsushi Yasumoto on June 1st, 2020.

document.addEventListener('DOMContentLoaded', function() {
  const codeList = document.getElementsByClassName("sourceCode");
  for (var i = 0; i < codeList.length; i++) {
    var linkList = codeList[i].getElementsByTagName('a');
    for (var j = 0; j < linkList.length; j++) {
      if (linkList[j].innerHTML === "") {
        linkList[j].setAttribute('aria-hidden', 'true');
      }
    }
  }
});
</script>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
code.sourceCode > span { display: inline-block; line-height: 1.25; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>



<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Using GxEScanR</h1>


<div id="TOC">
<ul>
<li><a href="#introduction">Introduction</a></li>
<li><a href="#example-files">Example Files</a></li>
<li><a href="#examples">Examples</a><ul>
<li><a href="#linear-regression-gwas">Linear Regression GWAS</a></li>
<li><a href="#linear-regression-gweis">Linear Regression GWEIS</a></li>
<li><a href="#logistic-regression-gwas">Logistic Regression GWAS</a></li>
<li><a href="#logistic-regression-gweis">Logistic Regression GWEIS</a><ul>
<li><a href="#logistic-regression-gweis-with-binary-covariate">Logistic Regression GWEIS with Binary Covariate</a></li>
<li><a href="#logistic-regression-gweis-with-a-continuous-covariate">Logistic Regression GWEIS with a Continuous Covariate</a></li>
</ul></li>
</ul></li>
</ul>
</div>

<p>This vignette shows some examples of using GxEScanR to perform genome-wide association study (GWAS) and genome-wide by environment interaction study (GWEIS) scans using all the options available to the user.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a><span class="kw">library</span>(GxEScanR)</span></code></pre></div>
<div id="introduction" class="section level1">
<h1>Introduction</h1>
<p>With growing number of SNPs that can be imputed it is necessary to have software that can efficiently perform GWAS and GWEIS scans. GxEScanR can do this using files that were saved in the BinaryDosage format. The BinaryDosage package can convert VCF and GEN files into the BinaryDosage format. The BinaryDosage format was designed to keep the file with the genetic data small with fast read times. GxEScanR uses this and efficient large scale regression routines to perform GWAS and GWEIS scans quickly.</p>
</div>
<div id="example-files" class="section level1">
<h1>Example Files</h1>
<p>The examples below use three sample files. The first contains a data frame that has subject data. The second file is a genetic data file in the BinaryDosage format. The last file contains the information returned by the BinaryDosage::getbdinfo routine that returns information about the binary dosage file that makes reading it fast.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1"></a>covdatafile &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;covdata.rds&quot;</span>, <span class="dt">package =</span> <span class="st">&quot;GxEScanR&quot;</span>)</span>
<span id="cb2-2"><a href="#cb2-2"></a>covdata &lt;-<span class="st"> </span><span class="kw">readRDS</span>(covdatafile)</span></code></pre></div>
<table>
<caption>First 5 Subjects</caption>
<thead>
<tr class="header">
<th align="left">sid</th>
<th align="right">y</th>
<th align="right">e</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">I1</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="left">I2</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
<tr class="odd">
<td align="left">I3</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="left">I4</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
<tr class="odd">
<td align="left">I5</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
</tbody>
</table>
<p>To load the binary dosage information file, it is necessary to update the file name since the file has been moved from its original location during the installation process. The following loads the binary dosage information and updates the file name.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1"></a>bdinfofile &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;pdata_4_1.bdinfo&quot;</span>, <span class="dt">package =</span> <span class="st">&quot;GxEScanR&quot;</span>)</span>
<span id="cb3-2"><a href="#cb3-2"></a>bdinfo &lt;-<span class="st"> </span><span class="kw">readRDS</span>(bdinfofile)</span>
<span id="cb3-3"><a href="#cb3-3"></a>bdinfo<span class="op">$</span>filename &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;pdata_4_1.bdose&quot;</span>, <span class="dt">package =</span> <span class="st">&quot;GxEScanR&quot;</span>)</span></code></pre></div>
<table>
<caption>Models Fit</caption>
<thead>
<tr class="header">
<th align="left">model</th>
<th align="left">outcome</th>
<th align="left">predictors</th>
<th align="left">subjects</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">D|E,G</td>
<td align="left">Phenotype</td>
<td align="left">All covariates, gene</td>
<td align="left">All</td>
</tr>
<tr class="even">
<td align="left">D|E,G,GxE</td>
<td align="left">Phenotype</td>
<td align="left">All covariates, gene, gene x last covariate</td>
<td align="left">All</td>
</tr>
<tr class="odd">
<td align="left">E|G</td>
<td align="left">Last covariate</td>
<td align="left">All other covariates, gene</td>
<td align="left">All</td>
</tr>
<tr class="even">
<td align="left">E|G,D=1, case only</td>
<td align="left">Last covariate</td>
<td align="left">All other covariates, gene</td>
<td align="left">Cases</td>
</tr>
<tr class="odd">
<td align="left">E|G,D=0 control only</td>
<td align="left">Last covariate</td>
<td align="left">All other covariates, gene</td>
<td align="left">Controls</td>
</tr>
</tbody>
</table>
</div>
<div id="examples" class="section level1">
<h1>Examples</h1>
<div id="linear-regression-gwas" class="section level2">
<h2>Linear Regression GWAS</h2>
<p>The simplest scan to do is a linear regression GWAS. The following model is first when doing a linear regression GWAS.</p>
<table>
<caption>Model Fit</caption>
<thead>
<tr class="header">
<th align="left">model</th>
<th align="left">outcome</th>
<th align="left">predictors</th>
<th align="left">subjects</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">D|E,G</td>
<td align="left">Phenotype</td>
<td align="left">All covariates, gene</td>
<td align="left">All</td>
</tr>
</tbody>
</table>
<p>In the example data set, the phenotype, y, is coded 0,1. When GxEScanR sees the phenotype codes this way it assumes the outcome is binary and uses logistic regression. To perform a linear regression GWAS the binary option needs to be set to FALSE. The following shows how to do a linear GWAS along with the results.</p>
<p>The routine outputs the number of subjects used in the analysis and returns a data frame with the results.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1"></a>lingwas1 &lt;-<span class="st"> </span><span class="kw">gwas</span>(<span class="dt">data =</span> covdata,</span>
<span id="cb4-2"><a href="#cb4-2"></a>                 <span class="dt">bdinfo =</span> bdinfo,</span>
<span id="cb4-3"><a href="#cb4-3"></a>                 <span class="dt">binary =</span> <span class="ot">FALSE</span>)</span>
<span id="cb4-4"><a href="#cb4-4"></a><span class="co">#&gt; [1] &quot;200 subjects have complete data&quot;</span></span></code></pre></div>
<table>
<caption>Linear Regression GWAS</caption>
<thead>
<tr class="header">
<th align="left">snp</th>
<th align="right">betag</th>
<th align="right">lrtg</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">1:10001</td>
<td align="right">0.1605708</td>
<td align="right">9.0532054</td>
</tr>
<tr class="even">
<td align="left">1:10002</td>
<td align="right">-0.0220380</td>
<td align="right">0.1219035</td>
</tr>
<tr class="odd">
<td align="left">1:10003</td>
<td align="right">-0.0732687</td>
<td align="right">1.3812032</td>
</tr>
<tr class="even">
<td align="left">1:10004</td>
<td align="right">0.0326160</td>
<td align="right">0.2355224</td>
</tr>
<tr class="odd">
<td align="left">1:10005</td>
<td align="right">0.0627953</td>
<td align="right">1.2072107</td>
</tr>
</tbody>
</table>
<p>The output can be redirected to output file that produces a plain test version of the results in a tab delimited file that can be read into R using the read.table routine. In this case, the gwas routine returns a value of 0.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a>outfile &lt;-<span class="st"> </span><span class="kw">tempfile</span>()</span>
<span id="cb5-2"><a href="#cb5-2"></a>lingwas2 &lt;-<span class="st"> </span><span class="kw">gwas</span>(<span class="dt">data =</span> covdata,</span>
<span id="cb5-3"><a href="#cb5-3"></a>                 <span class="dt">bdinfo =</span> bdinfo,</span>
<span id="cb5-4"><a href="#cb5-4"></a>                 <span class="dt">outfile =</span> outfile,</span>
<span id="cb5-5"><a href="#cb5-5"></a>                 <span class="dt">binary =</span> <span class="ot">FALSE</span>)</span>
<span id="cb5-6"><a href="#cb5-6"></a><span class="co">#&gt; [1] &quot;200 subjects have complete data&quot;</span></span>
<span id="cb5-7"><a href="#cb5-7"></a>lingwas2</span>
<span id="cb5-8"><a href="#cb5-8"></a><span class="co">#&gt; [1] 0</span></span>
<span id="cb5-9"><a href="#cb5-9"></a>lingwas2 &lt;-<span class="st"> </span><span class="kw">read.table</span>(outfile, <span class="dt">header =</span> <span class="ot">TRUE</span>, <span class="dt">sep =</span><span class="st">&#39;</span><span class="ch">\t</span><span class="st">&#39;</span>)</span></code></pre></div>
<table>
<caption>Linear Regression GWAS</caption>
<thead>
<tr class="header">
<th align="left">SNPID</th>
<th align="right">betag</th>
<th align="right">lrtg</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">1:10001</td>
<td align="right">0.1605708</td>
<td align="right">9.0532054</td>
</tr>
<tr class="even">
<td align="left">1:10002</td>
<td align="right">-0.0220380</td>
<td align="right">0.1219035</td>
</tr>
<tr class="odd">
<td align="left">1:10003</td>
<td align="right">-0.0732687</td>
<td align="right">1.3812032</td>
</tr>
<tr class="even">
<td align="left">1:10004</td>
<td align="right">0.0326160</td>
<td align="right">0.2355224</td>
</tr>
<tr class="odd">
<td align="left">1:10005</td>
<td align="right">0.0627953</td>
<td align="right">1.2072107</td>
</tr>
</tbody>
</table>
</div>
<div id="linear-regression-gweis" class="section level2">
<h2>Linear Regression GWEIS</h2>
<p>The gweis routine takes the same parameters as the gwas function but performs additional tests. The models fit for a linear regression GWAS are.</p>
<table>
<caption>Models Fit</caption>
<thead>
<tr class="header">
<th align="left">model</th>
<th align="left">outcome</th>
<th align="left">predictors</th>
<th align="left">subjects</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">D|E,G</td>
<td align="left">Phenotype</td>
<td align="left">All covariates, gene</td>
<td align="left">All</td>
</tr>
<tr class="even">
<td align="left">D|E,G,GxE</td>
<td align="left">Phenotype</td>
<td align="left">All covariates, gene, gene x last covariate</td>
<td align="left">All</td>
</tr>
</tbody>
</table>
<p>Note: When doing a GWEIS the interaction covariate is in the last column of the subject data frame.</p>
<p>In this test the minmaf option was used. When minmaf is specified the minor allele for a SNP must exceed minmaf to be test. Notice that only 5 SNPs are in the output data frame. This is because one of the SNPs has a minor allele frequency below 0.2.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1"></a>lingweis1 &lt;-<span class="st"> </span><span class="kw">gweis</span>(<span class="dt">data =</span> covdata,</span>
<span id="cb6-2"><a href="#cb6-2"></a>                   <span class="dt">bdinfo =</span> bdinfo,</span>
<span id="cb6-3"><a href="#cb6-3"></a>                   <span class="dt">minmaf =</span> <span class="fl">0.2</span>,</span>
<span id="cb6-4"><a href="#cb6-4"></a>                   <span class="dt">binary =</span> <span class="ot">FALSE</span>)</span>
<span id="cb6-5"><a href="#cb6-5"></a><span class="co">#&gt; [1] &quot;200 subjects have complete data&quot;</span></span></code></pre></div>
<table>
<caption>Linear Regression GWEIS</caption>
<thead>
<tr class="header">
<th align="left">snp</th>
<th align="right">betadg</th>
<th align="right">lrtdg</th>
<th align="right">betagxe</th>
<th align="right">lrtgxe</th>
<th align="right">lrt2df</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">1:10001</td>
<td align="right">0.1605708</td>
<td align="right">9.0532054</td>
<td align="right">0.0777308</td>
<td align="right">0.5402899</td>
<td align="right">9.593495</td>
</tr>
<tr class="even">
<td align="left">1:10002</td>
<td align="right">-0.0220380</td>
<td align="right">0.1219035</td>
<td align="right">0.1484303</td>
<td align="right">1.3479068</td>
<td align="right">1.469810</td>
</tr>
<tr class="odd">
<td align="left">1:10004</td>
<td align="right">0.0326160</td>
<td align="right">0.2355224</td>
<td align="right">0.1968781</td>
<td align="right">2.0769149</td>
<td align="right">2.312437</td>
</tr>
<tr class="even">
<td align="left">1:10005</td>
<td align="right">0.0627953</td>
<td align="right">1.2072107</td>
<td align="right">-0.0442079</td>
<td align="right">0.1401332</td>
<td align="right">1.347344</td>
</tr>
</tbody>
</table>
<p>If the user is interested in see what happened to SNPs that werenâ€™t included in the data frame, the skipfile option can be used. The skipfile value is the name of a file to write the skipped SNPs to. The skipfile option can be used along with the outfile option. The skip file is in the same format as the output file. Below is an example using the skipfile option.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1"></a>skipfile =<span class="st"> </span><span class="kw">tempfile</span>()</span>
<span id="cb7-2"><a href="#cb7-2"></a>lingweis2 &lt;-<span class="st"> </span><span class="kw">gweis</span>(<span class="dt">data =</span> covdata,</span>
<span id="cb7-3"><a href="#cb7-3"></a>                   <span class="dt">bdinfo =</span> bdinfo,</span>
<span id="cb7-4"><a href="#cb7-4"></a>                   <span class="dt">skipfile =</span> skipfile,</span>
<span id="cb7-5"><a href="#cb7-5"></a>                   <span class="dt">minmaf =</span> <span class="fl">0.2</span>,</span>
<span id="cb7-6"><a href="#cb7-6"></a>                   <span class="dt">binary =</span> <span class="ot">FALSE</span>)</span>
<span id="cb7-7"><a href="#cb7-7"></a><span class="co">#&gt; [1] &quot;200 subjects have complete data&quot;</span></span></code></pre></div>
<table>
<caption>Linear Regression GWEIS</caption>
<thead>
<tr class="header">
<th align="left">snp</th>
<th align="right">betadg</th>
<th align="right">lrtdg</th>
<th align="right">betagxe</th>
<th align="right">lrtgxe</th>
<th align="right">lrt2df</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">1:10001</td>
<td align="right">0.1605708</td>
<td align="right">9.0532054</td>
<td align="right">0.0777308</td>
<td align="right">0.5402899</td>
<td align="right">9.593495</td>
</tr>
<tr class="even">
<td align="left">1:10002</td>
<td align="right">-0.0220380</td>
<td align="right">0.1219035</td>
<td align="right">0.1484303</td>
<td align="right">1.3479068</td>
<td align="right">1.469810</td>
</tr>
<tr class="odd">
<td align="left">1:10004</td>
<td align="right">0.0326160</td>
<td align="right">0.2355224</td>
<td align="right">0.1968781</td>
<td align="right">2.0769149</td>
<td align="right">2.312437</td>
</tr>
<tr class="even">
<td align="left">1:10005</td>
<td align="right">0.0627953</td>
<td align="right">1.2072107</td>
<td align="right">-0.0442079</td>
<td align="right">0.1401332</td>
<td align="right">1.347344</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1"></a>knitr<span class="op">::</span><span class="kw">kable</span>(skipsnps, <span class="dt">caption =</span> <span class="st">&quot;Skipped SNPs&quot;</span>)</span></code></pre></div>
<table>
<caption>Skipped SNPs</caption>
<thead>
<tr class="header">
<th align="left">SNPID</th>
<th align="right">reasondg</th>
<th align="right">reasongxe</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">1:10003</td>
<td align="right">2</td>
<td align="right">2</td>
</tr>
</tbody>
</table>
<p>The following table lists the reasons SNPs were skipped given the skipped value.</p>
<table>
<caption>Skipped Reasons</caption>
<thead>
<tr class="header">
<th align="right">code</th>
<th align="left">reason</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="right">1</td>
<td align="left">Excluded by user</td>
</tr>
<tr class="even">
<td align="right">2</td>
<td align="left">Minor allele frequency below threshold</td>
</tr>
<tr class="odd">
<td align="right">3</td>
<td align="left">X matrix singular</td>
</tr>
<tr class="even">
<td align="right">4</td>
<td align="left">Singular matrix updating beta top</td>
</tr>
<tr class="odd">
<td align="right">5</td>
<td align="left">Singular matrix updating beta bottom</td>
</tr>
<tr class="even">
<td align="right">6</td>
<td align="left">Maximum iterations exceeded</td>
</tr>
</tbody>
</table>
</div>
<div id="logistic-regression-gwas" class="section level2">
<h2>Logistic Regression GWAS</h2>
<p>In this example, the phenotype is coded (0,1). The gwas and gweis routines check for this an will run logistic regressions if the outcome is coded (0,1) unless binary is set to FALSE. If the use wants to make sure the outcome is coded (0,1), the user may set binary to TRUE. In this case, if the outcome is not coded (0,1) an error is produced.</p>
<p>The following model is fit when doing a logistic regression GWAS.</p>
<table>
<caption>Model Fit</caption>
<thead>
<tr class="header">
<th align="left">model</th>
<th align="left">outcome</th>
<th align="left">predictors</th>
<th align="left">subjects</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">D|E,G</td>
<td align="left">Phenotype</td>
<td align="left">All covariates, gene</td>
<td align="left">All</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1"></a>loggwas1 &lt;-<span class="st"> </span><span class="kw">gwas</span>(<span class="dt">data =</span> covdata,</span>
<span id="cb9-2"><a href="#cb9-2"></a>                 <span class="dt">bdinfo =</span> bdinfo,</span>
<span id="cb9-3"><a href="#cb9-3"></a>                 <span class="dt">blksize =</span> <span class="dv">2</span>,</span>
<span id="cb9-4"><a href="#cb9-4"></a>                 <span class="dt">binary =</span> <span class="ot">TRUE</span>)</span>
<span id="cb9-5"><a href="#cb9-5"></a><span class="co">#&gt; [1] &quot;200 subjects have complete data&quot;</span></span></code></pre></div>
<table>
<caption>Logistic Regression GWAS</caption>
<thead>
<tr class="header">
<th align="left">snp</th>
<th align="right">betag</th>
<th align="right">lrtg</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">1:10001</td>
<td align="right">0.7129</td>
<td align="right">9.0230</td>
</tr>
<tr class="even">
<td align="left">1:10002</td>
<td align="right">-0.0943</td>
<td align="right">0.1219</td>
</tr>
<tr class="odd">
<td align="left">1:10003</td>
<td align="right">-0.3144</td>
<td align="right">1.3777</td>
</tr>
<tr class="even">
<td align="left">1:10004</td>
<td align="right">0.1397</td>
<td align="right">0.2356</td>
</tr>
<tr class="odd">
<td align="left">1:10005</td>
<td align="right">0.2680</td>
<td align="right">1.2003</td>
</tr>
</tbody>
</table>
<p>In this example, the option blksize is used. When an analysis is run several SNPs are read in at one time. This saves disk time. The following are the default values for given the number of subjects. These values were chosen to keep the program running using less than 4GB of RAM. The user is allowed to specify a value up to twice the default value. Little performance gain is seen going with larger values. If the user enters 0 for blksize, the default value is used.</p>
<table>
<caption>Default blksize</caption>
<thead>
<tr class="header">
<th align="left">subjects</th>
<th align="right">blksize</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">less than 10,000</td>
<td align="right">5000</td>
</tr>
<tr class="even">
<td align="left">10,000 to 24,999</td>
<td align="right">2000</td>
</tr>
<tr class="odd">
<td align="left">25,000 to 49,999</td>
<td align="right">1000</td>
</tr>
<tr class="even">
<td align="left">50,000 to 99,999</td>
<td align="right">500</td>
</tr>
<tr class="odd">
<td align="left">100,000 to 249,999</td>
<td align="right">200</td>
</tr>
<tr class="even">
<td align="left">250,000 to 499,999</td>
<td align="right">100</td>
</tr>
<tr class="odd">
<td align="left">500,000 or greater</td>
<td align="right">50</td>
</tr>
</tbody>
</table>
</div>
<div id="logistic-regression-gweis" class="section level2">
<h2>Logistic Regression GWEIS</h2>
<p>A logistic regression GWEIS fits an additional 4 models that produce 7 more tests. 3 of these models use the the interaction covariate as the outcome. The following show all the models fit in a logistic regression GWEIS.</p>
<p>The following model is fit when doing a logistic regression GWAS.</p>
<table>
<caption>Models Fit</caption>
<thead>
<tr class="header">
<th align="left">model</th>
<th align="left">outcome</th>
<th align="left">predictors</th>
<th align="left">subjects</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">D|E,G</td>
<td align="left">Phenotype</td>
<td align="left">All covariates, gene</td>
<td align="left">All</td>
</tr>
<tr class="even">
<td align="left">D|E,G,GxE</td>
<td align="left">Phenotype</td>
<td align="left">All covariates, gene, gene x last covariate</td>
<td align="left">All</td>
</tr>
<tr class="odd">
<td align="left">E|G</td>
<td align="left">Last covariate</td>
<td align="left">All other covariates, gene</td>
<td align="left">All</td>
</tr>
<tr class="even">
<td align="left">E|G,D=1, case only</td>
<td align="left">Last covariate</td>
<td align="left">All other covariates, gene</td>
<td align="left">Cases</td>
</tr>
<tr class="odd">
<td align="left">E|G,D=0 control only</td>
<td align="left">Last covariate</td>
<td align="left">All other covariates, gene</td>
<td align="left">Controls</td>
</tr>
</tbody>
</table>
<p>Note: When doing a GWEIS the interaction covariate is in the last column of the data frame.</p>
<div id="logistic-regression-gweis-with-binary-covariate" class="section level3">
<h3>Logistic Regression GWEIS with Binary Covariate</h3>
<p>In the example subject data, the covariate is coded (0,1). In this case, the gweis routine will use logistic regression to fit the last 3 models.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1"></a>loggweis1 &lt;-<span class="st"> </span><span class="kw">gweis</span>(<span class="dt">data =</span> covdata,</span>
<span id="cb10-2"><a href="#cb10-2"></a>                   <span class="dt">bdinfo =</span> bdinfo,</span>
<span id="cb10-3"><a href="#cb10-3"></a>                   <span class="dt">snps =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">2</span>,</span>
<span id="cb10-4"><a href="#cb10-4"></a>                   <span class="dt">binary =</span> <span class="ot">TRUE</span>)</span>
<span id="cb10-5"><a href="#cb10-5"></a><span class="co">#&gt; [1] &quot;200 subjects have complete data&quot;</span></span></code></pre></div>
<table style="width:100%;">
<caption>Logistic Regression GWEIS</caption>
<colgroup>
<col width="7%"></col>
<col width="7%"></col>
<col width="6%"></col>
<col width="7%"></col>
<col width="6%"></col>
<col width="6%"></col>
<col width="6%"></col>
<col width="6%"></col>
<col width="7%"></col>
<col width="8%"></col>
<col width="7%"></col>
<col width="8%"></col>
<col width="7%"></col>
</colgroup>
<thead>
<tr class="header">
<th align="left">snp</th>
<th align="right">betadg</th>
<th align="right">lrtdg</th>
<th align="right">betagxe</th>
<th align="right">lrtgxe</th>
<th align="right">lrt2df</th>
<th align="right">betaeg</th>
<th align="right">lrteg</th>
<th align="right">lrt3df</th>
<th align="right">betacase</th>
<th align="right">lrtcase</th>
<th align="right">betactrl</th>
<th align="right">lrtctrl</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">1:10001</td>
<td align="right">0.7129</td>
<td align="right">9.0230</td>
<td align="right">0.4931</td>
<td align="right">0.9520</td>
<td align="right">9.9750</td>
<td align="right">0.4115</td>
<td align="right">3.3606</td>
<td align="right">13.3356</td>
<td align="right">0.4290</td>
<td align="right">2.0147</td>
<td align="right">-0.0477</td>
<td align="right">0.0147</td>
</tr>
<tr class="even">
<td align="left">1:10002</td>
<td align="right">-0.0943</td>
<td align="right">0.1219</td>
<td align="right">0.6508</td>
<td align="right">1.3569</td>
<td align="right">1.4789</td>
<td align="right">0.0798</td>
<td align="right">0.0910</td>
<td align="right">1.5699</td>
<td align="right">0.3882</td>
<td align="right">1.1018</td>
<td align="right">-0.2562</td>
<td align="right">0.3735</td>
</tr>
</tbody>
</table>
<p>In this example the snps options was used. The snps option can either be a vector of indices indicating what SNPs to include or a list of SNPs by SNP ID. A vector of indices was used in this example.</p>
</div>
<div id="logistic-regression-gweis-with-a-continuous-covariate" class="section level3">
<h3>Logistic Regression GWEIS with a Continuous Covariate</h3>
<p>In the example subject data, the covariate is coded (0,1) which the gweis routine sees a binary covariate to make the routine do a linear regression 1 can be added to the interaction covariate. This will change the coding to (1,2) which the routine sees as a continuous covariate.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1"></a>covdata2 &lt;-<span class="st"> </span>covdata</span>
<span id="cb11-2"><a href="#cb11-2"></a>covdata2<span class="op">$</span>e &lt;-<span class="st"> </span>covdata2<span class="op">$</span>e <span class="op">+</span><span class="st"> </span><span class="dv">1</span></span>
<span id="cb11-3"><a href="#cb11-3"></a>loggweis2 &lt;-<span class="st"> </span><span class="kw">gweis</span>(<span class="dt">data =</span> covdata2,</span>
<span id="cb11-4"><a href="#cb11-4"></a>                   <span class="dt">bdinfo =</span> bdinfo,</span>
<span id="cb11-5"><a href="#cb11-5"></a>                   <span class="dt">snps =</span> <span class="kw">c</span>(<span class="st">&quot;1:10001&quot;</span>, <span class="st">&quot;1:10002&quot;</span>))</span>
<span id="cb11-6"><a href="#cb11-6"></a><span class="co">#&gt; [1] &quot;200 subjects have complete data&quot;</span></span></code></pre></div>
<table style="width:100%;">
<caption>Logistic Regression GWEIS</caption>
<colgroup>
<col width="7%"></col>
<col width="7%"></col>
<col width="6%"></col>
<col width="7%"></col>
<col width="6%"></col>
<col width="6%"></col>
<col width="6%"></col>
<col width="6%"></col>
<col width="7%"></col>
<col width="8%"></col>
<col width="7%"></col>
<col width="8%"></col>
<col width="7%"></col>
</colgroup>
<thead>
<tr class="header">
<th align="left">snp</th>
<th align="right">betadg</th>
<th align="right">lrtdg</th>
<th align="right">betagxe</th>
<th align="right">lrtgxe</th>
<th align="right">lrt2df</th>
<th align="right">betaeg</th>
<th align="right">lrteg</th>
<th align="right">lrt3df</th>
<th align="right">betacase</th>
<th align="right">lrtcase</th>
<th align="right">betactrl</th>
<th align="right">lrtctrl</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">1:10001</td>
<td align="right">0.7129</td>
<td align="right">9.0230</td>
<td align="right">0.4931</td>
<td align="right">0.9520</td>
<td align="right">9.9750</td>
<td align="right">0.1001</td>
<td align="right">3.4031</td>
<td align="right">13.3781</td>
<td align="right">0.1044</td>
<td align="right">2.0128</td>
<td align="right">-0.0098</td>
<td align="right">0.0147</td>
</tr>
<tr class="even">
<td align="left">1:10002</td>
<td align="right">-0.0943</td>
<td align="right">0.1219</td>
<td align="right">0.6508</td>
<td align="right">1.3569</td>
<td align="right">1.4789</td>
<td align="right">0.0194</td>
<td align="right">0.0912</td>
<td align="right">1.5700</td>
<td align="right">0.0947</td>
<td align="right">1.0967</td>
<td align="right">-0.0516</td>
<td align="right">0.3688</td>
</tr>
</tbody>
</table>
<p>In this example the snps options was used with the SNP IDs.</p>
</div>
</div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
